{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2b39090e",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-09-17T15:24:46.939359Z",
     "iopub.status.busy": "2025-09-17T15:24:46.938763Z",
     "iopub.status.idle": "2025-09-17T15:24:57.092400Z",
     "shell.execute_reply": "2025-09-17T15:24:57.091499Z"
    },
    "papermill": {
     "duration": 10.172254,
     "end_time": "2025-09-17T15:24:57.094996",
     "exception": false,
     "start_time": "2025-09-17T15:24:46.922742",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m→\u001b[0m connected lamindb: jkobject/scprint2\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/pasteur/appa/homes/jkalfon/simpler_flash/src/simpler_flash/layer_norm.py:1044: FutureWarning: `torch.cuda.amp.custom_fwd(args...)` is deprecated. Please use `torch.amp.custom_fwd(args..., device_type='cuda')` instead.\n",
      "  @custom_fwd\n",
      "/pasteur/appa/homes/jkalfon/simpler_flash/src/simpler_flash/layer_norm.py:1107: FutureWarning: `torch.cuda.amp.custom_bwd(args...)` is deprecated. Please use `torch.amp.custom_bwd(args..., device_type='cuda')` instead.\n",
      "  @custom_bwd\n"
     ]
    }
   ],
   "source": [
    "import scanpy as sc\n",
    "from scprint2 import scPRINT2\n",
    "from scdataloader import Preprocessor\n",
    "from scdataloader.utils import load_genes\n",
    "import numpy as np\n",
    "import anndata as ad\n",
    "from huggingface_hub import hf_hub_download\n",
    "import lamindb as ln\n",
    "\n",
    "from scprint2.tasks import Embedder\n",
    "from scprint2.tasks.cell_emb import display_confusion_matrix\n",
    "import pandas as pd\n",
    "\n",
    "from scib_metrics.benchmark import Benchmarker, BioConservation, BatchCorrection\n",
    "from anndata import AnnData\n",
    "from scdataloader.utils import translate\n",
    "import bionty as bt\n",
    "from scprint2.tasks.cell_emb import compute_classification\n",
    "\n",
    "from lightning.pytorch import Trainer\n",
    "from lightning.pytorch.loggers import TensorBoardLogger\n",
    "from scdataloader import SimpleAnnDataset, Collator, DataModule\n",
    "from torch.utils.data import DataLoader\n",
    "\n",
    "import lamindb as ln\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import torch\n",
    "import scipy.sparse\n",
    "\n",
    "torch.set_float32_matmul_precision(\"medium\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0dbc0c0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "LOC = \"/pasteur/appa/scratch/jkalfon/data/spcrint_data/temp/\" #\"../../data/temp/\"  #"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f81ff650",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read(\n",
    "    LOC + \"glio_smart_cort_area\" + \".h5ad\",\n",
    "    backup_url=\"https://datasets.cellxgene.cziscience.com/a1d40c84-c81c-406f-bef4-e25edeb651e5.h5ad\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ca3f3df4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "1353725.5\n",
      "Dropping layers:  KeysView(Layers with keys: exon, intron)\n",
      "checking raw counts\n",
      "removed 0 non primary cells, 49417 renamining\n",
      "filtered out 0 cells, 49417 renamining\n",
      "Removed 0 genes not known to the ontology\n",
      "Removed 0 duplicate genes\n",
      "Added 34323 genes in the ontology but not present in the dataset\n",
      "starting QC\n",
      "Seeing 10180 outliers (20.60% of total dataset):\n",
      "done\n",
      "AnnData object with n_obs × n_vars = 49417 × 70116\n",
      "    obs: 'suspension_type', 'cluster', 'class', 'subclass', 'sex_ontology_term_id', 'region', 'cortical_layer', 'cell_type_accession', 'cell_type_alias', 'cell_type_alt_alias', 'cell_type_designation', 'donor_id', 'outlier_call', 'outlier_type', 'tissue_ontology_term_id', 'disease_ontology_term_id', 'assay_ontology_term_id', 'is_primary_data', 'cell_type_ontology_term_id', 'Specimen ID', 'sample_tissue_type', 'development_stage_ontology_term_id', 'self_reported_ethnicity_ontology_term_id', 'cause_of_death', 'PMI (hr)', 'Tissue (RIN)', 'Hemisphere Sampled', 'tissue_type', 'cell_type', 'assay', 'disease', 'sex', 'tissue', 'self_reported_ethnicity', 'development_stage', 'observation_joinid', 'organism_ontology_term_id', 'nnz', 'n_genes', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'pct_counts_in_top_20_genes', 'total_counts_mt', 'log1p_total_counts_mt', 'pct_counts_mt', 'total_counts_ribo', 'log1p_total_counts_ribo', 'pct_counts_ribo', 'total_counts_hb', 'log1p_total_counts_hb', 'pct_counts_hb', 'outlier', 'mt_outlier'\n",
      "    var: 'feature_is_filtered', 'feature_name', 'feature_reference', 'feature_biotype', 'feature_length', 'feature_type', 'uid', 'symbol', 'biotype', 'organism_id', 'branch_id', 'mt', 'ribo', 'hb', 'organism', 'ensembl_gene_id', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts'\n",
      "    uns: 'unseen_genes'\n"
     ]
    }
   ],
   "source": [
    "preprocessor = Preprocessor(\n",
    "    force_preprocess=True,\n",
    "    skip_validate=True,\n",
    "    # drop_non_primary=False,\n",
    "    do_postp=False,\n",
    ")\n",
    "print(\"\")\n",
    "print(adata.X.sum(1).mean())\n",
    "adata = preprocessor(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "be261ad9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# model_checkpoint_file = hf_hub_download(\n",
    "#    repo_id=\"jkobject/scPRINT\", filename=f\"v2-medium.ckpt\"\n",
    "# )\n",
    "# model_checkpoint_file = ../data/\n",
    "model_checkpoint_file = \"../models/ji9krimq.ckpt\"\n",
    "# w937u4o1\n",
    "# ji9krimq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "327b650f",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-09-17T15:24:57.213837Z",
     "iopub.status.busy": "2025-09-17T15:24:57.213211Z",
     "iopub.status.idle": "2025-09-17T15:27:47.486800Z",
     "shell.execute_reply": "2025-09-17T15:27:47.485977Z"
    },
    "papermill": {
     "duration": 170.289846,
     "end_time": "2025-09-17T15:27:47.489107",
     "exception": false,
     "start_time": "2025-09-17T15:24:57.199261",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FYI: scPrint is not attached to a `Trainer`.\n"
     ]
    }
   ],
   "source": [
    "model = scPRINT2.load_from_checkpoint(\n",
    "    model_checkpoint_file, precpt_gene_emb=None, gene_pos_file=None, max_cont_len=None,\n",
    ")\n",
    "model = model.to(\"cuda\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "38efc260",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "cell_type\n",
       "glutamatergic neuron                                              21841\n",
       "L2/3-6 intratelencephalic projecting glutamatergic neuron          4778\n",
       "VIP GABAergic cortical interneuron                                 3533\n",
       "pvalb GABAergic cortical interneuron                               2800\n",
       "L6 corticothalamic-projecting glutamatergic cortical neuron        2556\n",
       "lamp5 GABAergic cortical interneuron                               2434\n",
       "sst GABAergic cortical interneuron                                 2358\n",
       "unknown                                                            1985\n",
       "oligodendrocyte                                                    1930\n",
       "astrocyte                                                          1187\n",
       "L6b glutamatergic cortical neuron                                  1080\n",
       "near-projecting glutamatergic cortical neuron                       816\n",
       "oligodendrocyte precursor cell                                      773\n",
       "microglial cell                                                     750\n",
       "caudal ganglionic eminence derived cortical interneuron             325\n",
       "L5 extratelencephalic projecting glutamatergic cortical neuron      158\n",
       "endothelial cell                                                     70\n",
       "pericyte                                                             32\n",
       "vascular leptomeningeal cell                                         11\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs.cell_type.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "cc07e5ba",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-09-17T15:27:48.352264Z",
     "iopub.status.busy": "2025-09-17T15:27:48.351711Z",
     "iopub.status.idle": "2025-09-17T15:28:02.377363Z",
     "shell.execute_reply": "2025-09-17T15:28:02.376478Z"
    },
    "papermill": {
     "duration": 14.041652,
     "end_time": "2025-09-17T15:28:02.379841",
     "exception": false,
     "start_time": "2025-09-17T15:27:48.338189",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Training data: (35587, 70116)\n",
      "Validation data: (13830, 70116)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/pasteur/appa/homes/jkalfon/scdataloader/scdataloader/utils.py:427: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "  organismdf = pd.concat(organismdf)\n"
     ]
    }
   ],
   "source": [
    "# Prepare data for fine-tuning (using the cat/tiger dataset from above)\n",
    "# Split data into train/val\n",
    "\n",
    "train_ct = [\"oligodendrocyte\",\n",
    "            \"microglial cell\",\n",
    "            \"pericyte\",\n",
    "            \"lamp5 GABAergic cortical interneuron\",\n",
    "            \"L6b glutamatergic cortical neuron\",\n",
    "            \"astrocyte\",\n",
    "            \"VIP GABAergic cortical interneuron\",\n",
    "            \"glutamatergic neuron\",\n",
    "            \"pvalb GABAergic cortical interneuron\"]\n",
    "train_data = adata[adata.obs.cell_type.isin(train_ct)].copy()\n",
    "val_data = adata[~adata.obs.cell_type.isin(train_ct)].copy()\n",
    "\n",
    "print(f\"Training data: {train_data.shape}\")\n",
    "print(f\"Validation data: {val_data.shape}\")\n",
    "\n",
    "mencoders = {}\n",
    "for k, v in model.label_decoders.items():\n",
    "    mencoders[k] = {va: ke for ke, va in v.items()}\n",
    "# this needs to remain its original name as it is expect like that by collator, otherwise need to send org_to_id as params\n",
    "\n",
    "# Create datasets\n",
    "train_dataset = SimpleAnnDataset(\n",
    "    train_data,\n",
    "    obs_to_output=[\"cell_type_ontology_term_id\", \"organism_ontology_term_id\"],\n",
    "    get_knn_cells=model.expr_emb_style == \"metacell\",\n",
    "    encoder=mencoders,\n",
    ")\n",
    "\n",
    "# Create collator\n",
    "collator = Collator(\n",
    "    organisms=model.organisms,\n",
    "    valid_genes=model.genes,\n",
    "    class_names=[\"cell_type_ontology_term_id\", 'organism_ontology_term_id'],\n",
    "    how=\"random expr\",  # or \"all expr\" for full expression\n",
    "    max_len=6000,\n",
    "    add_zero_genes=0,\n",
    "    org_to_id=mencoders[\"organism_ontology_term_id\"],\n",
    ")\n",
    "\n",
    "# Create data loaders\n",
    "train_loader = DataLoader(\n",
    "    train_dataset,\n",
    "    collate_fn=collator,\n",
    "    batch_size=4,  # Adjust based on GPU memory\n",
    "    num_workers=4,\n",
    "    shuffle=True,\n",
    "    pin_memory=True,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "5544b3a4",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-09-17T15:28:02.504874Z",
     "iopub.status.busy": "2025-09-17T15:28:02.504582Z",
     "iopub.status.idle": "2025-09-17T15:28:02.571619Z",
     "shell.execute_reply": "2025-09-17T15:28:02.570813Z"
    },
    "papermill": {
     "duration": 0.083185,
     "end_time": "2025-09-17T15:28:02.573912",
     "exception": false,
     "start_time": "2025-09-17T15:28:02.490727",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "for val in model.transformer.blocks:\n",
    "    val.mixer.Wqkv.requires_grad = True\n",
    "for val in model.cell_transformer.blocks:\n",
    "    for i in val.cross_attn.parameters():\n",
    "        i.requires_grad = True\n",
    "#for val in model.gene_encoder[3].parameters():\n",
    "#    val.requires_grad = True\n",
    "#for val in model.metacell_encoder.parameters():\n",
    "#    val.requires_grad = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "0c28ce78",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-09-17T15:28:02.727698Z",
     "iopub.status.busy": "2025-09-17T15:28:02.727097Z",
     "iopub.status.idle": "2025-09-17T15:28:02.828638Z",
     "shell.execute_reply": "2025-09-17T15:28:02.827760Z"
    },
    "papermill": {
     "duration": 0.118252,
     "end_time": "2025-09-17T15:28:02.831046",
     "exception": false,
     "start_time": "2025-09-17T15:28:02.712794",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "def classif(batch, do_cls=False):\n",
    "    gene_pos = batch[\"genes\"].to(model.device)\n",
    "    expression = batch[\"x\"].to(model.device)\n",
    "    depth = batch[\"depth\"].to(model.device)\n",
    "    class_elem = batch[\"class\"].long().to(model.device)\n",
    "    total_loss = 0\n",
    "\n",
    "    # Forward pass with automatic mixed precisio^n\n",
    "    with torch.cuda.amp.autocast():\n",
    "        # Forward pass\n",
    "        output = model.forward(\n",
    "            gene_pos,\n",
    "            expression,\n",
    "            req_depth=depth,\n",
    "            depth_mult=expression.sum(1),\n",
    "            do_class=True,\n",
    "            metacell_token=torch.zeros_like(depth),\n",
    "        )\n",
    "        ## generate expr loss\n",
    "        output_gen = model._generate(\n",
    "            cell_embs=output[\"output_cell_embs\"],\n",
    "            gene_pos=gene_pos,\n",
    "            depth_mult=expression.sum(1),\n",
    "            req_depth=depth,\n",
    "        )\n",
    "        if \"zero_logits\" in output_gen:\n",
    "            loss_expr = loss.zinb(\n",
    "                theta=output_gen[\"disp\"],\n",
    "                pi=output_gen[\"zero_logits\"],\n",
    "                mu=output_gen[\"mean\"],\n",
    "                target=expression,\n",
    "            )\n",
    "            if model.zinb_and_mse:\n",
    "                loss_expr += (\n",
    "                    loss.mse(\n",
    "                        input=torch.log(output_gen[\"mean\"] + 1)\n",
    "                        * (1 - torch.sigmoid(output_gen[\"zero_logits\"])),\n",
    "                        target=torch.log(expression + 1),\n",
    "                    )\n",
    "                    / 10  # scale to make it more similar to the zinb\n",
    "                )\n",
    "        else:\n",
    "            loss_expr = loss.mse(\n",
    "                input=torch.log(output_gen[\"mean\"] + 1),\n",
    "                target=torch.log(expression + 1),\n",
    "            )\n",
    "        # Add expression loss to total\n",
    "        total_loss += loss_expr\n",
    "\n",
    "        # ct clss\n",
    "        if do_cls:\n",
    "            for i, k in enumerate([\"cell_type_ontology_term_id\", \"organism_ontology_term_id\"]):\n",
    "                cls_loss = loss.hierarchical_classification(\n",
    "                    pred=output.get(\"cls_output_\" + k),\n",
    "                    cl=class_elem[:, i],\n",
    "                    labels_hierarchy=model.mat_labels_hierarchy.get(\n",
    "                        k\n",
    "                ).to(\"cuda\") if k in model.mat_labels_hierarchy else None,\n",
    "            )\n",
    "            total_loss += cls_loss\n",
    "            return total_loss, loss_expr, cls_loss\n",
    "        # total_loss += output[\"vae_kl_loss\"] * 0.001\n",
    "    return total_loss, loss_expr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "771be887",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-09-17T15:28:02.859972Z",
     "iopub.status.busy": "2025-09-17T15:28:02.859561Z",
     "iopub.status.idle": "2025-09-17T15:28:02.952098Z",
     "shell.execute_reply": "2025-09-17T15:28:02.951179Z"
    },
    "papermill": {
     "duration": 0.108652,
     "end_time": "2025-09-17T15:28:02.954535",
     "exception": false,
     "start_time": "2025-09-17T15:28:02.845883",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/local/scratch/tmp/ipykernel_1386340/2614348468.py:19: FutureWarning: `torch.cuda.amp.GradScaler(args...)` is deprecated. Please use `torch.amp.GradScaler('cuda', args...)` instead.\n",
      "  scaler = torch.cuda.amp.GradScaler() if torch.cuda.is_available() else None\n"
     ]
    }
   ],
   "source": [
    "# Alternative: Manual Training Loop (for more control)\n",
    "# If you prefer to have more control over the training process\n",
    "from tqdm import tqdm\n",
    "import torch.nn.functional as F\n",
    "from scprint2.model import loss\n",
    "\n",
    "num_epochs = 1\n",
    "lr = 0.00002\n",
    "\n",
    "# Setup optimizer\n",
    "all_params = (\n",
    "    list(model.parameters())\n",
    ")\n",
    "optimizer = torch.optim.AdamW(\n",
    "    all_params, lr=lr, weight_decay=0.01, betas=(0.9, 0.999), eps=1e-8\n",
    ")\n",
    "\n",
    "# Setup automatic mixed precision\n",
    "scaler = torch.cuda.amp.GradScaler() if torch.cuda.is_available() else None\n",
    "\n",
    "for k, i in model.mat_labels_hierarchy.items():\n",
    "    model.mat_labels_hierarchy[k] = i.to(model.device)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a2dd5b2",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/pasteur/appa/homes/jkalfon/scdataloader/scdataloader/utils.py:427: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "  organismdf = pd.concat(organismdf)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "not on wandb, could not set name\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 0/79 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "> \u001b[32m<string>\u001b[39m(\u001b[92m29\u001b[39m)\u001b[36m_encoder\u001b[39m\u001b[34m()\u001b[39m\n",
      "\n"
     ]
    }
   ],
   "source": [
    "metrics = {}\n",
    "for n in [400, 2000, 6000]:\n",
    "    ad = train_data[(train_data.X > 0).sum(1) >= n]\n",
    "    # Get 5,000 random cells\n",
    "    n_cells = min(5000, ad.n_obs)  # In case you have fewer than 5,000 cells\n",
    "    random_indices = np.random.choice(ad.n_obs, n_cells, replace=False)\n",
    "    adata_subset = ad[random_indices, :].copy()\n",
    "    embed = Embedder(\n",
    "        how=\"random expr\",\n",
    "        max_len=n,\n",
    "        num_workers=8,\n",
    "        pred_embedding=[\"cell_type_ontology_term_id\"],\n",
    "        doplot=False,\n",
    "    )\n",
    "    _, metrics[n] = embed(model, adata_subset.copy())\n",
    "    \n",
    "metrics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a06917e6",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-09-29T09:52:20.405972Z",
     "iopub.status.busy": "2025-09-29T09:52:20.405829Z",
     "iopub.status.idle": "2025-09-29T10:08:55.187257Z",
     "shell.execute_reply": "2025-09-29T10:08:55.186662Z"
    },
    "papermill": {
     "duration": 994.793516,
     "end_time": "2025-09-29T10:08:55.193969",
     "exception": false,
     "start_time": "2025-09-29T09:52:20.400453",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Epoch 1/1\n",
      "Current learning rate: 2.00e-05\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Training:   0%|          | 0/8897 [00:00<?, ?it/s]/local/scratch/tmp/ipykernel_1386340/3266719282.py:9: FutureWarning: `torch.cuda.amp.autocast(args...)` is deprecated. Please use `torch.amp.autocast('cuda', args...)` instead.\n",
      "  with torch.cuda.amp.autocast():\n",
      "Training:  27%|██▋       | 2401/8897 [25:29<1:08:57,  1.57it/s]\n",
      "/pasteur/appa/homes/jkalfon/scdataloader/scdataloader/utils.py:427: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "  organismdf = pd.concat(organismdf)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "not on wandb, could not set name\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 79/79 [00:11<00:00,  6.60it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "logging the anndata\n",
      "AnnData object with n_obs × n_vars = 5000 × 17917\n",
      "    obs: 'pred_cell_type_ontology_term_id', 'pred_tissue_ontology_term_id', 'pred_disease_ontology_term_id', 'pred_age_group', 'pred_assay_ontology_term_id', 'pred_self_reported_ethnicity_ontology_term_id', 'pred_sex_ontology_term_id', 'pred_organism_ontology_term_id', 'pred_cell_culture', 'conv_pred_cell_type_ontology_term_id', 'conv_pred_tissue_ontology_term_id', 'conv_pred_disease_ontology_term_id', 'conv_pred_age_group', 'conv_pred_assay_ontology_term_id', 'conv_pred_self_reported_ethnicity_ontology_term_id'\n",
      "    obsm: 'scprint_emb_cell_type_ontology_term_id'\n",
      "    layers: 'scprint_mu', 'scprint_theta', 'scprint_pi'\n",
      "too few cells to embed into a umap\n",
      "too few cells to compute a clustering\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/pasteur/appa/homes/jkalfon/scdataloader/scdataloader/utils.py:427: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "  organismdf = pd.concat(organismdf)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "not on wandb, could not set name\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 79/79 [00:37<00:00,  2.08it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "logging the anndata\n",
      "AnnData object with n_obs × n_vars = 5000 × 18604\n",
      "    obs: 'pred_cell_type_ontology_term_id', 'pred_tissue_ontology_term_id', 'pred_disease_ontology_term_id', 'pred_age_group', 'pred_assay_ontology_term_id', 'pred_self_reported_ethnicity_ontology_term_id', 'pred_sex_ontology_term_id', 'pred_organism_ontology_term_id', 'pred_cell_culture', 'conv_pred_cell_type_ontology_term_id', 'conv_pred_tissue_ontology_term_id', 'conv_pred_disease_ontology_term_id', 'conv_pred_age_group', 'conv_pred_assay_ontology_term_id', 'conv_pred_self_reported_ethnicity_ontology_term_id'\n",
      "    obsm: 'scprint_emb_cell_type_ontology_term_id'\n",
      "    layers: 'scprint_mu', 'scprint_theta', 'scprint_pi'\n",
      "too few cells to embed into a umap\n",
      "too few cells to compute a clustering\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/pasteur/appa/homes/jkalfon/scdataloader/scdataloader/utils.py:427: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "  organismdf = pd.concat(organismdf)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "not on wandb, could not set name\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 79/79 [02:23<00:00,  1.81s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "logging the anndata\n",
      "AnnData object with n_obs × n_vars = 5000 × 20004\n",
      "    obs: 'pred_cell_type_ontology_term_id', 'pred_tissue_ontology_term_id', 'pred_disease_ontology_term_id', 'pred_age_group', 'pred_assay_ontology_term_id', 'pred_self_reported_ethnicity_ontology_term_id', 'pred_sex_ontology_term_id', 'pred_organism_ontology_term_id', 'pred_cell_culture', 'conv_pred_cell_type_ontology_term_id', 'conv_pred_tissue_ontology_term_id', 'conv_pred_disease_ontology_term_id', 'conv_pred_age_group', 'conv_pred_assay_ontology_term_id', 'conv_pred_self_reported_ethnicity_ontology_term_id'\n",
      "    obsm: 'scprint_emb_cell_type_ontology_term_id'\n",
      "    layers: 'scprint_mu', 'scprint_theta', 'scprint_pi'\n",
      "too few cells to embed into a umap\n",
      "too few cells to compute a clustering\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{400: {'cell_type_ontology_term_id_accuracy': 0.36425075739920765,\n",
       "  'tissue_ontology_term_id_accuracy': 0.0604,\n",
       "  'disease_ontology_term_id_accuracy': 0.933,\n",
       "  'assay_ontology_term_id_accuracy': 0.037,\n",
       "  'self_reported_ethnicity_ontology_term_id_accuracy': 0.6556,\n",
       "  'sex_ontology_term_id_accuracy': 0.5738,\n",
       "  'organism_ontology_term_id_accuracy': 1.0},\n",
       " 2000: {'cell_type_ontology_term_id_accuracy': 0.39084753677328976,\n",
       "  'tissue_ontology_term_id_accuracy': 0.065,\n",
       "  'disease_ontology_term_id_accuracy': 0.9432,\n",
       "  'assay_ontology_term_id_accuracy': 0.04,\n",
       "  'self_reported_ethnicity_ontology_term_id_accuracy': 0.6598,\n",
       "  'sex_ontology_term_id_accuracy': 0.825,\n",
       "  'organism_ontology_term_id_accuracy': 1.0},\n",
       " 6000: {'cell_type_ontology_term_id_accuracy': 0.4239738805970149,\n",
       "  'tissue_ontology_term_id_accuracy': 0.0696,\n",
       "  'disease_ontology_term_id_accuracy': 0.9924,\n",
       "  'assay_ontology_term_id_accuracy': 0.0038,\n",
       "  'self_reported_ethnicity_ontology_term_id_accuracy': 0.6494,\n",
       "  'sex_ontology_term_id_accuracy': 0.9218,\n",
       "  'organism_ontology_term_id_accuracy': 1.0}}"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for epoch in range(num_epochs):\n",
    "    print(f\"\\nEpoch {epoch + 1}/{num_epochs}\")\n",
    "    print(f\"Current learning rate: {optimizer.param_groups[0]['lr']:.2e}\")\n",
    "    ## Training phase\n",
    "    train_loss = 0.0\n",
    "    train_steps = 0\n",
    "    avg_cls = 0\n",
    "    avg_expr = 0\n",
    "    model.train()\n",
    "\n",
    "    pbar = tqdm(train_loader, desc=\"Training\")\n",
    "    for batch_idx, batch in enumerate(pbar):\n",
    "        optimizer.zero_grad()\n",
    "        total_loss, loss_expr = classif(batch, do_cls=False)\n",
    "        # Backward pass\n",
    "        scaler.scale(total_loss).backward()\n",
    "        scaler.unscale_(optimizer)\n",
    "        torch.nn.utils.clip_grad_norm_(model.parameters(), max_norm=1.0)\n",
    "        scaler.step(optimizer)\n",
    "        scaler.update()\n",
    "\n",
    "        train_loss += total_loss.item()\n",
    "        train_steps += 1\n",
    "        avg_expr += loss_expr.item()\n",
    "        #avg_cls += loss_cls.item()\n",
    "        # print(loss_cls.item(), loss_expr.item())\n",
    "        if batch_idx > 2400:\n",
    "            break\n",
    "    # Validation phase\n",
    "    model.eval()\n",
    "\n",
    "metrics = {}\n",
    "for n in [400, 2000, 6000]:\n",
    "    ad = train_data[(train_data.X > 0).sum(1) >= n]\n",
    "    # Get 5,000 random cells\n",
    "    n_cells = min(5000, ad.n_obs)  # In case you have fewer than 5,000 cells\n",
    "    random_indices = np.random.choice(ad.n_obs, n_cells, replace=False)\n",
    "    adata_subset = ad[random_indices, :].copy()\n",
    "    embed = Embedder(\n",
    "        how=\"random expr\",\n",
    "        max_len=n,\n",
    "        num_workers=8,\n",
    "        pred_embedding=[\"cell_type_ontology_term_id\"],\n",
    "        doplot=False,\n",
    "    )\n",
    "    _, metrics[n] = embed(model, adata_subset.copy())\n",
    "    \n",
    "metrics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ab336035",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/pasteur/appa/homes/jkalfon/scdataloader/scdataloader/utils.py:427: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "  organismdf = pd.concat(organismdf)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "not on wandb, could not set name\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 79/79 [00:08<00:00,  9.19it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "logging the anndata\n",
      "AnnData object with n_obs × n_vars = 5000 × 17889\n",
      "    obs: 'pred_cell_type_ontology_term_id', 'pred_tissue_ontology_term_id', 'pred_disease_ontology_term_id', 'pred_age_group', 'pred_assay_ontology_term_id', 'pred_self_reported_ethnicity_ontology_term_id', 'pred_sex_ontology_term_id', 'pred_organism_ontology_term_id', 'pred_cell_culture', 'conv_pred_cell_type_ontology_term_id', 'conv_pred_tissue_ontology_term_id', 'conv_pred_disease_ontology_term_id', 'conv_pred_age_group', 'conv_pred_assay_ontology_term_id', 'conv_pred_self_reported_ethnicity_ontology_term_id'\n",
      "    obsm: 'scprint_emb_cell_type_ontology_term_id'\n",
      "    layers: 'scprint_mu', 'scprint_theta', 'scprint_pi'\n",
      "too few cells to embed into a umap\n",
      "too few cells to compute a clustering\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/pasteur/appa/homes/jkalfon/scdataloader/scdataloader/utils.py:427: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "  organismdf = pd.concat(organismdf)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "not on wandb, could not set name\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 79/79 [00:26<00:00,  2.99it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "logging the anndata\n",
      "AnnData object with n_obs × n_vars = 5000 × 18581\n",
      "    obs: 'pred_cell_type_ontology_term_id', 'pred_tissue_ontology_term_id', 'pred_disease_ontology_term_id', 'pred_age_group', 'pred_assay_ontology_term_id', 'pred_self_reported_ethnicity_ontology_term_id', 'pred_sex_ontology_term_id', 'pred_organism_ontology_term_id', 'pred_cell_culture', 'conv_pred_cell_type_ontology_term_id', 'conv_pred_tissue_ontology_term_id', 'conv_pred_disease_ontology_term_id', 'conv_pred_age_group', 'conv_pred_assay_ontology_term_id', 'conv_pred_self_reported_ethnicity_ontology_term_id'\n",
      "    obsm: 'scprint_emb_cell_type_ontology_term_id'\n",
      "    layers: 'scprint_mu', 'scprint_theta', 'scprint_pi'\n",
      "too few cells to embed into a umap\n",
      "too few cells to compute a clustering\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/pasteur/appa/homes/jkalfon/scdataloader/scdataloader/utils.py:427: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "  organismdf = pd.concat(organismdf)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "not on wandb, could not set name\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 79/79 [02:09<00:00,  1.64s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "logging the anndata\n",
      "AnnData object with n_obs × n_vars = 5000 × 20004\n",
      "    obs: 'pred_cell_type_ontology_term_id', 'pred_tissue_ontology_term_id', 'pred_disease_ontology_term_id', 'pred_age_group', 'pred_assay_ontology_term_id', 'pred_self_reported_ethnicity_ontology_term_id', 'pred_sex_ontology_term_id', 'pred_organism_ontology_term_id', 'pred_cell_culture', 'conv_pred_cell_type_ontology_term_id', 'conv_pred_tissue_ontology_term_id', 'conv_pred_disease_ontology_term_id', 'conv_pred_age_group', 'conv_pred_assay_ontology_term_id', 'conv_pred_self_reported_ethnicity_ontology_term_id'\n",
      "    obsm: 'scprint_emb_cell_type_ontology_term_id'\n",
      "    layers: 'scprint_mu', 'scprint_theta', 'scprint_pi'\n",
      "too few cells to embed into a umap\n",
      "too few cells to compute a clustering\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{400: {'cell_type_ontology_term_id_accuracy': 0.6121981681931724,\n",
       "  'tissue_ontology_term_id_accuracy': 0.2196,\n",
       "  'disease_ontology_term_id_accuracy': 0.991,\n",
       "  'assay_ontology_term_id_accuracy': 0.0278,\n",
       "  'self_reported_ethnicity_ontology_term_id_accuracy': 0.6266,\n",
       "  'sex_ontology_term_id_accuracy': 0.585,\n",
       "  'organism_ontology_term_id_accuracy': 1.0},\n",
       " 2000: {'cell_type_ontology_term_id_accuracy': 0.650836820083682,\n",
       "  'tissue_ontology_term_id_accuracy': 0.23,\n",
       "  'disease_ontology_term_id_accuracy': 0.9982,\n",
       "  'assay_ontology_term_id_accuracy': 0.03,\n",
       "  'self_reported_ethnicity_ontology_term_id_accuracy': 0.626,\n",
       "  'sex_ontology_term_id_accuracy': 0.7358,\n",
       "  'organism_ontology_term_id_accuracy': 1.0},\n",
       " 8000: {'cell_type_ontology_term_id_accuracy': 0.7434017595307918,\n",
       "  'tissue_ontology_term_id_accuracy': 0.4376,\n",
       "  'disease_ontology_term_id_accuracy': 1.0,\n",
       "  'assay_ontology_term_id_accuracy': 0.0002,\n",
       "  'self_reported_ethnicity_ontology_term_id_accuracy': 0.5972,\n",
       "  'sex_ontology_term_id_accuracy': 0.8998,\n",
       "  'organism_ontology_term_id_accuracy': 1.0}}"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metrics = {}\n",
    "for n in [400, 2000, 8000]:\n",
    "    ad = adata[(adata.X > 0).sum(1) >= n]\n",
    "    # Get 5,000 random cells\n",
    "    n_cells = min(5000, ad.n_obs)  # In case you have fewer than 5,000 cells\n",
    "    random_indices = np.random.choice(ad.n_obs, n_cells, replace=False)\n",
    "    adata_subset = ad[random_indices, :].copy()\n",
    "    embed = Embedder(\n",
    "        how=\"random expr\",\n",
    "        max_len=n,\n",
    "        num_workers=8,\n",
    "        pred_embedding=[\"cell_type_ontology_term_id\"],\n",
    "        doplot=False,\n",
    "    )\n",
    "    _, metrics[n] = embed(model, adata_subset.copy())\n",
    "    \n",
    "metrics"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8762984b",
   "metadata": {},
   "source": [
    " 8000: {'cell_type_ontology_term_id_accuracy': 0.7407175636211931,\n",
    "  'tissue_ontology_term_id_accuracy': 0.1876,\n",
    "  'disease_ontology_term_id_accuracy': 0.9948,\n",
    "  'assay_ontology_term_id_accuracy': 0.897, #0.9274\n",
    "  'self_reported_ethnicity_ontology_term_id_accuracy': 0.599, #0.6376\n",
    "  'sex_ontology_term_id_accuracy': 0.9828,\n",
    "  'organism_ontology_term_id_accuracy': 1.0}"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scPRINT (.venv)",
   "language": "python",
   "name": "scprint-venv"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  },
  "papermill": {
   "default_parameters": {},
   "duration": 203.438373,
   "end_time": "2025-09-17T15:28:09.167342",
   "environment_variables": {},
   "exception": null,
   "input_path": "fine_tuning_cross_species_emb_mmd.ipynb",
   "output_path": "fine_tuning_cross_species_emb_mmd.ipynb",
   "parameters": {},
   "start_time": "2025-09-17T15:24:45.728969",
   "version": "2.6.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
