{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "d899f01a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "from bengrn import get_GT_db, get_perturb_gt, get_sroy_gt, BenGRN, FILEDIR\n",
    "\n",
    "import numpy as np\n",
    "import tqdm\n",
    "import scanpy as sc\n",
    "import os\n",
    "\n",
    "from huggingface_hub import hf_hub_download\n",
    "\n",
    "from scprint import scPrint\n",
    "from scprint.tasks import GNInfer\n",
    "from scdataloader import Preprocessor\n",
    "\n",
    "from bengrn import BenGRN\n",
    "from anndata.utils import make_index_unique\n",
    "\n",
    "from bengrn.base import train_classifier\n",
    "from bengrn import compute_genie3\n",
    "from grnndata import utils as grnutils\n",
    "import joblib\n",
    "\n",
    "from grnndata import read_h5ad\n",
    "\n",
    "from matplotlib import pyplot as plt\n",
    "from scdataloader import utils as data_utils\n",
    "\n",
    "from pyvis import network as pnx\n",
    "import networkx as nx\n",
    "import scipy.sparse\n",
    "import gseapy as gp\n",
    "from gseapy import dotplot\n",
    "\n",
    "from scdataloader.utils import revert_to_raw\n",
    "\n",
    "import scipy.sparse\n",
    "\n",
    "import torch\n",
    "\n",
    "torch.set_float32_matmul_precision(\"medium\")\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "160a51b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "LOC = '/pasteur/appa/scratch/jkalfon/data/'\n",
    "\n",
    "name = \"Macrophage\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "8c1dd934",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "grn_m = read_h5ad(LOC+ f\"grn_task4_mouse_{name}.h5ad\")\n",
    "grn_h = read_h5ad(LOC+ f\"grn_task4_human_{name}.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "5da0d461",
   "metadata": {},
   "outputs": [],
   "source": [
    "grn_m.var['symbol'] = grn_m.var.symbol.str.upper()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "86c82284",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2978"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# between the two networks we have ~3000 genes in common over their 4000 genes\n",
    "grn_m.var.index = grn_m.var.symbol.str.upper()\n",
    "common = set(grn_h.var.index) & set(grn_m.var.index)\n",
    "len(common)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "20fdc8b4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "index\n",
       "FTL         67.522400\n",
       "B2M         63.816372\n",
       "TMSB4X      54.556316\n",
       "NLN         51.200565\n",
       "FTH1        43.270775\n",
       "HLA-DRA     40.650234\n",
       "TXNIP       38.062641\n",
       "EIF1        35.565845\n",
       "C1QA        25.740410\n",
       "UBXN1       25.120731\n",
       "LYZ         23.978209\n",
       "ITGB2       23.338360\n",
       "CD74        23.209900\n",
       "RPS14       22.008221\n",
       "GABARAP     21.861885\n",
       "SOD2        21.790438\n",
       "ACTB        20.383188\n",
       "C1QB        20.347918\n",
       "PPIA        20.202841\n",
       "C1QC        19.605148\n",
       "RPS27A      19.256315\n",
       "SPIC        19.180021\n",
       "RPS6        18.481428\n",
       "ALB         18.100208\n",
       "ALOX5AP     17.430592\n",
       "SRGN        17.301922\n",
       "DMRT2       16.415487\n",
       "IFI30       16.027319\n",
       "SH3BGRL3    14.680469\n",
       "RPL7        13.885112\n",
       "dtype: float32"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# hub genes based on edge centrality (number of connections / strength of connections)\n",
    "grn_h.grn.sum(0).sort_values(ascending=False).head(30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "fbaf54fc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "index\n",
       "B2M        63.816372\n",
       "TMSB4X     54.556316\n",
       "FTH1       43.270775\n",
       "ITGB2      23.338360\n",
       "CD74       23.209900\n",
       "RPS14      22.008221\n",
       "SOD2       21.790438\n",
       "PPIA       20.202841\n",
       "C1QC       19.605148\n",
       "SPIC       19.180021\n",
       "RPS6       18.481428\n",
       "ALB        18.100208\n",
       "SRGN       17.301922\n",
       "IFI30      16.027319\n",
       "RPS19      13.683068\n",
       "CD163      13.305556\n",
       "CST3       13.139258\n",
       "RPLP1      11.228999\n",
       "MYL6       11.163845\n",
       "PTP4A1     11.134760\n",
       "S100A11    10.099115\n",
       "CD3D       10.072865\n",
       "VSIG4       9.813706\n",
       "OLR1        9.525496\n",
       "S100A4      9.370368\n",
       "HMOX1       9.352107\n",
       "CTSD        8.292619\n",
       "XAF1        7.860425\n",
       "VIM         7.456419\n",
       "CXCL3       7.263704\n",
       "dtype: float32"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# without taking in account genes that are not present in the mouse network\n",
    "grn_h.grn.sum(0).loc[list(common)].sort_values(ascending=False).head(30)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dda66aac",
   "metadata": {},
   "source": [
    "ACTB=ACTB (most expressed in B cell, motility, key linked gene is CFL1 top 10 in mouse)\n",
    "\n",
    "B2M=B2M (serum protein found in association with the major histocompatibility complex, linked to HLA-C, top 20 in human genes)\n",
    "\n",
    "TMSB4X = TMSB10, TMSB4X (Thymosin Beta, cell prolif, highly linked to actin and ACTB)\n",
    "\n",
    "CD83, CD52 = CD52 CD3D, (CD74)  ( immunoglobulin superfamily of receptors, Regulate b cell function, esp CD83, CD52 less)\n",
    "\n",
    "S100A5 = S100A6 (key cell process regulation with calcium binding, both in top 3 relation to TOP53)\n",
    "\n",
    "RPL41, RPL3 = RPL10, RPLP1 (ribosomal proteins, much more for human in top50, highly co-regulated, often together, have in important impact on mRNA content by working on splicing, mRNA stability, sequestration of TFs )\n",
    "\n",
    "[RPS6] = RPS27A, RPS24, (RPS28) (same)\n",
    "\n",
    "[FTL] = FTH1 (iron storage)\n",
    "    MEF2C, JUNB (key TF in b cells, https://pubmed.ncbi.nlm.nih.gov/12907453/, https://pmc.ncbi.nlm.nih.gov/articles/PMC2518613/)\n",
    "\n",
    "RGS2, RGS1, GPR183, GPX1 (G protein Receptor, signaling members, super important b cell members)\n",
    "\n",
    "JCHAIN, ITM2C, OGT, UBE2S, IFITM1, CREM, FCER1G, CST3\n",
    "    TOP2A, CRIP1, PPP1R15A, MS4A1, PPIA, PTMA\n",
    "\n",
    "\"Immunoglobulin Receptor Binding (GO:0034987)\" JCHAIN\n",
    "\"B cell activation\", \"host pathogen pathways\" IFITM1, FCER1G \"B cell activation\" MS4A1 == \"virus detection\" PPIA, PPP1R15A\n",
    "\"MHC Class II Protein Complex Binding\" MS4A1, \"HSPA1A\" (MHC related)(PPP1R15A;PTMA)\n",
    "\"circadian rythm\" CST3;CREM;OGT = TOP2A\n",
    "\n",
    "() = in the non matching genes\n",
    "[] = in the top 50\n",
    "\n",
    "can find homologs and might use some genes that are similar when others are not available or expressed\n",
    "\n",
    "bias on the 1000 cells and which ones had which genes expressed\n",
    "worked using only human / mouse known orhologs, but some relationships might not have been found\n",
    "\n",
    "Generically, many genes with high importance in generic cellular processes, and some with high importance in B cell function.\n",
    "\n",
    "Even more interesting, many of these genes are listed in papers about the transcriptional control dynamic of B cell function using paired single cell ATAC-seq - CHip-seq - RNA-seq  https://www.biorxiv.org/content/10.1101/2022.02.23.481342v2.full.pdf \"Fate-resolved gene regulatory signatures of individual B lymphocytes in the early1stages of Epstein-Barr Virus infection2\"\n",
    "(PPP1R-, TOP2A, MS4A1, FCER2/FCER1G, JCHAIN, FTH1, FTL, TMSB10, TMSB4X, GPR183, (likely other members of the G protein family), numerous ribosomal subunit genes (listed as example: RPS27A) and also))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d03c6d2f",
   "metadata": {},
   "outputs": [],
   "source": [
    "INTerferon induced genes are among the top hub genes in human but not in mouse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "530d0bbd",
   "metadata": {},
   "outputs": [],
   "source": [
    "APOE\n",
    "FCNA\n",
    "SRRM2\n",
    "MRC1\n",
    "IGFBP7\n",
    "WFDC17\n",
    "CFP\n",
    "ITM2B\n",
    "NDFIP1\n",
    "NME2\n",
    "DDX5\n",
    "VSIG4\n",
    "ZBTB20\n",
    "TUBA4A"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f63a7613",
   "metadata": {},
   "outputs": [],
   "source": [
    "IFI30 - IFITM3 INTerferon induced transmembrane protein 3 and thiol reductase involved in MHC class II antigen processing\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8403999a",
   "metadata": {},
   "outputs": [],
   "source": [
    "NLN\n",
    "HLA-DRA\n",
    "TXNIP\n",
    "EIF1\n",
    "UBXN1\n",
    "LYZ\n",
    "ITGB2\n",
    "GABARAP\n",
    "SOD2\n",
    "ACTB\n",
    "PPIA\n",
    "SPIC\n",
    "ALOX5AP\n",
    "SRGN\n",
    "DMRT2\n",
    "SH3BGRL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d9742c8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "CD27"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "2115d2aa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_set</th>\n",
       "      <th>Term</th>\n",
       "      <th>P-value</th>\n",
       "      <th>Adjusted P-value</th>\n",
       "      <th>Old P-value</th>\n",
       "      <th>Old adjusted P-value</th>\n",
       "      <th>Odds Ratio</th>\n",
       "      <th>Combined Score</th>\n",
       "      <th>Genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WikiPathway_2023_Human</td>\n",
       "      <td>Complement System WP2806</td>\n",
       "      <td>0.002909</td>\n",
       "      <td>0.042997</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>31.708683</td>\n",
       "      <td>185.175704</td>\n",
       "      <td>VSIG4;CFP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>WikiPathway_2023_Human</td>\n",
       "      <td>ApoE And miR 146 In Inflammation And Atheroscl...</td>\n",
       "      <td>0.009407</td>\n",
       "      <td>0.042997</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>142.650000</td>\n",
       "      <td>665.640291</td>\n",
       "      <td>APOE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>WikiPathway_2023_Human</td>\n",
       "      <td>Interactions Between LOXL4 And Oxidative Stres...</td>\n",
       "      <td>0.010968</td>\n",
       "      <td>0.042997</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>118.854167</td>\n",
       "      <td>536.364611</td>\n",
       "      <td>IGFBP7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>WikiPathway_2023_Human</td>\n",
       "      <td>Lipid Particles Composition WP3601</td>\n",
       "      <td>0.010968</td>\n",
       "      <td>0.042997</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>118.854167</td>\n",
       "      <td>536.364611</td>\n",
       "      <td>APOE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>WikiPathway_2023_Human</td>\n",
       "      <td>Alzheimer 39 S Disease And miRNA Effects WP2059</td>\n",
       "      <td>0.013818</td>\n",
       "      <td>0.042997</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>13.780788</td>\n",
       "      <td>59.006095</td>\n",
       "      <td>APOE;TUBA4A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>WikiPathway_2023_Human</td>\n",
       "      <td>Alzheimer 39 S Disease WP5124</td>\n",
       "      <td>0.013818</td>\n",
       "      <td>0.042997</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>13.780788</td>\n",
       "      <td>59.006095</td>\n",
       "      <td>APOE;TUBA4A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>WikiPathway_2023_Human</td>\n",
       "      <td>Familial Hyperlipidemia Type 3 WP5110</td>\n",
       "      <td>0.014082</td>\n",
       "      <td>0.042997</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>89.109375</td>\n",
       "      <td>379.862471</td>\n",
       "      <td>APOE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>WikiPathway_2023_Human</td>\n",
       "      <td>Metabolic Pathway Of LDL HDL And TG Including ...</td>\n",
       "      <td>0.015635</td>\n",
       "      <td>0.042997</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>79.194444</td>\n",
       "      <td>329.307681</td>\n",
       "      <td>APOE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>WikiPathway_2023_Human</td>\n",
       "      <td>Complement Activation WP545</td>\n",
       "      <td>0.018736</td>\n",
       "      <td>0.045800</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>64.772727</td>\n",
       "      <td>257.620082</td>\n",
       "      <td>CFP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>WikiPathway_2023_Human</td>\n",
       "      <td>Melatonin Metabolism And Effects WP3298</td>\n",
       "      <td>0.023371</td>\n",
       "      <td>0.051417</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>50.866071</td>\n",
       "      <td>191.065387</td>\n",
       "      <td>APOE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>WikiPathway_2023_Human</td>\n",
       "      <td>Statin Inhibition Of Cholesterol Production WP430</td>\n",
       "      <td>0.026451</td>\n",
       "      <td>0.052901</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>44.492188</td>\n",
       "      <td>161.616875</td>\n",
       "      <td>APOE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>WikiPathway_2023_Human</td>\n",
       "      <td>Fatty Acids And Lipoproteins Transport In Hepa...</td>\n",
       "      <td>0.029521</td>\n",
       "      <td>0.054122</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>39.534722</td>\n",
       "      <td>139.266920</td>\n",
       "      <td>APOE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>WikiPathway_2023_Human</td>\n",
       "      <td>Pathogenic Escherichia Coli Infection WP2272</td>\n",
       "      <td>0.040201</td>\n",
       "      <td>0.067937</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>28.430000</td>\n",
       "      <td>91.370476</td>\n",
       "      <td>TUBA4A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>WikiPathway_2023_Human</td>\n",
       "      <td>Vitamin B12 Metabolism WP1533</td>\n",
       "      <td>0.043233</td>\n",
       "      <td>0.067937</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>26.314815</td>\n",
       "      <td>82.659104</td>\n",
       "      <td>APOE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>WikiPathway_2023_Human</td>\n",
       "      <td>Apoptosis Related Network Due To Altered Notch...</td>\n",
       "      <td>0.058265</td>\n",
       "      <td>0.081185</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>19.168919</td>\n",
       "      <td>54.492453</td>\n",
       "      <td>APOE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>WikiPathway_2023_Human</td>\n",
       "      <td>Parkin Ubiquitin Proteasomal System Pathway WP...</td>\n",
       "      <td>0.061246</td>\n",
       "      <td>0.081185</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>18.179487</td>\n",
       "      <td>50.772584</td>\n",
       "      <td>TUBA4A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>WikiPathway_2023_Human</td>\n",
       "      <td>Cholesterol Metabolism WP5304</td>\n",
       "      <td>0.062734</td>\n",
       "      <td>0.081185</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>17.721875</td>\n",
       "      <td>49.069283</td>\n",
       "      <td>APOE</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  Gene_set                                               Term  \\\n",
       "0   WikiPathway_2023_Human                           Complement System WP2806   \n",
       "1   WikiPathway_2023_Human  ApoE And miR 146 In Inflammation And Atheroscl...   \n",
       "2   WikiPathway_2023_Human  Interactions Between LOXL4 And Oxidative Stres...   \n",
       "3   WikiPathway_2023_Human                 Lipid Particles Composition WP3601   \n",
       "4   WikiPathway_2023_Human    Alzheimer 39 S Disease And miRNA Effects WP2059   \n",
       "5   WikiPathway_2023_Human                      Alzheimer 39 S Disease WP5124   \n",
       "6   WikiPathway_2023_Human              Familial Hyperlipidemia Type 3 WP5110   \n",
       "7   WikiPathway_2023_Human  Metabolic Pathway Of LDL HDL And TG Including ...   \n",
       "8   WikiPathway_2023_Human                        Complement Activation WP545   \n",
       "9   WikiPathway_2023_Human            Melatonin Metabolism And Effects WP3298   \n",
       "10  WikiPathway_2023_Human  Statin Inhibition Of Cholesterol Production WP430   \n",
       "11  WikiPathway_2023_Human  Fatty Acids And Lipoproteins Transport In Hepa...   \n",
       "12  WikiPathway_2023_Human       Pathogenic Escherichia Coli Infection WP2272   \n",
       "13  WikiPathway_2023_Human                      Vitamin B12 Metabolism WP1533   \n",
       "14  WikiPathway_2023_Human  Apoptosis Related Network Due To Altered Notch...   \n",
       "15  WikiPathway_2023_Human  Parkin Ubiquitin Proteasomal System Pathway WP...   \n",
       "16  WikiPathway_2023_Human                      Cholesterol Metabolism WP5304   \n",
       "\n",
       "     P-value  Adjusted P-value  Old P-value  Old adjusted P-value  Odds Ratio  \\\n",
       "0   0.002909          0.042997            0                     0   31.708683   \n",
       "1   0.009407          0.042997            0                     0  142.650000   \n",
       "2   0.010968          0.042997            0                     0  118.854167   \n",
       "3   0.010968          0.042997            0                     0  118.854167   \n",
       "4   0.013818          0.042997            0                     0   13.780788   \n",
       "5   0.013818          0.042997            0                     0   13.780788   \n",
       "6   0.014082          0.042997            0                     0   89.109375   \n",
       "7   0.015635          0.042997            0                     0   79.194444   \n",
       "8   0.018736          0.045800            0                     0   64.772727   \n",
       "9   0.023371          0.051417            0                     0   50.866071   \n",
       "10  0.026451          0.052901            0                     0   44.492188   \n",
       "11  0.029521          0.054122            0                     0   39.534722   \n",
       "12  0.040201          0.067937            0                     0   28.430000   \n",
       "13  0.043233          0.067937            0                     0   26.314815   \n",
       "14  0.058265          0.081185            0                     0   19.168919   \n",
       "15  0.061246          0.081185            0                     0   18.179487   \n",
       "16  0.062734          0.081185            0                     0   17.721875   \n",
       "\n",
       "    Combined Score        Genes  \n",
       "0       185.175704    VSIG4;CFP  \n",
       "1       665.640291         APOE  \n",
       "2       536.364611       IGFBP7  \n",
       "3       536.364611         APOE  \n",
       "4        59.006095  APOE;TUBA4A  \n",
       "5        59.006095  APOE;TUBA4A  \n",
       "6       379.862471         APOE  \n",
       "7       329.307681         APOE  \n",
       "8       257.620082          CFP  \n",
       "9       191.065387         APOE  \n",
       "10      161.616875         APOE  \n",
       "11      139.266920         APOE  \n",
       "12       91.370476       TUBA4A  \n",
       "13       82.659104         APOE  \n",
       "14       54.492453         APOE  \n",
       "15       50.772584       TUBA4A  \n",
       "16       49.069283         APOE  "
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Assuming 'node_names' contains the list of gene names.\n",
    "# Note: it seems with enrichr, one can only use one gene set at a time, choose accordingly\n",
    "enr = gp.enrichr(\n",
    "    gene_list=[#\"JCHAIN\", \"ITM2C\", \"OGT\", \"UBE2S\", \"IFITM1\", \"CREM\", \"FCER1G\", \"CST3\"],\n",
    "    #\"NLN\",\n",
    "    #\"HLA-DRA\",\n",
    "    #\"TXNIP\",\n",
    "    #\"EIF1\",\n",
    "    #\"UBXN1\",\n",
    "    #\"LYZ\",\n",
    "    #\"ITGB2\",\n",
    "    #\"GABARAP\",\n",
    "    #\"SOD2\",\n",
    "    #\"ACTB\",\n",
    "    #\"PPIA\",\n",
    "    #\"SPIC\",\n",
    "    #\"ALOX5AP\",\n",
    "    #\"SRGN\",\n",
    "    #\"DMRT2\",\n",
    "    #\"SH3BGRL\",\n",
    "    #],\n",
    "    \"APOE\",\n",
    "    \"FCNA\",\n",
    "    \"SRRM2\",\n",
    "    \"MRC1\",\n",
    "    \"IGFBP7\",\n",
    "    \"WFDC17\",\n",
    "    \"CFP\",\n",
    "    \"ITM2B\",\n",
    "    \"NDFIP1\",\n",
    "    \"NME2\",\n",
    "    \"DDX5\",\n",
    "    \"VSIG4\",\n",
    "    \"ZBTB20\",\n",
    "    \"TUBA4A\",\n",
    "    ],\n",
    "    # [\n",
    "    #    \"TOP2A\",\n",
    "    #    \"CRIP1\",\n",
    "    #    \"PPP1R15A\",\n",
    "    #    \"MS4A1\",\n",
    "    #    \"PPIA\",\n",
    "    #    \"PTMA\",\n",
    "    # ], #[\"JCHAIN\", \"ITM2C\", \"OGT\", \"UBE2S\", \"IFITM1\", \"CREM\", \"FCER1G\", \"CST3\"],\n",
    "    gene_sets=[\n",
    "        \"WikiPathway_2023_Human\"\n",
    "    ],  # , 'MSigDB_Hallmark_2020', 'Reactome_2022', 'Tabula_Sapiens', 'WikiPathway_2023_Human', 'TF_Perturbations_Followed_by_Expression', 'PPI_Hub_Proteins', 'OMIM_Disease', 'GO_Molecular_Function_2023'],\n",
    "    organism=\"Human\",  # change accordingly\n",
    "    # description='pathway',\n",
    "    cutoff=0.05,  # test dataset, use lower value for real case\n",
    "    background=grn_h.var.symbol.tolist(),\n",
    ")\n",
    "enr.res2d[enr.res2d[\"Adjusted P-value\"] < 0.1].head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a0b7fde",
   "metadata": {},
   "outputs": [],
   "source": [
    "HUMAN\n",
    "1\tWikiPathway_2023_Human\tMechanoregulation And Pathology Of YAP TAZ Via...\t0.001860\t0.051145\t0\t0\t38.006689\t238.958204\tITGB2;ACTB\n",
    "2\tWikiPathway_2023_Human\tMale Infertility WP4673\t0.005202\t0.052236\t0\t0\t21.788462\t114.578700\tHLA-DRA;SOD2\n",
    "3\tWikiPathway_2023_Human\tAcute Viral Myocarditis WP4298\t0.005202\t0.052236\t0\t0\t21.788462\t114.578700\tITGB2;ACTB\n",
    "4\tWikiPathway_2023_Human\tGlucocorticoid Receptor Pathway WP2880\t0.005448\t0.052236\t0\t0\t21.253283\t110.784267\tSRGN;ALOX5AP\n",
    "5\tWikiPathway_2023_Human\tSelenium Micronutrient Network WP15\t0.005698\t0.052236\t0\t0\t20.743590\t107.193841\tALOX5AP;SOD2\n",
    "Allograft Rejection\t0.003408\t0.047707\t0\t0\t11.940171\t67.840978\tSRGN;ITGB2;HLA-DRA\n",
    "0\tGO_Molecular_Function_2023\tK48-linked Polyubiquitin Modification-Dependen...\t0.002622\t0.055033\t0\t0\tinf\tinf\tUBXN1\n",
    "1\tGO_Molecular_Function_2023\tUbiquitin Protein Ligase Binding (GO:0031625)\t0.004653\t0.055033\t0\t0\t10.637405\t57.124879\tUBXN1;TXNIP;GABARAP\n",
    "2\tGO_Molecular_Function_2023\tUbiquitin-Like Protein Ligase Binding (GO:0044...\t0.005693\t0.055033\t0\t0\t9.865248\t50.988665\tUBXN1;TXNIP;GABARAP\n",
    "3\tGO_Molecular_Function_2023\tMHC Class II Receptor Activity (GO:0032395)\t0.007848\t0.056897\t0\t0\t203.678571\t987.336365\tHLA-DRA\n",
    "4\tGO_Molecular_Function_2023\tGABA Receptor Binding (GO:0050811)\t0.010451\t0.060615\t0\t0\t135.761905\t619.218567\tGABARAP\n",
    "5\tGO_Molecular_Function_2023\tComplement Component C3b Binding (GO:0001851)\t0.015638\t0.075584\t0\t0\t81.428571\t338.583502\tITGB2\n",
    "6\tGO_Molecular_Function_2023\tCyclosporin A Binding (GO:0016018)\t0.023371\t0.088566\t0\t0\t50.866071\t191.065387\tPPIA\n",
    "7\tGO_Molecular_Function_2023\tUbiquitin Ligase-Substrate Adaptor Activity (G...\t0.028495\t0.088566\t0\t0\t40.678571\t144.735018\tSH3BGRL\n",
    "8\tGO_Molecular_Function_2023\tRibosomal Small Subunit Binding (GO:0043024)\t0.028495\t0.088566\t0\t0\t40.678571\t144.735018\tEIF1\n",
    "9\tGO_Molecular_Function_2023\tBeta-Tubulin Binding (GO:0048487)\t0.031048\t0.088566\t0\t0\t36.974026\t128.382165\tGABARAP\n",
    "\n",
    "MOUSE\n",
    "0\tGO_Molecular_Function_2023\tAmyloid-Beta Binding (GO:0001540)\t0.002012\t0.044950\t0\t0\t38.564626\t239.436015\tAPOE;ITM2B\n",
    "1\tGO_Molecular_Function_2023\tG-quadruplex DNA Binding (GO:0051880)\t0.004714\t0.044950\t0\t0\t356.812500\t1911.551692\tNME2\n",
    "2\tGO_Molecular_Function_2023\tPhosphatidylcholine-Sterol O-acyltransferase A...\t0.004714\t0.044950\t0\t0\t356.812500\t1911.551692\tAPOE\n",
    "3\tGO_Molecular_Function_2023\tC2H2 Zinc Finger Domain Binding (GO:0070742)\t0.006280\t0.044950\t0\t0\t237.833333\t1205.890220\tSRRM2\n",
    "4\tGO_Molecular_Function_2023\tComplement Component C3b Binding (GO:0001851)\t0.009407\t0.044950\t0\t0\t142.650000\t665.640291\tVSIG4\n",
    "5\tGO_Molecular_Function_2023\tGuanyl Ribonucleotide Binding (GO:0032561)\t0.011047\t0.044950\t0\t0\t15.556172\t70.090327\tNME2;TUBA4A\n",
    "6\tGO_Molecular_Function_2023\tWW Domain Binding (GO:0050699)\t0.012526\t0.044950\t0\t0\t101.857143\t446.130544\tNDFIP1\n",
    "7\tGO_Molecular_Function_2023\tNucleoside Diphosphate Kinase Activity (GO:000...\t0.012526\t0.044950\t0\t0\t101.857143\t446.130544\tNME2\n",
    "8\tGO_Molecular_Function_2023\tLow-Density Lipoprotein Particle Receptor Bind...\t0.014082\t0.044950\t0\t0\t89.109375\t379.862471\tAPOE\n",
    "9\tGO_Molecular_Function_2023\tPhosphotransferase Activity, Phosphate Group A...\t0.014082\t0.044950\t0\t0\t89.109375\t379.862471\tNME2\n",
    "10\tGO_Molecular_Function_2023\tmRNA Binding (GO:0003729)\t0.016865\t0.044950\t0\t0\t12.363234\t50.473384\tSRRM2;DDX5\n",
    "11\tGO_Molecular_Function_2023\tNucleobase-Containing Compound Kinase Activity...\t0.017187\t0.044950\t0\t0\t71.262500\t289.582676\tNME2\n",
    "12\tGO_Molecular_Function_2023\tpre-mRNA Binding (GO:0036002)\t0.017187\t0.044950\t0\t0\t71.262500\t289.582676\tDDX5\n",
    "13\tGO_Molecular_Function_2023\tLipoprotein Particle Receptor Binding (GO:0070...\t0.021829\t0.046386\t0\t0\t54.788462\t209.540593\tAPOE\n",
    "14\tGO_Molecular_Function_2023\tNuclear Androgen Receptor Binding (GO:0050681)\t0.021829\t0.046386\t0\t0\t54.788462\t209.540593\tDDX5\n",
    "15\tGO_Molecular_Function_2023\tR-SMAD Binding (GO:0070412)\t0.021829\t0.046386\t0\t0\t54.788462\t209.540593\tDDX5\n",
    "16\tGO_Molecular_Function_2023\tPromoter-Specific Chromatin Binding (GO:1990841)\t0.024912\t0.049824\t0\t0\t47.466667\t175.266110\tDDX5\n",
    "0\tReactome_2022\tMetabolism Of Proteins R-HSA-392499\t0.000344\t0.034396\t0\t0\t12.930719\t103.122267\tDDX5;IGFBP7;APOE;CFP;ITM2B;TUBA4A\n",
    "1\tReactome_2022\tAmyloid Fiber Formation R-HSA-977225\t0.000936\t0.046788\t0\t0\t57.989796\t404.429819\tAPOE;ITM2B\n",
    "2\tReactome_2022\tChylomicron Clearance R-HSA-8964026\t0.003145\t0.082834\t0\t0\t713.750000\t4112.675946\tAPOE\n",
    "3\tReactome_2022\tPost-translational Protein Phosphorylation R-H...\t0.003358\t0.082834\t0\t0\t29.381818\t167.367878\tIGFBP7;APOE\n",
    "4\tReactome_2022\tRegulation Of IGF Transport And Uptake By IGFB...\t0.004218\t0.082834\t0\t0\t26.032258\t142.355222\tIGFBP7;APOE\n",
    "5\tReactome_2022\tChylomicron Remodeling R-HSA-8963901\t0.006280\t0.082834\t0\t0\t237.833333\t1205.890220\tAPOE\n",
    "6\tReactome_2022\tModulation By Mtb Of Host Immune System R-HSA-...\t0.006280\t0.082834\t0\t0\t237.833333\t1205.890220\tMRC1\n",
    "7\tReactome_2022\tPost-translational Protein Modification R-HSA-...\t0.007187\t0.082834\t0\t0\t7.567766\t37.350912\tDDX5;IGFBP7;APOE;CFP\n",
    "8\tReactome_2022\tChylomicron Assembly R-HSA-8963888\t0.007845\t0.082834\t0\t0\t178.343750\t864.587514\tAPOE\n",
    "9\tReactome_2022\tmRNA Splicing - Major Pathway R-HSA-72163\t0.009863\t0.082834\t0\t0\t16.536082\t76.379846\tSRRM2;DDX5\n",
    "10\tReactome_2022\tmRNA Splicing R-HSA-72172\t0.010845\t0.082834\t0\t0\t15.711485\t71.079400\tSRRM2;DDX5\n",
    "0\tWikiPathway_2023_Human\tComplement System WP2806\t0.002909\t0.042997\t0\t0\t31.708683\t185.175704\tVSIG4;CFP\n",
    "1\tWikiPathway_2023_Human\tApoE And miR 146 In Inflammation And Atheroscl...\t0.009407\t0.042997\t0\t0\t142.650000\t665.640291\tAPOE\n",
    "2\tWikiPathway_2023_Human\tInteractions Between LOXL4 And Oxidative Stres...\t0.010968\t0.042997\t0\t0\t118.854167\t536.364611\tIGFBP7\n",
    "3\tWikiPathway_2023_Human\tLipid Particles Composition WP3601\t0.010968\t0.042997\t0\t0\t118.854167\t536.364611\tAPOE\n",
    "4\tWikiPathway_2023_Human\tAlzheimer 39 S Disease And miRNA Effects WP2059\t0.013818\t0.042997\t0\t0\t13.780788\t59.006095\tAPOE;TUBA4A\n",
    "5\tWikiPathway_2023_Human\tAlzheimer 39 S Disease WP5124\t0.013818\t0.042997\t0\t0\t13.780788\t59.006095\tAPOE;TUBA4A\n",
    "6\tWikiPathway_2023_Human\tFamilial Hyperlipidemia Type 3 WP5110\t0.014082\t0.042997\t0\t0\t89.109375\t379.862471\tAPOE\n",
    "7\tWikiPathway_2023_Human\tMetabolic Pathway Of LDL HDL And TG Including ...\t0.015635\t0.042997\t0\t0\t79.194444\t329.307681\tAPOE\n",
    "8\tWikiPathway_2023_Human\tComplement Activation WP545\t0.018736\t0.045800\t0\t0\t64.772727\t257.620082\tCFP\n",
    "9\tWikiPathway_2023_Human\tMelatonin Metabolism And Effects WP3298\t0.023371\t0.051417\t0\t0\t50.866071\t191.065387\tAPOE\n",
    "10\tWikiPathway_2023_Human\tStatin Inhibition Of Cholesterol Production WP430\t0.026451\t0.052901\t0\t0\t44.492188\t161.616875\tAPOE\n",
    "11\tWikiPathway_2023_Human\tFatty Acids And Lipoproteins Transport In Hepa...\t0.029521\t0.054122\t0\t0\t39.534722\t139.266920\tAPOE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21d47914",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ACTB',\n",
       " 'ATP1A1',\n",
       " 'ATP4A',\n",
       " 'B2M',\n",
       " 'BIRC5',\n",
       " 'CCDC89',\n",
       " 'CCNB2',\n",
       " 'CD3D',\n",
       " 'CD52',\n",
       " 'CD74',\n",
       " 'CFL1',\n",
       " 'CHAD',\n",
       " 'COLQ',\n",
       " 'COX7B',\n",
       " 'COX8B',\n",
       " 'CRIP1',\n",
       " 'CTLA2A',\n",
       " 'CYP2F2',\n",
       " 'DUSP11',\n",
       " 'FAM120B',\n",
       " 'FCMR',\n",
       " 'FTH1',\n",
       " 'GIMAP4',\n",
       " 'GM266',\n",
       " 'H2-AA',\n",
       " 'H2-AB1',\n",
       " 'H2-EB1',\n",
       " 'H3F3B',\n",
       " 'HSP90AB1',\n",
       " 'JCHAIN',\n",
       " 'JUNB',\n",
       " 'LIN54',\n",
       " 'LY6D',\n",
       " 'LYPD2',\n",
       " 'MEF2C',\n",
       " 'MPC1',\n",
       " 'MRGPRF',\n",
       " 'MS4A1',\n",
       " 'MT-CYTB',\n",
       " 'MT-ND1',\n",
       " 'MT-ND2',\n",
       " 'MT-ND4',\n",
       " 'NDC80',\n",
       " 'NWD1',\n",
       " 'OXCT1',\n",
       " 'PCDHGA12',\n",
       " 'PCLAF',\n",
       " 'PEX11G',\n",
       " 'PPIA',\n",
       " 'PPP1R15A',\n",
       " 'PSME1',\n",
       " 'PTMA',\n",
       " 'PTPRB',\n",
       " 'PTPRCAP',\n",
       " 'RPL10',\n",
       " 'RPL13',\n",
       " 'RPL13A',\n",
       " 'RPL14',\n",
       " 'RPL18',\n",
       " 'RPL18A',\n",
       " 'RPL3',\n",
       " 'RPL37',\n",
       " 'RPL37A',\n",
       " 'RPL38',\n",
       " 'RPL39',\n",
       " 'RPL4',\n",
       " 'RPL6',\n",
       " 'RPLP0',\n",
       " 'RPLP1',\n",
       " 'RPS12',\n",
       " 'RPS16',\n",
       " 'RPS17',\n",
       " 'RPS18',\n",
       " 'RPS19',\n",
       " 'RPS23',\n",
       " 'RPS24',\n",
       " 'RPS27',\n",
       " 'RPS27A',\n",
       " 'RPS28',\n",
       " 'RPS29',\n",
       " 'RPS8',\n",
       " 'RPSA',\n",
       " 'RRM2',\n",
       " 'S100A4',\n",
       " 'S100A6',\n",
       " 'SELENOP',\n",
       " 'SLC25A5',\n",
       " 'SNRPG',\n",
       " 'TAF10',\n",
       " 'TIFA',\n",
       " 'TMEM132E',\n",
       " 'TMEM176B',\n",
       " 'TMSB10',\n",
       " 'TMSB4X',\n",
       " 'TOMM7',\n",
       " 'TOP1',\n",
       " 'TOP2A',\n",
       " 'UBB',\n",
       " 'XRN2',\n",
       " 'YPEL1'}"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "TOP = 100\n",
    "set(grn_m.grn.sum(0).sort_values(ascending=False).head(TOP).index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "ac3b7257",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "symbol\n",
       "C1QC       77.519073\n",
       "TMSB4X     41.802864\n",
       "APOE       40.588585\n",
       "FCNA       39.359177\n",
       "SRRM2      33.876957\n",
       "MRC1       33.289558\n",
       "RPL13A     32.056759\n",
       "IGFBP7     31.896709\n",
       "RPL11      27.574696\n",
       "FTL1       27.434568\n",
       "B2M        26.620033\n",
       "IFITM3     26.154570\n",
       "WFDC17     25.864525\n",
       "CFP        21.936113\n",
       "ITM2B      20.361074\n",
       "NDFIP1     20.023716\n",
       "CD52       19.125843\n",
       "MT-CYTB    18.294144\n",
       "NME2       17.928022\n",
       "MT-ND1     17.800694\n",
       "MT-ND4     17.678120\n",
       "RPS24      17.041389\n",
       "FTH1       16.653090\n",
       "CD5L       16.438406\n",
       "DDX5       15.556078\n",
       "VSIG4      15.311343\n",
       "ALB        15.164091\n",
       "CD74       15.150368\n",
       "ZBTB20     15.142496\n",
       "TUBA4A     14.925520\n",
       "dtype: float32"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# mouse hub genes\n",
    "grn_m.grn.sum(0).sort_values(ascending=False).head(30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "4bffb4d9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "symbol\n",
       "C1QC       77.519073\n",
       "TMSB4X     41.802864\n",
       "APOE       40.588585\n",
       "SRRM2      33.876957\n",
       "MRC1       33.289558\n",
       "IGFBP7     31.896709\n",
       "B2M        26.620033\n",
       "IFITM3     26.154570\n",
       "CFP        21.936113\n",
       "CD52       19.125843\n",
       "NME2       17.928022\n",
       "FTH1       16.653090\n",
       "CD5L       16.438406\n",
       "VSIG4      15.311343\n",
       "ALB        15.164091\n",
       "CD74       15.150368\n",
       "ZBTB20     15.142496\n",
       "TUBA4A     14.925520\n",
       "CTSD       14.200592\n",
       "FN1        13.714249\n",
       "TMEM218    12.782110\n",
       "SLC25A5    12.569665\n",
       "RPLP1      12.023596\n",
       "FOLR2      11.515420\n",
       "GRN        11.379998\n",
       "HSPA8      11.121265\n",
       "CXCL2      10.497538\n",
       "ID3         9.898645\n",
       "CLEC4F      9.694858\n",
       "PRDX1       9.653316\n",
       "dtype: float32"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# without taking in account genes not present in the human network\n",
    "grn_m.grn.sum(0).loc[list(common)].sort_values(ascending=False).head(30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "d7e1e7d1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'B2M', 'C1QC', 'TMSB4X'}"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TOP = 20\n",
    "set(grn_m.grn.sum(0).sort_values(ascending=False).head(TOP).index) & set(\n",
    "    grn_h.grn.sum(0).sort_values(ascending=False).head(TOP).index\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "e26d0042",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'TMEM218', 'SRRM2', 'GRN', 'SLC25A5', 'CFP', 'MPEG1', 'CHMP1B', 'ARG1', 'FOLR2', 'MX1', 'CLEC4F', 'APOA1', 'IFIT3', 'NUSAP1', 'ZFP36', 'ZBTB20', 'TUBA4A', 'KIF22', 'ID3', 'CD5L', 'HSPA8', 'APOE', 'MRC1', 'LPL', 'PLTP', 'HSPA5', 'IGFBP7', 'LGALS3', 'PTPRB', 'NME2', 'CD52', 'CXCL2', 'FN1', 'PRDX1', 'IFITM3'}\n",
      "{'HMGN2', 'RPS14', 'EPSTI1', 'IFI30', 'S100A11', 'NR4A3', 'OLR1', 'SPARCL1', 'HMGB2', 'EGFL7', 'RPS6', 'CXCL3', 'S100A9', 'MYL6', 'CD163', 'NAMPT', 'PTP4A1', 'CTSL', 'HSP90AB1', 'S100A4', 'SRGN', 'COL1A2', 'VIM', 'SOD2', 'ITGB2', 'CST3', 'CD3D', 'XAF1', 'SPIC', 'HMOX1', 'FOXRED2', 'P2RY13', 'CCL4', 'MARCO', 'MS4A7'}\n"
     ]
    }
   ],
   "source": [
    "# top differential hubs\n",
    "TOP = 100\n",
    "# in m not in h\n",
    "print(\n",
    "    (\n",
    "        set(grn_m.grn.sum(0).sort_values(ascending=False).head(TOP).index)\n",
    "        - set(grn_h.grn.sum(0).sort_values(ascending=False).head(TOP).index)\n",
    "    )\n",
    "    & (set(grn_h.var.index) & set(grn_m.var.index))\n",
    ")\n",
    "# in h not in m\n",
    "print(\n",
    "    (\n",
    "        set(grn_h.grn.sum(0).sort_values(ascending=False).head(TOP).index)\n",
    "        - set(grn_m.grn.sum(0).sort_values(ascending=False).head(TOP).index)\n",
    "    )\n",
    "    & (set(grn_h.var.index) & set(grn_m.var.index))\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e55a2e61",
   "metadata": {},
   "outputs": [],
   "source": [
    "{'TMEM218', 'SRRM2', 'GRN', 'SLC25A5', 'CFP', 'MPEG1', 'CHMP1B', 'ARG1', 'FOLR2', 'MX1', 'CLEC4F', 'APOA1', 'IFIT3', 'NUSAP1', 'ZFP36', 'ZBTB20', 'TUBA4A', 'KIF22', 'ID3', 'CD5L', 'HSPA8', 'APOE', 'MRC1', 'LPL', 'PLTP', 'HSPA5', 'IGFBP7', 'LGALS3', 'PTPRB', 'NME2', 'CD52', 'CXCL2', 'FN1', 'PRDX1', 'IFITM3'}\n",
    "\n",
    "{'HMGN2', 'RPS14', 'EPSTI1', 'IFI30', 'S100A11', 'NR4A3', 'OLR1', 'SPARCL1', 'HMGB2', 'EGFL7', 'RPS6', 'CXCL3', 'S100A9', 'MYL6', 'CD163', 'NAMPT', 'PTP4A1', 'CTSL', 'HSP90AB1', 'S100A4', 'SRGN', 'COL1A2', 'VIM', 'SOD2', 'ITGB2', 'CST3', 'CD3D', 'XAF1', 'SPIC', 'HMOX1', 'FOXRED2', 'P2RY13', 'CCL4', 'MARCO', 'MS4A7'}\n",
    "\n",
    "CXCL2 - CXCL3\n",
    "CD52 - CD3D\n",
    "PTP4A1 - PTPRB\n",
    "HSPA8 / HSPA5 - HSP90AB1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f0e9c36e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2287b5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "TREM\n",
    "PPARG\n",
    "CD36, CD9"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "4b561829",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'IGFBP7', 'NME2', 'SRRM2', 'CD52', 'APOE', 'MRC1', 'CFP', 'IFITM3'}\n",
      "{'PPIA', 'ITGB2', 'FTH1', 'RPS14', 'CD74', 'SOD2'}\n"
     ]
    }
   ],
   "source": [
    "TOP = 20\n",
    "# in m not in h\n",
    "print(\n",
    "    (\n",
    "        set(grn_m.grn.sum(0).sort_values(ascending=False).head(TOP).index)\n",
    "        - set(grn_h.grn.sum(0).sort_values(ascending=False).head(TOP).index)\n",
    "    )\n",
    "    & (set(grn_h.var.index) & set(grn_m.var.index))\n",
    ")\n",
    "# in h not in m\n",
    "print(\n",
    "    (\n",
    "        set(grn_h.grn.sum(0).sort_values(ascending=False).head(TOP).index)\n",
    "        - set(grn_m.grn.sum(0).sort_values(ascending=False).head(TOP).index)\n",
    "    )\n",
    "    & (set(grn_h.var.index) & set(grn_m.var.index))\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "6a48f78c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'IGFBP7', 'TUBA4A', 'FOLR2', 'NME2', 'TMEM218', 'SRRM2', 'CD52', 'CD5L', 'GRN', 'CTSD', 'HSPA8', 'APOE', 'FN1', 'MRC1', 'SLC25A5', 'CFP', 'IFITM3', 'ZBTB20'}\n",
      "{'PPIA', 'ITGB2', 'PTP4A1', 'CST3', 'RPS14', 'CD3D', 'SPIC', 'IFI30', 'RPS6', 'SRGN', 'RPS19', 'SOD2', 'S100A11', 'OLR1', 'MYL6', 'CD163'}\n"
     ]
    }
   ],
   "source": [
    "TOP = 50\n",
    "# in m not in h\n",
    "print(\n",
    "    (\n",
    "        set(grn_m.grn.sum(0).sort_values(ascending=False).head(TOP).index)\n",
    "        - set(grn_h.grn.sum(0).sort_values(ascending=False).head(TOP).index)\n",
    "    )\n",
    "    & (set(grn_h.var.index) & set(grn_m.var.index))\n",
    ")\n",
    "# in h not in m\n",
    "print(\n",
    "    (\n",
    "        set(grn_h.grn.sum(0).sort_values(ascending=False).head(TOP).index)\n",
    "        - set(grn_m.grn.sum(0).sort_values(ascending=False).head(TOP).index)\n",
    "    )\n",
    "    & (set(grn_h.var.index) & set(grn_m.var.index))\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "7c3b9dc4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Top central genes: []\n",
      "Top central genes: []\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# we now compute eigen centrality creating a sparse network by only keeping the top 20 neighbors for each gene in the network\n",
    "TOP = 20\n",
    "\n",
    "grnutils.get_centrality(grn_h, TOP, top_k_to_disp=0)\n",
    "grnutils.get_centrality(grn_m, TOP, top_k_to_disp=0)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "9178cda6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "index\n",
       "TMSB4X     0.276147\n",
       "B2M        0.276147\n",
       "FTH1       0.276147\n",
       "FTL        0.276147\n",
       "HLA-DRA    0.276147\n",
       "EIF1       0.276147\n",
       "CD74       0.275399\n",
       "C1QB       0.242367\n",
       "C1QA       0.231694\n",
       "PPIA       0.206782\n",
       "Name: centrality, dtype: float64"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "grn_h.var.centrality.sort_values(ascending=False).head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "be94dc7a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "symbol\n",
       "C1QC       0.274250\n",
       "APOE       0.274250\n",
       "TMSB4X     0.274250\n",
       "B2M        0.274249\n",
       "MT-CYTB    0.274011\n",
       "MT-ND1     0.274002\n",
       "FTH1       0.268971\n",
       "MT-ND4     0.260279\n",
       "FTL1       0.252664\n",
       "IFITM3     0.217691\n",
       "Name: centrality, dtype: float64"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "grn_m.var.centrality.sort_values(ascending=False).head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "84e42298",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'C1QC', 'IFITM3', 'APOE'}\n",
      "{'PPIA', 'CD74'}\n"
     ]
    }
   ],
   "source": [
    "TOP = 10\n",
    "# in m not in h\n",
    "print(\n",
    "    (\n",
    "        set(grn_m.var.centrality.sort_values(ascending=False).head(TOP).index)\n",
    "        - set(grn_h.var.centrality.sort_values(ascending=False).head(TOP).index)\n",
    "    )\n",
    "    & (set(grn_h.var.index) & set(grn_m.var.index))\n",
    ")\n",
    "# in h not in m\n",
    "print(\n",
    "    (\n",
    "        set(grn_h.var.centrality.sort_values(ascending=False).head(TOP).index)\n",
    "        - set(grn_m.var.centrality.sort_values(ascending=False).head(TOP).index)\n",
    "    )\n",
    "    & (set(grn_h.var.index) & set(grn_m.var.index))\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "1867ed56",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[18., 61., 82., ..., 67.,  1., 20.]], dtype=float32)"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.array(grn_m.X.sum(0))[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "f3b5399b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "SignificanceResult(statistic=0.34178112327196214, pvalue=2.832630747887451e-174)"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "scipy.stats.spearmanr(np.array(grn_m.X.sum(0))[0], grn_m.grn.sum(1).values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "e141d38e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.6804855772300833"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.corrcoef(np.array(grn_m.X.sum(0))[0], grn_m.grn.sum(0).values)[0, 1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "440f1708",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'CD5L', 'CTSD', 'APOE', 'RPS19', 'IFITM3'}\n",
      "{'PPIA', 'ITGB2', 'RPS6', 'S100A11', 'FN1', 'SOD2'}\n"
     ]
    }
   ],
   "source": [
    "TOP = 20\n",
    "# in m not in h\n",
    "print(\n",
    "    (\n",
    "        set(grn_m.var.centrality.sort_values(ascending=False).head(TOP).index)\n",
    "        - set(grn_h.var.centrality.sort_values(ascending=False).head(TOP).index)\n",
    "    )\n",
    "    & (set(grn_h.var.index) & set(grn_m.var.index))\n",
    ")\n",
    "# in h not in m\n",
    "print(\n",
    "    (\n",
    "        set(grn_h.var.centrality.sort_values(ascending=False).head(TOP).index)\n",
    "        - set(grn_m.var.centrality.sort_values(ascending=False).head(TOP).index)\n",
    "    )\n",
    "    & (set(grn_h.var.index) & set(grn_m.var.index))\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "15fbc504",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'FOLR2', 'IFITM3', 'CD52', 'CLEC4F', 'GRN', 'CTSD', 'CD5L', 'APOE', 'CFP', 'TTR', 'ZBTB20'}\n",
      "{'PPIA', 'RPS14', 'IFI30', 'S100A11', 'SPARCL1', 'RPS27', 'RPS6', 'LGALS1', 'CD163', 'PTP4A1', 'APOC1', 'S100A4', 'SRGN', 'SOD2', 'ITGB2', 'CST3', 'NME2', 'CCL4', 'FN1'}\n"
     ]
    }
   ],
   "source": [
    "TOP = 50\n",
    "# in m not in h\n",
    "print(\n",
    "    (\n",
    "        set(grn_m.var.centrality.sort_values(ascending=False).head(TOP).index)\n",
    "        - set(grn_h.var.centrality.sort_values(ascending=False).head(TOP).index)\n",
    "    )\n",
    "    & (set(grn_h.var.index) & set(grn_m.var.index))\n",
    ")\n",
    "# in h not in m\n",
    "print(\n",
    "    (\n",
    "        set(grn_h.var.centrality.sort_values(ascending=False).head(TOP).index)\n",
    "        - set(grn_m.var.centrality.sort_values(ascending=False).head(TOP).index)\n",
    "    )\n",
    "    & (set(grn_h.var.index) & set(grn_m.var.index))\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b54349fa",
   "metadata": {},
   "source": [
    "# 2. Network similarity\n",
    "\n",
    "We now look at the similarity of the two networks based on general overlap of their top K edges across their common nodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "e3ab1120",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.29805238415043656"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "K = 20\n",
    "subgrn_h = grn_h.get(common).grn\n",
    "subgrn_h = subgrn_h.apply(lambda row: row >= row.nlargest(K).min(), axis=1)\n",
    "\n",
    "subgrn_m = grn_m.get(common).grn\n",
    "subgrn_m = subgrn_m.apply(lambda row: row >= row.nlargest(K).min(), axis=1)\n",
    "(subgrn_m & subgrn_h).sum(1).mean() / K\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "c89986b5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "genes with most dissimilar top-K connections (mouse vs human):\n",
      "RPLP1        7\n",
      "HSPA8        8\n",
      "RPS14        8\n",
      "RPS27        8\n",
      "RPS19        9\n",
      "HMGN2        9\n",
      "HSPA1B       9\n",
      "HSPA1A       9\n",
      "HSPE1        9\n",
      "IFIT2       10\n",
      "SERBP1      10\n",
      "KLF4        10\n",
      "PPIA        10\n",
      "IFIT1       10\n",
      "COX5A       10\n",
      "SEC61G      10\n",
      "RAN         10\n",
      "EGR1        10\n",
      "RPS6        10\n",
      "IER2        10\n",
      "CCL4        10\n",
      "NME2        10\n",
      "HSPA5       10\n",
      "IFIT3       10\n",
      "HSP90AA1    10\n",
      "HNRNPA1     10\n",
      "CD74        11\n",
      "REL         11\n",
      "PPP1R14B    11\n",
      "FOSL2       11\n",
      "S100A9      11\n",
      "MAF         11\n",
      "S100A8      11\n",
      "NR4A2       11\n",
      "STAB1       11\n",
      "MRPL21      11\n",
      "ISG15       11\n",
      "MRC1        11\n",
      "SET         11\n",
      "HMGB2       11\n",
      "U2AF1       11\n",
      "PSMB5       11\n",
      "CCRL2       11\n",
      "HSPD1       11\n",
      "DNAJB1      11\n",
      "NR4A1       11\n",
      "HSPB1       11\n",
      "IGF1        11\n",
      "STAT1       11\n",
      "DCN         11\n",
      "SRGN        11\n",
      "POLR2E      11\n",
      "NSA2        11\n",
      "ZFP36       11\n",
      "NAA38       11\n",
      "LYVE1       11\n",
      "IFITM3      11\n",
      "HEXB        11\n",
      "IFITM1      11\n",
      "TUBB4B      11\n",
      "dtype: int64\n",
      "\n",
      "Most similar genes: ['RPLP1', 'HSPA8', 'RPS14', 'RPS27', 'RPS19', 'HMGN2', 'HSPA1B', 'HSPA1A', 'HSPE1', 'IFIT2', 'SERBP1', 'KLF4', 'PPIA', 'IFIT1', 'COX5A', 'SEC61G', 'RAN', 'EGR1', 'RPS6', 'IER2', 'CCL4', 'NME2', 'HSPA5', 'IFIT3', 'HSP90AA1', 'HNRNPA1', 'CD74', 'REL', 'PPP1R14B', 'FOSL2', 'S100A9', 'MAF', 'S100A8', 'NR4A2', 'STAB1', 'MRPL21', 'ISG15', 'MRC1', 'SET', 'HMGB2', 'U2AF1', 'PSMB5', 'CCRL2', 'HSPD1', 'DNAJB1', 'NR4A1', 'HSPB1', 'IGF1', 'STAT1', 'DCN', 'SRGN', 'POLR2E', 'NSA2', 'ZFP36', 'NAA38', 'LYVE1', 'IFITM3', 'HEXB', 'IFITM1', 'TUBB4B']\n"
     ]
    }
   ],
   "source": [
    "# Calculate similar for each common CD gene in human compared to mouse\n",
    "dissim_scores = {}\n",
    "for gene in subgrn_m.index:\n",
    "    # Count overlapping connections (human vs mouse)\n",
    "    overlap = (subgrn_m.loc[gene] & subgrn_h.loc[gene]).sum()\n",
    "    # Dissimilarity score: how many top-K human connections are NOT present in mouse\n",
    "    dissim_scores[gene] = (subgrn_m.loc[gene] & ~subgrn_h.loc[gene]).sum()\n",
    "\n",
    "# Sort genes by similar (highest first)\n",
    "sim_df = pd.Series(dissim_scores).sort_values(ascending=True)\n",
    "print(\"genes with most dissimilar top-K connections (mouse vs human):\")\n",
    "print(sim_df.head(60))\n",
    "\n",
    "# Show the most similar genes\n",
    "most_similar = sim_df.head(60).index.tolist()\n",
    "print(f\"\\nMost similar genes: {most_similar}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "2d7a52e2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "genes with most dissimilar top-K connections (mouse vs human):\n",
      "TSPAN6       17\n",
      "SLCO2A1      17\n",
      "RPP38        17\n",
      "TSPAN15      17\n",
      "PLXNA3       17\n",
      "RANGRF       17\n",
      "ANLN         17\n",
      "EFHD1        17\n",
      "ID4          17\n",
      "MRPL44       16\n",
      "JMJD6        16\n",
      "AIMP2        16\n",
      "PRR5         16\n",
      "SPATS2L      16\n",
      "RPP25L       16\n",
      "ATIC         16\n",
      "XRCC5        16\n",
      "DNAJC9       16\n",
      "TMEM107      16\n",
      "ITGAE        16\n",
      "TSPAN2       16\n",
      "GEMIN6       16\n",
      "PES1         16\n",
      "RWDD2A       16\n",
      "EPOR         16\n",
      "PFN2         16\n",
      "SCAMP3       16\n",
      "ENPEP        16\n",
      "ARC          16\n",
      "ROR1         16\n",
      "ATP2A2       16\n",
      "ANGPTL4      16\n",
      "ADH4         16\n",
      "MAFK         16\n",
      "BORCS8       16\n",
      "DDX27        16\n",
      "CCDC43       16\n",
      "MTFMT        16\n",
      "GTF2F1       16\n",
      "SYAP1        16\n",
      "TMEM70       16\n",
      "ALDH5A1      16\n",
      "EXOSC10      16\n",
      "TPRKB        16\n",
      "DARS2        16\n",
      "MTMR11       16\n",
      "RGS5         16\n",
      "NUF2         16\n",
      "SLC26A4      16\n",
      "TACC3        16\n",
      "IBA57        16\n",
      "PPP1R15B     16\n",
      "NFE2L3       16\n",
      "RIPK4        16\n",
      "GCH1         16\n",
      "EEF1AKMT1    16\n",
      "GJB2         16\n",
      "TMEM242      16\n",
      "ODF3B        16\n",
      "RASD1        16\n",
      "dtype: int64\n",
      "\n",
      "Most dissimilar genes: ['TSPAN6', 'SLCO2A1', 'RPP38', 'TSPAN15', 'PLXNA3', 'RANGRF', 'ANLN', 'EFHD1', 'ID4', 'MRPL44', 'JMJD6', 'AIMP2', 'PRR5', 'SPATS2L', 'RPP25L', 'ATIC', 'XRCC5', 'DNAJC9', 'TMEM107', 'ITGAE', 'TSPAN2', 'GEMIN6', 'PES1', 'RWDD2A', 'EPOR', 'PFN2', 'SCAMP3', 'ENPEP', 'ARC', 'ROR1', 'ATP2A2', 'ANGPTL4', 'ADH4', 'MAFK', 'BORCS8', 'DDX27', 'CCDC43', 'MTFMT', 'GTF2F1', 'SYAP1', 'TMEM70', 'ALDH5A1', 'EXOSC10', 'TPRKB', 'DARS2', 'MTMR11', 'RGS5', 'NUF2', 'SLC26A4', 'TACC3', 'IBA57', 'PPP1R15B', 'NFE2L3', 'RIPK4', 'GCH1', 'EEF1AKMT1', 'GJB2', 'TMEM242', 'ODF3B', 'RASD1']\n"
     ]
    }
   ],
   "source": [
    "# Calculate dissimilarity for each common CD gene in human compared to mouse\n",
    "dissim_scores = {}\n",
    "for gene in subgrn_m.index:\n",
    "    # Count overlapping connections (human vs mouse)\n",
    "    overlap = (subgrn_m.loc[gene] & subgrn_h.loc[gene]).sum()\n",
    "    # Dissimilarity score: how many top-K human connections are NOT present in mouse\n",
    "    dissim_scores[gene] = (subgrn_m.loc[gene] & ~subgrn_h.loc[gene]).sum()\n",
    "\n",
    "# Sort genes by dissimilarity (highest first)\n",
    "dissim_df = pd.Series(dissim_scores).sort_values(ascending=False)\n",
    "print(\"genes with most dissimilar top-K connections (mouse vs human):\")\n",
    "print(dissim_df.head(60))\n",
    "\n",
    "# Show the most dissimilar genes\n",
    "most_dissimilar = dissim_df.head(60).index.tolist()\n",
    "print(f\"\\nMost dissimilar genes: {most_dissimilar}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "d5652e69",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_set</th>\n",
       "      <th>Term</th>\n",
       "      <th>P-value</th>\n",
       "      <th>Adjusted P-value</th>\n",
       "      <th>Old P-value</th>\n",
       "      <th>Old adjusted P-value</th>\n",
       "      <th>Odds Ratio</th>\n",
       "      <th>Combined Score</th>\n",
       "      <th>Genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GO_Molecular_Function_2023</td>\n",
       "      <td>Hydroxymethyl-, Formyl- And Related Transferas...</td>\n",
       "      <td>0.000399</td>\n",
       "      <td>0.043522</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>inf</td>\n",
       "      <td>inf</td>\n",
       "      <td>MTFMT;ATIC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Gene_set  \\\n",
       "0  GO_Molecular_Function_2023   \n",
       "\n",
       "                                                Term   P-value  \\\n",
       "0  Hydroxymethyl-, Formyl- And Related Transferas...  0.000399   \n",
       "\n",
       "   Adjusted P-value  Old P-value  Old adjusted P-value  Odds Ratio  \\\n",
       "0          0.043522            0                     0         inf   \n",
       "\n",
       "   Combined Score       Genes  \n",
       "0             inf  MTFMT;ATIC  "
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "enr = gp.enrichr(\n",
    "    gene_list=most_dissimilar,\n",
    "    gene_sets=[\n",
    "        \"GO_Molecular_Function_2023\"\n",
    "    ],  # , 'MSigDB_Hallmark_2020', 'Reactome_2022', 'Tabula_Sapiens', 'WikiPathway_2023_Human', 'TF_Perturbations_Followed_by_Expression', 'PPI_Hub_Proteins', 'OMIM_Disease', 'GO_Molecular_Function_2023'],\n",
    "    organism=\"Human\",  # change accordingly\n",
    "    # description='pathway',\n",
    "    cutoff=0.05,  # test dataset, use lower value for real case\n",
    "    background=common,\n",
    ")\n",
    "enr.res2d[enr.res2d[\"Adjusted P-value\"] < 0.05].head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53dc894e",
   "metadata": {},
   "outputs": [],
   "source": [
    "subgrn_m.pl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4ab0a832",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "dd3ec4f4",
   "metadata": {},
   "source": [
    "# res \n",
    "\n",
    "- seeing CD34 first is really cool (when most dissim human / mouse)\n",
    "  \n",
    "# pathway analysis\n",
    "## mouse vs human\n",
    "- many cell cyle genes\n",
    "- replication genes\n",
    "- G2-M Checkpoint / Mitotic Spindle\n",
    "- DNA synthesis and replication, unwinding of dna...\n",
    "- microtubulue, and tubulin stuff\n",
    "- CDK1, MCM2, ..\n",
    "## human vs mouse\n",
    "- cytokine signaling\n",
    "- Fluoropyrimidine, Pyrimidine and serine metabolism\n",
    "- same otherwise\n",
    "\n",
    "key differences in B-cell this also changes how they respond to cytokines and influence in how they divide and replicate, \n",
    "That difference can be in part driven by differences in nucleotide metabolism, which is key for DNA synthesis and replication,\n",
    "\n",
    "\t\tOrtholog expression / enzyme activities differ: The DPD example shows mouse > human in hepatic activity. That suggests species have evolved different baseline enzyme capacities.\n",
    "\t\tOrgan/tissue distribution of enzymes differs: The capecitabine study showed different organ/intestinal vs hepatic enzyme distributions in mice vs humans. That implies that even with the same genes, tissueexpression context differs across species.\n",
    "\t\tDifferences in precursor supply / flux: For serine, de novo synthesis vs uptake differs between species/tissue contexts. So B-cells in mouse may rely more on one mode (say, uptake) vs human B-cells another.\n",
    "\t\tDifferences in regulation by signalling or metabolism: Immune cells in different species may use different metabolic wiring (e.g., nutrient transporter expression, mitochondrial vs glycolytic flux). These differences may ripple through nucleotide and serine metabolism.\n",
    "\t\tEvolutionary adaptation to lifespan, cell turnover, and metabolic demand: Mice have much faster cell proliferation rates, higher basal metabolic rates, shorter lifespans; this may drive differences in how strongly they rely on de novo synthesis vs salvage pathways, or how tightly pyrimidine catabolism is regulated.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c867b46e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "genes with most dissimilar top-K connections (human vs mouse):\n",
      "GINS1       18\n",
      "KIF22       18\n",
      "UBE3A       18\n",
      "MELK        18\n",
      "SDC4        18\n",
      "RFC3        18\n",
      "GCH1        18\n",
      "ACTR10      18\n",
      "NSD1        18\n",
      "PRR11       18\n",
      "SHCBP1      18\n",
      "SIVA1       18\n",
      "BAZ1B       18\n",
      "UBE2C       18\n",
      "DMKN        18\n",
      "PARPBP      18\n",
      "RRM2        18\n",
      "TXN2        18\n",
      "CENPK       18\n",
      "SPDL1       18\n",
      "RAB11A      18\n",
      "MCM7        18\n",
      "MCM4        18\n",
      "SLC25A39    18\n",
      "CCNB1       18\n",
      "TOP2A       18\n",
      "CCNL2       18\n",
      "CENPH       18\n",
      "NUSAP1      18\n",
      "KIF23       18\n",
      "DHFR        18\n",
      "SEC63       18\n",
      "BRCA1       18\n",
      "SNRNP27     18\n",
      "GLRX5       18\n",
      "ASF1B       18\n",
      "GSDMD       18\n",
      "SPPL2A      18\n",
      "GINS2       18\n",
      "RRM1        18\n",
      "RRP7A       18\n",
      "RACGAP1     18\n",
      "FEN1        18\n",
      "TYMS        18\n",
      "PLK1        18\n",
      "CCNB2       18\n",
      "AK2         18\n",
      "SUCLG1      18\n",
      "CDC20       18\n",
      "GMNN        18\n",
      "RFC4        18\n",
      "DNAJC19     18\n",
      "DLGAP5      18\n",
      "PAICS       18\n",
      "NSRP1       18\n",
      "MRPL34      18\n",
      "EXOSC8      18\n",
      "SPC25       18\n",
      "KCNK1       18\n",
      "NDUFA4L2    18\n",
      "dtype: int64\n",
      "\n",
      "Most dissimilar genes: ['GINS1', 'KIF22', 'UBE3A', 'MELK', 'SDC4', 'RFC3', 'GCH1', 'ACTR10', 'NSD1', 'PRR11', 'SHCBP1', 'SIVA1', 'BAZ1B', 'UBE2C', 'DMKN', 'PARPBP', 'RRM2', 'TXN2', 'CENPK', 'SPDL1', 'RAB11A', 'MCM7', 'MCM4', 'SLC25A39', 'CCNB1', 'TOP2A', 'CCNL2', 'CENPH', 'NUSAP1', 'KIF23', 'DHFR', 'SEC63', 'BRCA1', 'SNRNP27', 'GLRX5', 'ASF1B', 'GSDMD', 'SPPL2A', 'GINS2', 'RRM1', 'RRP7A', 'RACGAP1', 'FEN1', 'TYMS', 'PLK1', 'CCNB2', 'AK2', 'SUCLG1', 'CDC20', 'GMNN', 'RFC4', 'DNAJC19', 'DLGAP5', 'PAICS', 'NSRP1', 'MRPL34', 'EXOSC8', 'SPC25', 'KCNK1', 'NDUFA4L2']\n"
     ]
    }
   ],
   "source": [
    "# Calculate dissimilarity for each common CD gene in human compared to mouse\n",
    "dissim_scores = {}\n",
    "for gene in subgrn_h.index:\n",
    "    # Count overlapping connections (human vs mouse)\n",
    "    overlap = (subgrn_h.loc[gene] & subgrn_m.loc[gene]).sum()\n",
    "    # Dissimilarity score: how many top-K human connections are NOT present in mouse\n",
    "    dissim_scores[gene] = (subgrn_h.loc[gene] & ~subgrn_m.loc[gene]).sum()\n",
    "\n",
    "# Sort genes by dissimilarity (highest first)\n",
    "dissim_df = pd.Series(dissim_scores).sort_values(ascending=False)\n",
    "print(\"genes with most dissimilar top-K connections (human vs mouse):\")\n",
    "print(dissim_df.head(60))\n",
    "\n",
    "# Show the most dissimilar genes\n",
    "most_dissimilar = dissim_df.head(60).index.tolist()\n",
    "print(f\"\\nMost dissimilar genes: {most_dissimilar}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19b95975",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['CDC20', 'CDK1', 'CDC16', 'CD34', 'CDCA3', 'CDKAL1', 'CDO1', 'CDK4',\n",
       "       'CD200R1', 'CDC34', 'CDKN2AIPNL', 'CDK14', 'CDKN1A', 'CDC37L1', 'CDC26',\n",
       "       'CD36', 'CD93', 'CD68', 'CD83', 'CD14', 'CD69', 'CD40', 'CD47', 'CDC37',\n",
       "       'CD63', 'CD46', 'CD52', 'CD3D'],\n",
       "      dtype='object')"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "dissim_df.index[dissim_df.index.str.startswith(\"CD\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "35a49081",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "np.float64(0.5857142857142856)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "K = 5\n",
    "subgrn_h = grn_h.get(\n",
    "    set(common) & set(grn_h.var.index[grn_h.var.index.str.match(r\"CD\\d{1,4}\")])\n",
    ").grn\n",
    "subgrn_h = subgrn_h.apply(lambda row: row >= row.nlargest(K).min(), axis=1)\n",
    "\n",
    "subgrn_m = grn_m.get(set(common) & set(grn_m.var.index[grn_m.var.index.str.match(r\"CD\\d{1,4}\")])).grn\n",
    "subgrn_m = subgrn_m.apply(lambda row: row >= row.nlargest(K).min(), axis=1)\n",
    "(subgrn_m & subgrn_h).sum(1).mean() / K"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ddc21e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "https://chatgpt.com/share/68ee40f1-7380-8000-b5ef-585db46704e9"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1546a0f8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "54627e63",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "47fe8ff0",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scPRINT (.venv)",
   "language": "python",
   "name": "scprint-venv"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
