{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## working with ESM2 embeddings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! esm-extract esm2_t33_650M_UR50D data/temp/test_esm.fasta data/temp/ --include mean"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [],
   "source": [
    "import io\n",
    "import os\n",
    "from biomart import BiomartServer\n",
    "\n",
    "def _fetchFromServer(ensemble_server, attributes):\n",
    "    server = BiomartServer(ensemble_server, verbose=True)\n",
    "    ensmbl = server.datasets[\"hsapiens_gene_ensembl\"]\n",
    "    print(attributes)\n",
    "    res = pd.read_csv(\n",
    "        io.StringIO(\n",
    "            ensmbl.search({\"attributes\": attributes}, header=1).content.decode()\n",
    "        ),\n",
    "        sep=\"\\t\",\n",
    "    )\n",
    "    return res\n",
    "\n",
    "def createFoldersFor(filepath):\n",
    "    \"\"\"\n",
    "    will recursively create folders if needed until having all the folders required to save the file in this filepath\n",
    "    \"\"\"\n",
    "    prevval = \"\"\n",
    "    for val in os.path.expanduser(filepath).split(\"/\")[:-1]:\n",
    "        prevval += val + \"/\"\n",
    "        if not os.path.exists(prevval):\n",
    "            os.mkdir(prevval)\n",
    "\n",
    "def getBiomartTable(\n",
    "    ensemble_server=\"http://feb2023.archive.ensembl.org/biomart\",\n",
    "    useCache=False,\n",
    "    cache_folder=\"/tmp/biomart/\",\n",
    "    attributes=[],\n",
    "    bypass_attributes=False,\n",
    "):\n",
    "    \"\"\"generate a genelist dataframe from ensembl's biomart\n",
    "\n",
    "    Args:\n",
    "        ensemble_server ([type], optional): [description]. Defaults to ENSEMBL_SERVER_V.\n",
    "        useCache (bool, optional): [description]. Defaults to False.\n",
    "        cache_folder ([type], optional): [description]. Defaults to CACHE_PATH.\n",
    "\n",
    "    Raises:\n",
    "        ValueError: [description]\n",
    "\n",
    "    Returns:\n",
    "        [type]: [description]\n",
    "    \"\"\"\n",
    "    attr = (\n",
    "        [\n",
    "            \"ensembl_gene_id\",\n",
    "            \"hgnc_symbol\",\n",
    "            \"gene_biotype\",\n",
    "            \"entrezgene_id\",\n",
    "        ]\n",
    "        if not bypass_attributes\n",
    "        else []\n",
    "    )\n",
    "    assert cache_folder[-1] == \"/\"\n",
    "\n",
    "    cache_folder = os.path.expanduser(cache_folder)\n",
    "    createFoldersFor(cache_folder)\n",
    "    cachefile = os.path.join(cache_folder, \".biomart.csv\")\n",
    "    if useCache & os.path.isfile(cachefile):\n",
    "        print(\"fetching gene names from biomart cache\")\n",
    "        res = pd.read_csv(cachefile)\n",
    "    else:\n",
    "        print(\"downloading gene names from biomart\")\n",
    "\n",
    "        res = _fetchFromServer(ensemble_server, attr + attributes)\n",
    "        res.to_csv(cachefile, index=False)\n",
    "    \n",
    "    res.columns = attr + attributes\n",
    "    if type(res) is not type(pd.DataFrame()):\n",
    "        raise ValueError(\"should be a dataframe\")\n",
    "    res = res[~(res[\"ensembl_gene_id\"].isna() & res[\"hgnc_symbol\"].isna())]\n",
    "    res.loc[res[res.hgnc_symbol.isna()].index, \"hgnc_symbol\"] = res[\n",
    "        res.hgnc_symbol.isna()\n",
    "    ][\"ensembl_gene_id\"]\n",
    "\n",
    "    return res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "downloading gene names from biomart\n",
      "[BiomartServer:'http://feb2023.archive.ensembl.org/biomart/martservice'] is alive.\n",
      "[BiomartServer:'http://feb2023.archive.ensembl.org/biomart/martservice'] Fetching datasets\n",
      "[BiomartServer:'http://feb2023.archive.ensembl.org/biomart/martservice'] Fetching databases\n",
      "[BiomartDatabase:'Ensembl Genes 109'] Fetching datasets\n",
      "[BiomartDatabase:'Mouse strains 109'] Fetching datasets\n",
      "[BiomartDatabase:'Sequence'] Fetching datasets\n",
      "[BiomartDatabase:'Ontology'] Fetching datasets\n",
      "[BiomartDatabase:'Genomic features 109'] Fetching datasets\n",
      "[BiomartDatabase:'Ensembl Variation 109'] Fetching datasets\n",
      "[BiomartDatabase:'Ensembl Regulation 109'] Fetching datasets\n",
      "['ensembl_gene_id', 'hgnc_symbol', 'gene_biotype', 'entrezgene_id']\n",
      "[BiomartDataset:'hsapiens_gene_ensembl'] Searching using following params:\n",
      "{'attributes': ['ensembl_gene_id',\n",
      "                'hgnc_symbol',\n",
      "                'gene_biotype',\n",
      "                'entrezgene_id']}\n",
      "[BiomartDataset:'hsapiens_gene_ensembl'] Fetching attributes\n",
      "[BiomartDataset] search query:\n",
      "b'<Query virtualSchemaName=\"default\" formatter=\"TSV\" header=\"1\" uniqueRows=\"1\" datasetConfigVersion=\"0.6\" count=\"\"><Dataset name=\"hsapiens_gene_ensembl\" interface=\"default\"><Attribute name=\"ensembl_gene_id\" /><Attribute name=\"hgnc_symbol\" /><Attribute name=\"gene_biotype\" /><Attribute name=\"entrezgene_id\" /></Dataset></Query>'\n"
     ]
    }
   ],
   "source": [
    "biomart2 = getBiomartTable()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "downloading gene names from biomart\n",
      "[BiomartServer:'http://feb2023.archive.ensembl.org/biomart/martservice'] is alive.\n",
      "[BiomartServer:'http://feb2023.archive.ensembl.org/biomart/martservice'] Fetching datasets\n",
      "[BiomartServer:'http://feb2023.archive.ensembl.org/biomart/martservice'] Fetching databases\n",
      "[BiomartDatabase:'Ensembl Genes 109'] Fetching datasets\n",
      "[BiomartDatabase:'Mouse strains 109'] Fetching datasets\n",
      "[BiomartDatabase:'Sequence'] Fetching datasets\n",
      "[BiomartDatabase:'Ontology'] Fetching datasets\n",
      "[BiomartDatabase:'Genomic features 109'] Fetching datasets\n",
      "[BiomartDatabase:'Ensembl Variation 109'] Fetching datasets\n",
      "[BiomartDatabase:'Ensembl Regulation 109'] Fetching datasets\n",
      "['ensembl_gene_id', 'hgnc_symbol', 'gene_biotype', 'entrezgene_id', 'ensembl_transcript_id', 'protein_id', 'ensembl_peptide_id']\n",
      "[BiomartDataset:'hsapiens_gene_ensembl'] Searching using following params:\n",
      "{'attributes': ['ensembl_gene_id',\n",
      "                'hgnc_symbol',\n",
      "                'gene_biotype',\n",
      "                'entrezgene_id',\n",
      "                'ensembl_transcript_id',\n",
      "                'protein_id',\n",
      "                'ensembl_peptide_id']}\n",
      "[BiomartDataset:'hsapiens_gene_ensembl'] Fetching attributes\n",
      "[BiomartDataset] search query:\n",
      "b'<Query virtualSchemaName=\"default\" formatter=\"TSV\" header=\"1\" uniqueRows=\"1\" datasetConfigVersion=\"0.6\" count=\"\"><Dataset name=\"hsapiens_gene_ensembl\" interface=\"default\"><Attribute name=\"ensembl_gene_id\" /><Attribute name=\"hgnc_symbol\" /><Attribute name=\"gene_biotype\" /><Attribute name=\"entrezgene_id\" /><Attribute name=\"ensembl_transcript_id\" /><Attribute name=\"protein_id\" /><Attribute name=\"ensembl_peptide_id\" /></Dataset></Query>'\n"
     ]
    }
   ],
   "source": [
    "biomart = getBiomartTable(attributes=               \n",
    "    [\"ensembl_transcript_id\",\n",
    "        \"protein_id\",\n",
    "        \"ensembl_peptide_id\",\n",
    "        #\"peptide\",\n",
    "       # \"uniprotswissprot\",\n",
    "    ], bypass_attributes=False, useCache=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>hgnc_symbol</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>entrezgene_id</th>\n",
       "      <th>ensembl_transcript_id</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ensembl_peptide_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [ensembl_gene_id, hgnc_symbol, gene_biotype, entrezgene_id, ensembl_transcript_id, protein_id, ensembl_peptide_id]\n",
       "Index: []"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "biomart[biomart.gene_biotype.isin(['Mt_tRNA', 'snRNA', 'sRNA'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>hgnc_symbol</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>entrezgene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000210049</td>\n",
       "      <td>MT-TF</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000210077</td>\n",
       "      <td>MT-TV</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000209082</td>\n",
       "      <td>MT-TL1</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENSG00000210100</td>\n",
       "      <td>MT-TI</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENSG00000210107</td>\n",
       "      <td>MT-TQ</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENSG00000210112</td>\n",
       "      <td>MT-TM</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ENSG00000210117</td>\n",
       "      <td>MT-TW</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENSG00000210127</td>\n",
       "      <td>MT-TA</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ENSG00000210135</td>\n",
       "      <td>MT-TN</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ENSG00000210140</td>\n",
       "      <td>MT-TC</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ENSG00000210144</td>\n",
       "      <td>MT-TY</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ENSG00000210151</td>\n",
       "      <td>MT-TS1</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>113219467.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ENSG00000210154</td>\n",
       "      <td>MT-TD</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ENSG00000210156</td>\n",
       "      <td>MT-TK</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>ENSG00000210164</td>\n",
       "      <td>MT-TG</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>ENSG00000210174</td>\n",
       "      <td>MT-TR</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>ENSG00000210176</td>\n",
       "      <td>MT-TH</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>ENSG00000210184</td>\n",
       "      <td>MT-TS2</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>ENSG00000210191</td>\n",
       "      <td>MT-TL2</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>ENSG00000210194</td>\n",
       "      <td>MT-TE</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    ensembl_gene_id hgnc_symbol gene_biotype  entrezgene_id\n",
       "0   ENSG00000210049       MT-TF      Mt_tRNA            NaN\n",
       "2   ENSG00000210077       MT-TV      Mt_tRNA            NaN\n",
       "4   ENSG00000209082      MT-TL1      Mt_tRNA            NaN\n",
       "6   ENSG00000210100       MT-TI      Mt_tRNA            NaN\n",
       "7   ENSG00000210107       MT-TQ      Mt_tRNA            NaN\n",
       "8   ENSG00000210112       MT-TM      Mt_tRNA            NaN\n",
       "10  ENSG00000210117       MT-TW      Mt_tRNA            NaN\n",
       "11  ENSG00000210127       MT-TA      Mt_tRNA            NaN\n",
       "12  ENSG00000210135       MT-TN      Mt_tRNA            NaN\n",
       "13  ENSG00000210140       MT-TC      Mt_tRNA            NaN\n",
       "14  ENSG00000210144       MT-TY      Mt_tRNA            NaN\n",
       "16  ENSG00000210151      MT-TS1      Mt_tRNA    113219467.0\n",
       "17  ENSG00000210154       MT-TD      Mt_tRNA            NaN\n",
       "19  ENSG00000210156       MT-TK      Mt_tRNA            NaN\n",
       "23  ENSG00000210164       MT-TG      Mt_tRNA            NaN\n",
       "25  ENSG00000210174       MT-TR      Mt_tRNA            NaN\n",
       "28  ENSG00000210176       MT-TH      Mt_tRNA            NaN\n",
       "29  ENSG00000210184      MT-TS2      Mt_tRNA            NaN\n",
       "30  ENSG00000210191      MT-TL2      Mt_tRNA            NaN\n",
       "33  ENSG00000210194       MT-TE      Mt_tRNA            NaN"
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "biomart2[biomart2.gene_biotype.isin(['Mt_tRNA', 'snRNA', 'sRNA'])].head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>hgnc_symbol</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>entrezgene_id</th>\n",
       "      <th>ensembl_transcript_id</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ensembl_peptide_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000198888</td>\n",
       "      <td>MT-ND1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>4535.0</td>\n",
       "      <td>ENST00000361390</td>\n",
       "      <td>CAA24026</td>\n",
       "      <td>ENSP00000354687</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000198888</td>\n",
       "      <td>MT-ND1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>4535.0</td>\n",
       "      <td>ENST00000361390</td>\n",
       "      <td>AAB58943</td>\n",
       "      <td>ENSP00000354687</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000198888</td>\n",
       "      <td>MT-ND1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>4535.0</td>\n",
       "      <td>ENST00000361390</td>\n",
       "      <td>BAA07290</td>\n",
       "      <td>ENSP00000354687</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000198888</td>\n",
       "      <td>MT-ND1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>4535.0</td>\n",
       "      <td>ENST00000361390</td>\n",
       "      <td>AAP89036</td>\n",
       "      <td>ENSP00000354687</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000198888</td>\n",
       "      <td>MT-ND1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>4535.0</td>\n",
       "      <td>ENST00000361390</td>\n",
       "      <td>AAP89049</td>\n",
       "      <td>ENSP00000354687</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>901591</th>\n",
       "      <td>ENSG00000162437</td>\n",
       "      <td>RAVER2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>55225.0</td>\n",
       "      <td>ENST00000418058</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000397069</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>901592</th>\n",
       "      <td>ENSG00000122432</td>\n",
       "      <td>SPATA1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>100505741.0</td>\n",
       "      <td>ENST00000697276</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000514413</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>901593</th>\n",
       "      <td>ENSG00000122432</td>\n",
       "      <td>SPATA1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>100505741.0</td>\n",
       "      <td>ENST00000699524</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000514414</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>901594</th>\n",
       "      <td>ENSG00000122432</td>\n",
       "      <td>SPATA1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>100505741.0</td>\n",
       "      <td>ENST00000697277</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000514416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>901595</th>\n",
       "      <td>ENSG00000122432</td>\n",
       "      <td>SPATA1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>100505741.0</td>\n",
       "      <td>ENST00000460286</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000514417</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>900922 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        ensembl_gene_id hgnc_symbol    gene_biotype  entrezgene_id  \\\n",
       "0       ENSG00000198888      MT-ND1  protein_coding         4535.0   \n",
       "1       ENSG00000198888      MT-ND1  protein_coding         4535.0   \n",
       "2       ENSG00000198888      MT-ND1  protein_coding         4535.0   \n",
       "3       ENSG00000198888      MT-ND1  protein_coding         4535.0   \n",
       "4       ENSG00000198888      MT-ND1  protein_coding         4535.0   \n",
       "...                 ...         ...             ...            ...   \n",
       "901591  ENSG00000162437      RAVER2  protein_coding        55225.0   \n",
       "901592  ENSG00000122432      SPATA1  protein_coding    100505741.0   \n",
       "901593  ENSG00000122432      SPATA1  protein_coding    100505741.0   \n",
       "901594  ENSG00000122432      SPATA1  protein_coding    100505741.0   \n",
       "901595  ENSG00000122432      SPATA1  protein_coding    100505741.0   \n",
       "\n",
       "       ensembl_transcript_id protein_id ensembl_peptide_id  \n",
       "0            ENST00000361390   CAA24026    ENSP00000354687  \n",
       "1            ENST00000361390   AAB58943    ENSP00000354687  \n",
       "2            ENST00000361390   BAA07290    ENSP00000354687  \n",
       "3            ENST00000361390   AAP89036    ENSP00000354687  \n",
       "4            ENST00000361390   AAP89049    ENSP00000354687  \n",
       "...                      ...        ...                ...  \n",
       "901591       ENST00000418058        NaN    ENSP00000397069  \n",
       "901592       ENST00000697276        NaN    ENSP00000514413  \n",
       "901593       ENST00000699524        NaN    ENSP00000514414  \n",
       "901594       ENST00000697277        NaN    ENSP00000514416  \n",
       "901595       ENST00000460286        NaN    ENSP00000514417  \n",
       "\n",
       "[900922 rows x 7 columns]"
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "main_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [],
   "source": [
    "import gget"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "47.0"
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(ls)/2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ENST00000576342'"
      ]
     },
     "execution_count": 124,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ls[val*2].split(' ')[0][1:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "res = []\n",
    "size = 600\n",
    "from multiprocessing import Pool\n",
    "\n",
    "\n",
    "genes = biomart2[biomart2.gene_biotype.isin(['protein_coding'])].ensembl_gene_id.tolist()\n",
    "groups = [genes[i*size:(i+1)*size] for i in range(len(genes)//size)] + [genes[(len(genes)//size)*size:]]\n",
    "with Pool() as p:\n",
    "    results = [p.apply(gget.seq, args=(sub, True, False)) for sub in groups]\n",
    "\n",
    "res = [item for sublist in results for item in sublist]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scprint.utils import get_seq\n",
    "%reload_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Sun Nov 12 16:33:29 2023 INFO Requesting amino acid sequence of the canonical transcript ENST00000361390 of gene ENSG00000198888 from UniProt.\n",
      "Sun Nov 12 16:33:29 2023 INFO Requesting amino acid sequence of the canonical transcript ENST00000361453 of gene ENSG00000198763 from UniProt.\n",
      "Sun Nov 12 16:33:29 2023 INFO Requesting amino acid sequence of the canonical transcript ENST00000361624 of gene ENSG00000198804 from UniProt.\n",
      "Sun Nov 12 16:33:29 2023 INFO Requesting amino acid sequence of the canonical transcript ENST00000361739 of gene ENSG00000198712 from UniProt.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "> \u001b[0;32m/Users/jkobject/Documents/code/scPRINT/scprint/utils/get_seq.py\u001b[0m(300)\u001b[0;36mseq\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32m    298 \u001b[0;31m            \u001b[0;32mimport\u001b[0m \u001b[0mpdb\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0m\u001b[0;32m    299 \u001b[0;31m            \u001b[0mpdb\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mset_trace\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0m\u001b[0;32m--> 300 \u001b[0;31m            \u001b[0mdf_uniprot\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdf_uniprot\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0minfo_df\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mon\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"canonical_transcript\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mhow\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"inner\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0m\u001b[0;32m    301 \u001b[0;31m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0m\u001b[0;32m    302 \u001b[0;31m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0m\n",
      "             index         ensembl_id uniprot_id pdb_id ncbi_gene_id  \\\n",
      "0  ENSG00000198888  ENSG00000198888.2        NaN    NaN          NaN   \n",
      "1  ENSG00000198763  ENSG00000198763.3        NaN    NaN          NaN   \n",
      "2  ENSG00000198804  ENSG00000198804.2        NaN    NaN          NaN   \n",
      "3  ENSG00000198712  ENSG00000198712.1        NaN    NaN          NaN   \n",
      "\n",
      "        species assembly_name primary_gene_name ensembl_gene_name synonyms  \\\n",
      "0  homo_sapiens        GRCh38               NaN            MT-ND1      NaN   \n",
      "1  homo_sapiens        GRCh38               NaN            MT-ND2      NaN   \n",
      "2  homo_sapiens        GRCh38               NaN            MT-CO1      NaN   \n",
      "3  homo_sapiens        GRCh38               NaN            MT-CO2      NaN   \n",
      "\n",
      "   ... transcript_names transcript_strands transcript_starts transcript_ends  \\\n",
      "0  ...     [MT-ND1-201]                [1]            [3307]          [4262]   \n",
      "1  ...     [MT-ND2-201]                [1]            [4470]          [5511]   \n",
      "2  ...     [MT-CO1-201]                [1]            [5904]          [7445]   \n",
      "3  ...     [MT-CO2-201]                [1]            [7586]          [8269]   \n",
      "\n",
      "  all_exons exon_starts exon_ends all_translations translation_starts  \\\n",
      "0       NaN         NaN       NaN              NaN                NaN   \n",
      "1       NaN         NaN       NaN              NaN                NaN   \n",
      "2       NaN         NaN       NaN              NaN                NaN   \n",
      "3       NaN         NaN       NaN              NaN                NaN   \n",
      "\n",
      "  translation_ends  \n",
      "0              NaN  \n",
      "1              NaN  \n",
      "2              NaN  \n",
      "3              NaN  \n",
      "\n",
      "[4 rows x 35 columns]\n",
      "*** KeyError: 'canonical_transcript'\n",
      "  uniprot_id      organism                                           sequence  \\\n",
      "0     P03886  Homo sapiens  MPMANLLLLIVPILIAMAFLMLTERKILGYMQLRKGPNVVGPYGLL...   \n",
      "0     P03891  Homo sapiens  MNPLAQPVIYSTIFAGTLITALSSHWFFTWVGLEMNMLAFIPVLTK...   \n",
      "0     P00395  Homo sapiens  MFADRWLFSTNHKDIGTLYLLFGAWAGVLGTALSLLIRAELGQPGN...   \n",
      "0     P00403  Homo sapiens  MAHAAQVGLQDATSPIMEELITFHDHALMIIFLICFLVLYALFLTL...   \n",
      "\n",
      "   sequence_length gene_name            query  \n",
      "0              318    MT-ND1  ENST00000361390  \n",
      "0              347    MT-ND2  ENST00000361453  \n",
      "0              513    MT-CO1  ENST00000361624  \n",
      "0              227    MT-CO2  ENST00000361739  \n"
     ]
    }
   ],
   "source": [
    "get_seq(biomart2[biomart2.gene_biotype.isin(['protein_coding'])].ensembl_gene_id.tolist()[:4], translate=True, isoforms=False, save=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "python launch/predict.py --config=\"pretrained/extract_embedding.yml\" \\\n",
    "--data_path=\"./data/examples/example.fasta\" --save_dir=\"./resuts\" \\\n",
    "--save_frequency 1 --save_embeddings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>hgnc_symbol</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>entrezgene_id</th>\n",
       "      <th>ensembl_transcript_id</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ensembl_peptide_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000198888</td>\n",
       "      <td>MT-ND1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>4535.0</td>\n",
       "      <td>ENST00000361390</td>\n",
       "      <td>CAA24026</td>\n",
       "      <td>ENSP00000354687</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000198888</td>\n",
       "      <td>MT-ND1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>4535.0</td>\n",
       "      <td>ENST00000361390</td>\n",
       "      <td>AAB58943</td>\n",
       "      <td>ENSP00000354687</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000198888</td>\n",
       "      <td>MT-ND1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>4535.0</td>\n",
       "      <td>ENST00000361390</td>\n",
       "      <td>BAA07290</td>\n",
       "      <td>ENSP00000354687</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000198888</td>\n",
       "      <td>MT-ND1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>4535.0</td>\n",
       "      <td>ENST00000361390</td>\n",
       "      <td>AAP89036</td>\n",
       "      <td>ENSP00000354687</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000198888</td>\n",
       "      <td>MT-ND1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>4535.0</td>\n",
       "      <td>ENST00000361390</td>\n",
       "      <td>AAP89049</td>\n",
       "      <td>ENSP00000354687</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>901591</th>\n",
       "      <td>ENSG00000162437</td>\n",
       "      <td>RAVER2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>55225.0</td>\n",
       "      <td>ENST00000418058</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000397069</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>901592</th>\n",
       "      <td>ENSG00000122432</td>\n",
       "      <td>SPATA1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>100505741.0</td>\n",
       "      <td>ENST00000697276</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000514413</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>901593</th>\n",
       "      <td>ENSG00000122432</td>\n",
       "      <td>SPATA1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>100505741.0</td>\n",
       "      <td>ENST00000699524</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000514414</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>901594</th>\n",
       "      <td>ENSG00000122432</td>\n",
       "      <td>SPATA1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>100505741.0</td>\n",
       "      <td>ENST00000697277</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000514416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>901595</th>\n",
       "      <td>ENSG00000122432</td>\n",
       "      <td>SPATA1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>100505741.0</td>\n",
       "      <td>ENST00000460286</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000514417</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>901596 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        ensembl_gene_id hgnc_symbol    gene_biotype  entrezgene_id  \\\n",
       "0       ENSG00000198888      MT-ND1  protein_coding         4535.0   \n",
       "1       ENSG00000198888      MT-ND1  protein_coding         4535.0   \n",
       "2       ENSG00000198888      MT-ND1  protein_coding         4535.0   \n",
       "3       ENSG00000198888      MT-ND1  protein_coding         4535.0   \n",
       "4       ENSG00000198888      MT-ND1  protein_coding         4535.0   \n",
       "...                 ...         ...             ...            ...   \n",
       "901591  ENSG00000162437      RAVER2  protein_coding        55225.0   \n",
       "901592  ENSG00000122432      SPATA1  protein_coding    100505741.0   \n",
       "901593  ENSG00000122432      SPATA1  protein_coding    100505741.0   \n",
       "901594  ENSG00000122432      SPATA1  protein_coding    100505741.0   \n",
       "901595  ENSG00000122432      SPATA1  protein_coding    100505741.0   \n",
       "\n",
       "       ensembl_transcript_id protein_id ensembl_peptide_id  \n",
       "0            ENST00000361390   CAA24026    ENSP00000354687  \n",
       "1            ENST00000361390   AAB58943    ENSP00000354687  \n",
       "2            ENST00000361390   BAA07290    ENSP00000354687  \n",
       "3            ENST00000361390   AAP89036    ENSP00000354687  \n",
       "4            ENST00000361390   AAP89049    ENSP00000354687  \n",
       "...                      ...        ...                ...  \n",
       "901591       ENST00000418058        NaN    ENSP00000397069  \n",
       "901592       ENST00000697276        NaN    ENSP00000514413  \n",
       "901593       ENST00000699524        NaN    ENSP00000514414  \n",
       "901594       ENST00000697277        NaN    ENSP00000514416  \n",
       "901595       ENST00000460286        NaN    ENSP00000514417  \n",
       "\n",
       "[901596 rows x 7 columns]"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "biomart"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{33: tensor([-0.0224,  0.0423, -0.0291,  ..., -0.1396, -0.0235, -0.0449])}"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "import glob\n",
    "import torch\n",
    "\n",
    "tensor_list = []\n",
    "for file in glob.glob(\"../../data/temp/*.pt\"):\n",
    "    tensor = torch.load(file)['mean_representations'][33]\n",
    "    tensor_list.append(tensor)\n",
    "\n",
    "concatenated_tensor = torch.cat(tensor_list, dim=0)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get all genes in the dataset, all species\n",
    "# load them from biomart (for each species)\n",
    "# for all protein codings\n",
    "    # get the fasta file from uniprot using gget seq\n",
    "    # if not available use gget seq from ensembl\n",
    "        # use https://github.com/prestevez/dna2proteins to convert to protein\n",
    "        # merge with the uniprot fasta\n",
    "    # get the embedding of the fasta file using esm\n",
    "    \n",
    "\n",
    "# for all non protein codings (RNA based)\n",
    "    # get the fasta file from ensembl using gget seq\n",
    "    # use https://github.com/ml4bio/RNA-FM to embed the sequence\n",
    "\n",
    "# can we have learned embeddings for just a subset of the elements of the transformer?\n",
    "# can we have two KQV matrices, one for the protein coding, one for the RNA based? -> we would need at least to not have a skip connection for this first layer\n",
    "# else we have an additional FCN layer that maps both to the actual embedding size\n",
    "# adds \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_structural_embeddings(biotype, ids):\n",
    "\n",
    "    if biotype == \"proteins\":\n",
    "\n",
    "    if biotype == \"protein_coding_genes\":\n",
    "\n",
    "    if biotype == \"non_coding_genes\":\n",
    "\n",
    "    if biotype == \"coding_transcripts\":\n",
    "\n",
    "    if biotype == \"non_coding_transcripts\":\n",
    "        # get fasta file\n",
    "        # subset fasta file\n",
    "        # "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "store it as an additional array in the dataset (make it behave as a varm)\n",
    "\n",
    "add anndata idioms in the dataset\n",
    "\n",
    "add GRN as a sparse array in the data (make it behave as a varp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "https://ftp.ensembl.org/pub/release-110/fasta/mus_musculus/pep/Mus_musculus.GRCm39.pep.all.fa.gz\n",
    "\n",
    "https://ftp.ensembl.org/pub/release-110/fasta/homo_sapiens/ncrna/Homo_sapiens.GRCh38.ncrna.fa.gz\n",
    "\n",
    "https://ftp.ensembl.org/pub/release-110/fasta/mus_musculus/ncrna/Mus_musculus.GRCm39.ncrna.fa.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'221 Goodbye.'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import ftplib\n",
    "import os\n",
    "\n",
    "def list_files(ftp, match=''):\n",
    "    files = ftp.nlst()\n",
    "    return [file for file in files if file.endswith(match)]\n",
    "\n",
    "def load_fasta_species(species=\"homo_sapiens\")\n",
    "    ftp = ftplib.FTP('ftp.ensembl.org')\n",
    "    ftp.login()\n",
    "    ftp.cwd('/pub/release-110/fasta/homo_sapiens/pep/')\n",
    "    file = list_files(ftp, '.all.fa.gz')[0]\n",
    "    local_file_path = '../../data/fasta/' + file\n",
    "    if not os.path.exists(local_file_path):\n",
    "        with open(local_file_path, 'wb') as local_file:\n",
    "            ftp.retrbinary('RETR ' + file, local_file.write)\n",
    "    ftp.cwd('/pub/release-110/fasta/homo_sapiens/ncrna/')\n",
    "    file = list_files(ftp, '.ncrna.fa.gz')[0]\n",
    "    local_file_path = '../../data/fasta/' + file\n",
    "    if not os.path.exists(local_file_path):\n",
    "        with open(local_file_path, 'wb') as local_file:\n",
    "            ftp.retrbinary('RETR ' + file, local_file.write)\n",
    "    ftp.quit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "seqs, label, test_dl  = data.load_data_EMB(args.data_embedding) \n",
    "features = train.make_feature(model, test_dl, seqs)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "training-gpt",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
