<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="description" content="None" /><link rel="canonical" href="https://www.jkobject.com/scPRINT/" />
      <link rel="shortcut icon" href="img/favicon.ico" />
    <title>scprint</title>
    <link rel="stylesheet" href="css/theme.css" />
    <link rel="stylesheet" href="css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" />
        <link href="assets/_mkdocstrings.css" rel="stylesheet" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "Home";
        var mkdocs_page_input_path = "index.md";
        var mkdocs_page_url = "/scPRINT/";
      </script>
    
    <!--[if lt IE 9]>
      <script src="js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
      <script>hljs.highlightAll();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href="." class="icon icon-home"> scprint
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="./search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
                <li class="toctree-l1 current"><a class="reference internal current" href="#">Home</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#install-scprint">Install scPRINT</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#laminai">lamin.ai</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#usage">Usage</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#scprints-basic-commands">scPRINT's basic commands</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#notes-on-gpucpu-usage-with-triton">Notes on GPU/CPU usage with triton</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#faq">FAQ</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#i-want-to-generate-gene-networks-from-scrnaseq-data">I want to generate gene networks from scRNAseq data:</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#i-want-to-generate-cell-embeddings-and-cell-label-predictions-from-scrnaseq-data">I want to generate cell embeddings and cell label predictions from scRNAseq data:</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#i-want-to-denoising-my-scrnaseq-dataset">I want to denoising my scRNAseq dataset:</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#i-want-to-generate-an-atlas-level-embedding">I want to generate an atlas-level embedding</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#i-need-to-generate-gene-tokens-using-pllms">I need to generate gene tokens using pLLMs</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#i-want-to-pre-train-scprint-from-scratch-on-my-own-data">I want to pre-train scPRINT from scratch on my own data</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#how-can-i-find-if-scprint-was-trained-on-my-data">how can I find if scPRINT was trained on my data?</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#can-i-use-scprint-on-other-organisms-rather-than-human">can I use scPRINT on other organisms rather than human?</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#how-long-does-scprint-takes-what-kind-of-resources-do-i-need-or-in-alternative-can-i-run-scprint-locally">how long does scPRINT takes? what kind of resources do I need? (or in alternative: can i run scPRINT locally?)</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#i-have-different-scrnaseq-batches-should-i-integrate-my-data-before-running-scprint">I have different scRNASeq batches. Should I integrate my data before running scPRINT?</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#where-to-find-the-gene-embeddings">where to find the gene embeddings?</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#documentation">Documentation</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#model-weights">Model Weights</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#development">Development</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#work-in-progress">Work in progress:</a>
    </li>
    </ul>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="structure/">structure</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="pretrain/">pre-training</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="usage/">usage example</a>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">example notebooks</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="notebooks/cancer_usecase/">scPRINT use case on BPH</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="notebooks/cancer_usecase_part2/">scPRINT use case on BPH (part 2, GN analysis)</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">documentation</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="model/">model</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="tasks/">tasks</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="cli/">cli</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="embedder/">embedders</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="utils/">utils</a>
                  </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href=".">scprint</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="." class="icon icon-home" aria-label="Docs"></a></li>
      <li class="breadcrumb-item active">Home</li>
    <li class="wy-breadcrumbs-aside">
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <h1 id="scprint-large-cell-model-for-scrnaseq-data">scPRINT: Large Cell Model for scRNAseq data</h1>
<p><a href="https://badge.fury.io/py/scprint"><img alt="PyPI version" src="https://badge.fury.io/py/scprint.svg" /></a>
<a href="https://scprint.readthedocs.io/en/latest/?badge=latest"><img alt="Documentation Status" src="https://readthedocs.org/projects/scprint/badge/?version=latest" /></a>
<a href="https://pepy.tech/project/scprint"><img alt="Downloads" src="https://pepy.tech/badge/scprint" /></a>
<a href="https://pepy.tech/project/scprint"><img alt="Downloads" src="https://pepy.tech/badge/scprint/month" /></a>
<a href="https://pepy.tech/project/scprint"><img alt="Downloads" src="https://pepy.tech/badge/scprint/week" /></a>
<a href="https://img.shields.io/github/issues/jkobject/scPRINT"><img alt="GitHub issues" src="https://img.shields.io/github/issues/jkobject/scPRINT" /></a>
<a href="https://github.com/psf/black"><img alt="Code style: black" src="https://img.shields.io/badge/code%20style-black-000000.svg" /></a>
<a href=""><img alt="DOI" src="https://zenodo.org/badge/391909874.svg" /></a></p>
<p><img alt="logo" src="logo.png" /></p>
<p>scPRINT is a large transformer model built for the inference of gene networks (connections between genes explaining the cell's expression profile) from scRNAseq data.</p>
<p>It uses novel encoding and decoding of the cell expression profile and new pre-training methodologies to learn a cell model.</p>
<p>scPRINT can be used to perform the following analyses:</p>
<ul>
<li><strong>expression denoising</strong>: increase the resolution of your scRNAseq data</li>
<li><strong>cell embedding</strong>: generate a low-dimensional representation of your dataset</li>
<li><strong>label prediction</strong>: predict the cell type, disease, sequencer, sex, and ethnicity of your cells</li>
<li><strong>gene network inference</strong>: generate a gene network from any cell or cell cluster in your scRNAseq dataset</li>
</ul>
<p><a href="https://www.biorxiv.org/content/10.1101/2024.07.29.605556v1">Read the paper!</a> if you would like to know more about scPRINT.</p>
<p><img alt="figure1" src="figure1.png" /></p>
<h2 id="install-scprint">Install <code>scPRINT</code></h2>
<p>For the moment scPRINT has been tested on MacOS and Linux (Ubuntu 20.04) with Python 3.10.</p>
<p>If you want to be using flashattention2, know that it only supports triton 2.0 MLIR's version and torch==2.0.0 for now.</p>
<pre><code class="language-python">conda create -n &quot;[whatever]&quot; python==3.10
#one of
pip install scprint # OR
pip install scprint[dev] # for the dev dependencies (building etc..) AND/OR [dev,flash]
pip install scprint[flash] &amp;&amp; pip install -e &quot;git+https:/
/github.com/triton-lang/triton.git@legacy-backend
#egg=triton&amp;subdirectory=python&quot; # to use flashattention2, you will need to install triton 2.0.0.dev20221202 specifically, working on removing this dependency # only if you have a compatible gpu (e.g. not available for apple GPUs for now, see https://github.com/triton-lang/triton?tab=readme-ov-file#compatibility)
</code></pre>
<p>We make use of some additional packages we developed alongside scPRint.
Please refer to their documentation for more information:</p>
<ul>
<li><a href="https://github.com/jkobject/scDataLoader">scDataLoader</a>: a dataloader for training large cell models.</li>
<li><a href="https://github.com/cantinilab/GRnnData">GRnnData</a>: a package to work with gene networks from single cell data.</li>
<li><a href="https://github.com/jkobject/benGRN">benGRN</a>: a package to benchmark gene network inference methods from single cell data.</li>
</ul>
<h3 id="laminai">lamin.ai</h3>
<p>⚠️ if you want to use the scDataloader's multi-dataset mode or if you want to preprocess datasets and other functions of the model, you will need to use lamin.ai.</p>
<p>In that case, connect with google or github to <a href="https://lamin.ai/login">lamin.ai</a>, then be sure to connect before running anything (or before starting a notebook): <code>lamin login &lt;email&gt; --key &lt;API-key&gt;</code>. Follow the instructions on <a href="https://docs.lamin.ai/guide">their website</a>.</p>
<h2 id="usage">Usage</h2>
<h3 id="scprints-basic-commands">scPRINT's basic commands</h3>
<p>This is the most minimal example of how scPRINT works:</p>
<pre><code class="language-py">from lightning.pytorch import Trainer
from scprint import scPrint
from scdataloader import DataModule

datamodule = DataModule(...)
model = scPrint(...)
# to train / fit / test the model
trainer = Trainer(...)
trainer.fit(model, datamodule=datamodule)
# to do predictions Denoiser, Embedder, GNInfer
denoiser = Denoiser(...)
adata = sc.read_h5ad(...)
denoiser(model, adata=adata)
...
</code></pre>
<p>or, from a bash command line</p>
<pre><code class="language-bash">$ scprint fit/train/predict/test/denoise/embed/gninfer --config config/[medium|large|vlarge] ...
</code></pre>
<p>find out more about the commands by running <code>scprint --help</code> or <code>scprint [command] --help</code>.</p>
<p>more examples of using the command line are available in the <a href="./docs/usage.md">docs</a>.</p>
<h3 id="notes-on-gpucpu-usage-with-triton">Notes on GPU/CPU usage with triton</h3>
<p>If you do not have <a href="https://triton-lang.org/main/python-api/triton.html">triton</a> installed you will not be able to take advantage of GPU acceleration, but you can still use the model on the CPU.</p>
<p>In that case, if loading from a checkpoint that was trained with flashattention, you will need to specify <code>transformer="normal"</code> in the <code>load_from_checkpoint</code> function like so:</p>
<pre><code class="language-python">model = scPrint.load_from_checkpoint(
    '../data/temp/last.ckpt', precpt_gene_emb=None,
    transformer=&quot;normal&quot;)
</code></pre>
<p>We now explore the different usages of scPRINT:</p>
<h2 id="faq">FAQ</h2>
<h3 id="i-want-to-generate-gene-networks-from-scrnaseq-data">I want to generate gene networks from scRNAseq data:</h3>
<p>-&gt; Refer to the section . gene network inference in <a href="notebooks/cancer_usecase/">this notebook</a>.</p>
<p>-&gt; More examples in this notebook <a href="../notebooks/bench_omni.ipynb">notebooks/assessments/bench_omni.ipynb</a>.</p>
<h3 id="i-want-to-generate-cell-embeddings-and-cell-label-predictions-from-scrnaseq-data">I want to generate cell embeddings and cell label predictions from scRNAseq data:</h3>
<p>-&gt; Refer to the embeddings and cell annotations section in <a href="notebooks/cancer_usecase/">this notebook</a>.</p>
<h3 id="i-want-to-denoising-my-scrnaseq-dataset">I want to denoising my scRNAseq dataset:</h3>
<p>-&gt; Refer to the Denoising of B-cell section in <a href="notebooks/cancer_usecase/">this notebook</a>.</p>
<p>-&gt; More example in our benchmark notebook <a href="../notebooks/bench_denoising.ipynb">notebooks/assessments/bench_denoising.ipynb</a>.</p>
<h3 id="i-want-to-generate-an-atlas-level-embedding">I want to generate an atlas-level embedding</h3>
<p>-&gt; Refer to the notebook <a href="../figures/nice_umap.ipynb">figures/nice_umap.ipynb</a>.</p>
<h3 id="i-need-to-generate-gene-tokens-using-pllms">I need to generate gene tokens using pLLMs</h3>
<p>To run scPRINT, you can use the option to define the gene tokens using protein language model embeddings of genes. This is done by providing the path to a parquet file of the precomputed set of embeddings for each gene name to scPRINT via "precpt_gene_emb"</p>
<p>-&gt; To generate this file please refer to the notebook <a href="../notebooks/generate_gene_embeddings.ipynb">notebooks/generate_gene_embeddings.ipynb</a>.</p>
<h3 id="i-want-to-pre-train-scprint-from-scratch-on-my-own-data">I want to pre-train scPRINT from scratch on my own data</h3>
<p>-&gt; Refer to the documentation page <a href="pretrain/">pretrain scprint</a></p>
<h3 id="how-can-i-find-if-scprint-was-trained-on-my-data">how can I find if scPRINT was trained on my data?</h3>
<p>If your data is available in cellxgene, scPRINT was likely trained on it. However some cells, datasets were dropped due to low quality data and some were randomly removed to be part of the validation / test sets.</p>
<h3 id="can-i-use-scprint-on-other-organisms-rather-than-human">can I use scPRINT on other organisms rather than human?</h3>
<p>scPRINT has been pretrained on both humans and mouse, and can be used on any organism with a similar gene set. If you want to use scPRINT on very different organisms, you will need to generate gene embeddings for that organism and re-train scPRINT</p>
<h3 id="how-long-does-scprint-takes-what-kind-of-resources-do-i-need-or-in-alternative-can-i-run-scprint-locally">how long does scPRINT takes? what kind of resources do I need? (or in alternative: can i run scPRINT locally?)</h3>
<p>please look at our supplementary tables in the <a href="https://www.biorxiv.org/content/10.1101/2024.07.29.605556v1">manuscript</a></p>
<h3 id="i-have-different-scrnaseq-batches-should-i-integrate-my-data-before-running-scprint">I have different scRNASeq batches. Should I integrate my data before running scPRINT?</h3>
<p>scPRINT takes raw count as inputs, so please don't use integrated data. Just give the raw counts to scPRINT and it will take care of the rest.</p>
<h3 id="where-to-find-the-gene-embeddings">where to find the gene embeddings?</h3>
<p>If you think you need the gene embeddings file for loading the model from a checkpoint, you don't, as the embeddings are also stored in the model weights. You just need to load the weights like this:</p>
<pre><code class="language-python">model = scPrint.load_from_checkpoint(
    '../../data/temp/last.ckpt',
    precpt_gene_emb=None,
)
</code></pre>
<p>You can also recreate the gene embedding file through <a href="notebooks/generate_gene_embeddings.ipynb">this notebook</a>. Just call the functions, and it should recreate the file itself.</p>
<p>the file itself is also available on <a href="https://huggingface.co/jkobject/scPRINT/tree/main">hugging face</a></p>
<h2 id="documentation">Documentation</h2>
<p>For more information on usage please see the documentation in <a href="https://www.jkobject.com/scPRINT/">https://www.jkobject.com/scPRINT/</a></p>
<h2 id="model-weights">Model Weights</h2>
<p>Model weights are available on <a href="https://huggingface.co/jkobject/scPRINT/">hugging face</a>.</p>
<h2 id="development">Development</h2>
<p>Read the <a href="CONTRIBUTING.md">CONTRIBUTING.md</a> file.</p>
<p>Read the <a href="https://wandb.ai/ml4ig/scprint_scale/reports/scPRINT-trainings--Vmlldzo4ODIxMjgx?accessToken=80metwx7b08hhourotpskdyaxiflq700xzmzymr6scvkp69agybt79l341tv68hp">training runs</a> document to know more about how pre-training was performed and the its behavior.</p>
<p>Acknowledgement:
<a href="https://github.com/rochacbruno/python-project-template">python template</a>
<a href="https://lamin.ai/">laminDB</a>
<a href="https://lightning.ai/">lightning</a></p>
<h2 id="work-in-progress">Work in progress:</h2>
<ol>
<li>remove the triton dependencies</li>
<li>add version with additional labels (tissues, age) and organisms (mouse, zebrafish) and more datasets from cellxgene</li>
<li>version with separate transformer blocks for the encoding part of the bottleneck learning and for the cell embeddings</li>
<li>improve classifier to output uncertainties and topK predictions when unsure</li>
<li></li>
</ol>
<p>Awesome Large Cell Model created by Jeremie Kalfon.</p>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="structure/" class="btn btn-neutral float-right" title="structure">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
    
    
      <span><a href="structure/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script src="js/jquery-3.6.0.min.js"></script>
    <script>var base_url = ".";</script>
    <script src="js/theme_extra.js"></script>
    <script src="js/theme.js"></script>
      <script src="search/main.js"></script>
    <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>

</body>
</html>

<!--
MkDocs version : 1.6.0
Build Date UTC : 2024-09-04 16:55:30.146389+00:00
-->
