{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "14401148",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-12-07T10:33:15.327491Z",
     "iopub.status.busy": "2025-12-07T10:33:15.327368Z",
     "iopub.status.idle": "2025-12-07T10:33:34.179121Z",
     "shell.execute_reply": "2025-12-07T10:33:34.178626Z"
    },
    "papermill": {
     "duration": 18.860713,
     "end_time": "2025-12-07T10:33:34.180435",
     "exception": false,
     "start_time": "2025-12-07T10:33:15.319722",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m→\u001b[0m connected lamindb: jkobject/scprint2\n"
     ]
    }
   ],
   "source": [
    "from scdataloader.utils import translate\n",
    "import scanpy as sc\n",
    "from anndata.experimental import concat_on_disk\n",
    "from umap import UMAP\n",
    "import numpy as np\n",
    "\n",
    "import bionty as bt\n",
    "from scdataloader.utils import get_all_ancestors\n",
    "\n",
    "import matplotlib\n",
    "\n",
    "import datamapplot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2fd3869b",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-12-07T10:33:34.185405Z",
     "iopub.status.busy": "2025-12-07T10:33:34.184971Z",
     "iopub.status.idle": "2025-12-07T10:33:34.187350Z",
     "shell.execute_reply": "2025-12-07T10:33:34.187089Z"
    },
    "papermill": {
     "duration": 0.004556,
     "end_time": "2025-12-07T10:33:34.187721",
     "exception": false,
     "start_time": "2025-12-07T10:33:34.183165",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "data_directory = \"/pasteur/appa/scratch/jkalfon/45322258/\"\n",
    "name = \"18hebyht\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "301d7458",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-12-07T10:33:34.190994Z",
     "iopub.status.busy": "2025-12-07T10:33:34.190911Z",
     "iopub.status.idle": "2025-12-07T10:33:45.937794Z",
     "shell.execute_reply": "2025-12-07T10:33:45.937431Z"
    },
    "papermill": {
     "duration": 11.749242,
     "end_time": "2025-12-07T10:33:45.938498",
     "exception": false,
     "start_time": "2025-12-07T10:33:34.189256",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 16406560 × 0\n",
       "    obs: 'pred_cell_type_ontology_term_id', 'pred_tissue_ontology_term_id', 'pred_disease_ontology_term_id', 'pred_age_group', 'pred_assay_ontology_term_id', 'pred_self_reported_ethnicity_ontology_term_id', 'pred_sex_ontology_term_id', 'pred_organism_ontology_term_id', 'pred_cell_culture', 'conv_pred_cell_type_ontology_term_id', 'conv_pred_tissue_ontology_term_id', 'conv_pred_disease_ontology_term_id', 'conv_pred_age_group', 'conv_pred_assay_ontology_term_id', 'conv_pred_self_reported_ethnicity_ontology_term_id', 'leiden_1.0'\n",
       "    uns: 'leiden_1.0', 'neighbors', 'umap'\n",
       "    obsm: 'X_umap', 'scprint_emb_cell_type_ontology_term_id'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata = sc.read_h5ad(data_directory + name + \"_predict.h5ad\")  # , backed=\"r\")\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3852dc4",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-12-07T10:33:45.943378Z",
     "iopub.status.busy": "2025-12-07T10:33:45.943288Z",
     "iopub.status.idle": "2025-12-07T10:33:57.081120Z",
     "shell.execute_reply": "2025-12-07T10:33:57.080707Z"
    },
    "papermill": {
     "duration": 11.140557,
     "end_time": "2025-12-07T10:33:57.081744",
     "exception": false,
     "start_time": "2025-12-07T10:33:45.941187",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pred_sex_ontology_term_id\n",
      "2\n",
      "2\n",
      "pred_cell_culture\n",
      "2\n",
      "2\n",
      "pred_organism_ontology_term_id\n",
      "16\n",
      "16\n",
      "conv_pred_cell_type_ontology_term_id\n",
      "337\n",
      "254\n",
      "conv_pred_self_reported_ethnicity_ontology_term_id\n",
      "16\n",
      "16\n"
     ]
    }
   ],
   "source": [
    "res = {}\n",
    "subres = {}\n",
    "for i in [\n",
    "    \"pred_sex_ontology_term_id\",\n",
    "    \"pred_cell_culture\",\n",
    "    \"pred_organism_ontology_term_id\",\n",
    "    \"conv_pred_cell_type_ontology_term_id\",\n",
    "    \"conv_pred_self_reported_ethnicity_ontology_term_id\",\n",
    "]:\n",
    "    print(i)\n",
    "    res[i] = adata.obs[i].value_counts()\n",
    "    print(len(res[i]))\n",
    "    subres[i] = {j: j if k > 150 else \"other\" for j, k in res[i].items()}\n",
    "    print(len(set(subres[i].values())))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "246d19e5",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-12-07T10:33:57.086891Z",
     "iopub.status.busy": "2025-12-07T10:33:57.086800Z",
     "iopub.status.idle": "2025-12-07T10:33:57.834168Z",
     "shell.execute_reply": "2025-12-07T10:33:57.833754Z"
    },
    "papermill": {
     "duration": 0.750386,
     "end_time": "2025-12-07T10:33:57.835076",
     "exception": false,
     "start_time": "2025-12-07T10:33:57.084690",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "rt = translate(\n",
    "    res[\"pred_organism_ontology_term_id\"].keys(), \"organism_ontology_term_id\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "9490a82c",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-12-07T10:33:57.840011Z",
     "iopub.status.busy": "2025-12-07T10:33:57.839905Z",
     "iopub.status.idle": "2025-12-07T10:33:57.842133Z",
     "shell.execute_reply": "2025-12-07T10:33:57.841856Z"
    },
    "papermill": {
     "duration": 0.004672,
     "end_time": "2025-12-07T10:33:57.842506",
     "exception": false,
     "start_time": "2025-12-07T10:33:57.837834",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "rt = {j: rt[i] for i, j in enumerate(res[\"pred_organism_ontology_term_id\"].keys())}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "b70d8e39",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-12-07T10:33:57.846266Z",
     "iopub.status.busy": "2025-12-07T10:33:57.846180Z",
     "iopub.status.idle": "2025-12-07T10:33:57.847984Z",
     "shell.execute_reply": "2025-12-07T10:33:57.847734Z"
    },
    "papermill": {
     "duration": 0.004057,
     "end_time": "2025-12-07T10:33:57.848331",
     "exception": false,
     "start_time": "2025-12-07T10:33:57.844274",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "rt[\"NCBITaxon:9483\"] = \"marmoset\"\n",
    "rt[\"NCBITaxon:7227\"] = \"drosophila\"\n",
    "rt[\"NCBITaxon:3702\"] = \"arabidopsis\"\n",
    "rt[\"NCBITaxon:4577\"] = \"maize\"\n",
    "rt[\"NCBITaxon:6239\"] = \"c. elegans\"\n",
    "rt[\"NCBITaxon:9940\"] = \"sheep\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "167f64a3",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-12-07T10:33:57.852027Z",
     "iopub.status.busy": "2025-12-07T10:33:57.851945Z",
     "iopub.status.idle": "2025-12-07T10:33:57.857791Z",
     "shell.execute_reply": "2025-12-07T10:33:57.857369Z"
    },
    "papermill": {
     "duration": 0.008049,
     "end_time": "2025-12-07T10:33:57.858210",
     "exception": false,
     "start_time": "2025-12-07T10:33:57.850161",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/local/scratch/tmp/ipykernel_3851655/1575082616.py:3: FutureWarning: The behavior of Series.replace (and DataFrame.replace) with CategoricalDtype is deprecated. In a future version, replace will only be used for cases that preserve the categories. To change the categories, use ser.cat.rename_categories instead.\n",
      "  ].replace(rt)\n"
     ]
    }
   ],
   "source": [
    "adata.obs[\"conv_pred_organism_ontology_term_id\"] = adata.obs[\n",
    "    \"pred_organism_ontology_term_id\"\n",
    "].replace(rt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c34eb48a",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.color_palette(\"husl\", 8)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "962b346f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "conv_pred_cell_type_ontology_term_id\n",
      "337\n",
      "183\n"
     ]
    }
   ],
   "source": [
    "subsubres = {}\n",
    "for i in [\n",
    "    \"conv_pred_cell_type_ontology_term_id\",\n",
    "]:\n",
    "    print(i)\n",
    "    res[i] = adata.obs[i].value_counts()\n",
    "    print(len(res[i]))\n",
    "    subsubres[i] = {j: j if k > 1500 else \"other\" for j, k in res[i].items()}\n",
    "    print(len(set(subsubres[i].values())))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "558c1ac1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[\"10x 3' v2\",\n",
       " \"10x 3' v3\",\n",
       " \"10x 5' v1\",\n",
       " '10x immune profiling',\n",
       " '10x multiome',\n",
       " 'sci-Plex',\n",
       " \"10x 5' v2\",\n",
       " 'Seq-Well S3',\n",
       " 'Drop-seq',\n",
       " 'SPLiT-seq',\n",
       " '10x feature profiling',\n",
       " 'ScaleBio single cell RNA sequencing',\n",
       " 'microwell-seq',\n",
       " 'TruDrop',\n",
       " 'Smart-seq v4',\n",
       " 'inDrop',\n",
       " \"10x 3' v1\",\n",
       " 'MARS-seq',\n",
       " 'Smart-seq2',\n",
       " 'BD Rhapsody Whole Transcriptome Analysis',\n",
       " 'GEXSCOPE technology',\n",
       " 'STRT-seq',\n",
       " 'CEL-seq2',\n",
       " 'DroNc-seq']"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata[: MAX * 2].obs[\"conv_pred_assay_ontology_term_id\"].unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "acbf1cb2",
   "metadata": {},
   "outputs": [],
   "source": [
    "subres[\"conv_pred_assay_ontology_term_id\"] = {\n",
    "    \"10x 3' v2\": \"10x 3'\",\n",
    "    \"10x 3' v3\": \"10x 3'\",\n",
    "    \"10x 3' v1\": \"10x 3'\",\n",
    "    \"10x immune profiling\": \"10x 3'\",\n",
    "    \"10x multiome\": \"10x 3'\",\n",
    "    \"Smart-seq v4\": \"Smart-seq\",\n",
    "    \"Smart-seq2\": \"Smart-seq\",\n",
    "    \"10x 5' v2\": \"10x 5'\",\n",
    "    \"10x 5' v1\": \"10x 5'\",\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "75339cf3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['kidney proximal straight tubule epithelial cell', 'MHC-II-positive classical monocyte', 'cerebellar granule cell', 'naive regulatory T cell', 'effector memory CD8-positive, alpha-beta T cell', ..., 'basal cell of epithelium of bronchus', 'type EC enteroendocrine cell', 'enterocyte of colon', 'IgG plasma cell', 'CD14-low, CD16-positive monocyte']\n",
       "Length: 500000\n",
       "Categories (295, object): ['Bergmann glial cell', 'CD1c-positive myeloid dendritic cell', 'CD4-positive, alpha-beta cytotoxic T cell', 'CD4-positive, alpha-beta thymocyte', ..., 'vascular leptomeningeal cell (Mmus)', 'vascular lymphangioblast', 'ventricular cardiac muscle cell', 'vip GABAergic cortical interneuron']"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata[:MAX].obs[\"conv_pred_cell_type_ontology_term_id\"].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "b257f870",
   "metadata": {},
   "outputs": [],
   "source": [
    "MAX = 500_000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "300adc21",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-12-07T10:33:57.867743Z",
     "iopub.status.busy": "2025-12-07T10:33:57.867669Z",
     "iopub.status.idle": "2025-12-07T10:34:13.340027Z",
     "shell.execute_reply": "2025-12-07T10:34:13.339475Z"
    },
    "papermill": {
     "duration": 15.474774,
     "end_time": "2025-12-07T10:34:13.340543",
     "exception": true,
     "start_time": "2025-12-07T10:33:57.865769",
     "status": "failed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/local/scratch/tmp/ipykernel_3851655/3474557632.py:6: FutureWarning: The behavior of Series.replace (and DataFrame.replace) with CategoricalDtype is deprecated. In a future version, replace will only be used for cases that preserve the categories. To change the categories, use ser.cat.rename_categories instead.\n",
      "  .replace(subsubres[\"conv_pred_cell_type_ontology_term_id\"]),\n",
      "/local/scratch/tmp/ipykernel_3851655/3474557632.py:11: FutureWarning: The behavior of Series.replace (and DataFrame.replace) with CategoricalDtype is deprecated. In a future version, replace will only be used for cases that preserve the categories. To change the categories, use ser.cat.rename_categories instead.\n",
      "  .replace(subres[\"conv_pred_cell_type_ontology_term_id\"])\n",
      "/local/scratch/tmp/ipykernel_3851655/3474557632.py:16: FutureWarning: The behavior of Series.replace (and DataFrame.replace) with CategoricalDtype is deprecated. In a future version, replace will only be used for cases that preserve the categories. To change the categories, use ser.cat.rename_categories instead.\n",
      "  .replace(subres[\"conv_pred_assay_ontology_term_id\"])\n",
      "/pasteur/appa/homes/jkalfon/scPRINT/.venv/lib/python3.11/site-packages/sklearn/base.py:1389: ConvergenceWarning: Number of distinct clusters (14) found smaller than n_clusters (337). Possibly due to duplicate points in X.\n",
      "  return fit_method(estimator, *args, **kwargs)\n"
     ]
    }
   ],
   "source": [
    "plot = datamapplot.create_interactive_plot(\n",
    "    adata[:MAX].obsm[\"X_umap\"],\n",
    "    adata[:MAX].obs[\"conv_pred_organism_ontology_term_id\"],\n",
    "    adata[:MAX]\n",
    "    .obs[\"conv_pred_cell_type_ontology_term_id\"]\n",
    "    .replace(subsubres[\"conv_pred_cell_type_ontology_term_id\"]),\n",
    "    noise_label=\"other\",\n",
    "    colormap_rawdata=[\n",
    "        adata[:MAX]\n",
    "        .obs[\"conv_pred_cell_type_ontology_term_id\"]\n",
    "        .replace(subres[\"conv_pred_cell_type_ontology_term_id\"])\n",
    "        .values,\n",
    "        adata[:MAX].obs[\"conv_pred_organism_ontology_term_id\"].values,\n",
    "        adata[:MAX]\n",
    "        .obs[\"conv_pred_assay_ontology_term_id\"]\n",
    "        .replace(subres[\"conv_pred_assay_ontology_term_id\"])\n",
    "        .values,\n",
    "        adata[:MAX].obs[\"pred_cell_culture\"].values,\n",
    "    ],\n",
    "    colormap_metadata=[\n",
    "        {\n",
    "            \"field\": \"cell type\",\n",
    "            \"description\": \"Predicted Cell Type\",\n",
    "            \"cmap\": \"gist_rainbow\",\n",
    "            \"kind\": \"categorical\",\n",
    "            \"n_colors\": len(subres[\"conv_pred_cell_type_ontology_term_id\"]),\n",
    "            # palete\n",
    "        },\n",
    "        # organism\n",
    "        {\n",
    "            \"field\": \"organism\",\n",
    "            \"description\": \"Predicted Organism\",\n",
    "            \"cmap\": \"tab20\",\n",
    "            \"kind\": \"categorical\",\n",
    "            \"n_colors\": len(subres[\"pred_organism_ontology_term_id\"]),\n",
    "        },\n",
    "        # assay\n",
    "        {\n",
    "            \"field\": \"assay\",\n",
    "            \"description\": \"Predicted Assay\",\n",
    "            \"cmap\": \"tab20\",\n",
    "            \"kind\": \"categorical\",\n",
    "            \"n_colors\": len(subres[\"conv_pred_assay_ontology_term_id\"]),\n",
    "        },\n",
    "        # cell culture\n",
    "        {\n",
    "            \"field\": \"culture\",\n",
    "            \"description\": \"Predicted Cellular Culture\",\n",
    "            \"cmap\": \"Set3\",\n",
    "            \"kind\": \"categorical\",\n",
    "            \"n_colors\": len(subres[\"pred_cell_culture\"]),\n",
    "        },\n",
    "    ],\n",
    "    hover_text=(\n",
    "        \"organism: \"\n",
    "        + adata[:MAX].obs[\"conv_pred_organism_ontology_term_id\"].astype(str)\n",
    "        + \"\\ncell type: \"\n",
    "        + adata[:MAX].obs[\"conv_pred_cell_type_ontology_term_id\"].astype(str)\n",
    "        + \"\\ncellular culture: \"\n",
    "        + adata[:MAX].obs[\"pred_cell_culture\"].astype(str)\n",
    "        + \"\\nassay: \"\n",
    "        + adata[:MAX].obs[\"conv_pred_assay_ontology_term_id\"].astype(str)\n",
    "        + \"\\ntissue of origin: \"\n",
    "        + adata[:MAX].obs[\"conv_pred_tissue_ontology_term_id\"].astype(str)\n",
    "        + \"\\ndisease of origin: \"\n",
    "        + adata[:MAX].obs[\"conv_pred_disease_ontology_term_id\"].astype(str)\n",
    "        + \"\\nage: \"\n",
    "        + adata[:MAX].obs[\"conv_pred_age_group\"].astype(str)\n",
    "        + \"\\nsex: \"\n",
    "        + adata[:MAX].obs[\"pred_sex_ontology_term_id\"].astype(str)\n",
    "        + \"\\n ethnicity: \"\n",
    "        + adata[:MAX]\n",
    "        .obs[\"conv_pred_self_reported_ethnicity_ontology_term_id\"]\n",
    "        .astype(str)\n",
    "    ),\n",
    "    font_family=\"DejaVu Sans\",\n",
    "    #cmap=\"viridis\",\n",
    "    enable_search=True,\n",
    "    inline_data=False,\n",
    "    # use_medoids=True,\n",
    "    initial_zoom_fraction=0.5,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5de4251a",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-12-05T23:14:42.042877Z",
     "iopub.status.busy": "2025-12-05T23:14:42.042757Z",
     "iopub.status.idle": "2025-12-05T23:14:42.046311Z",
     "shell.execute_reply": "2025-12-05T23:14:42.046102Z"
    },
    "papermill": {
     "duration": null,
     "end_time": null,
     "exception": null,
     "start_time": null,
     "status": "pending"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "plot.save(\"./nice_umap_scprint3.html\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scPRINT",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  },
  "papermill": {
   "default_parameters": {},
   "duration": 62.746261,
   "end_time": "2025-12-07T10:34:16.219836",
   "environment_variables": {},
   "exception": true,
   "input_path": "figures/nice_umap_inter.ipynb",
   "output_path": "figures/nice_umap_inter.ipynb",
   "parameters": {},
   "start_time": "2025-12-07T10:33:13.473575",
   "version": "2.6.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
