<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><link rel="canonical" href="https://www.jkobject.com/scPRINT/utils/" />
      <link rel="shortcut icon" href="../img/favicon.ico" />
    <title>utils - scprint</title>
    <link rel="stylesheet" href="../css/theme.css" />
    <link rel="stylesheet" href="../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" />
        <link href="../assets/_mkdocstrings.css" rel="stylesheet" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "utils";
        var mkdocs_page_input_path = "utils.md";
        var mkdocs_page_url = "/scPRINT/utils/";
      </script>
    
    <!--[if lt IE 9]>
      <script src="../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
      <script>hljs.highlightAll();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href=".." class="icon icon-home"> scprint
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="..">Home</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../structure/">structure</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../pretrain/">pre-training</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../usage/">usage example</a>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">example notebooks</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../notebooks/cancer_usecase/">Cancer usecase</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../notebooks/cancer_usecase_part2/">Cancer usecase (part 2)</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">documentation</span></p>
              <ul class="current">
                  <li class="toctree-l1"><a class="reference internal" href="../model/">model</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../tasks/">tasks</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../cli/">cli</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../embedder/">embedders</a>
                  </li>
                  <li class="toctree-l1 current"><a class="reference internal current" href="#">utils</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#scprint.utils.sinkhorn">sinkhorn</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#scprint.utils.sinkhorn.SinkhornDistance">SinkhornDistance</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#scprint.utils.sinkhorn.SinkhornDistance.M">M</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#scprint.utils.sinkhorn.SinkhornDistance.ave">ave</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#scprint.utils.sinkhorn.SinkhornDistance.forward">forward</a>
    </li>
        </ul>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#scprint.utils.utils">utils</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#scprint.utils.utils.category_str2int">category_str2int</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#scprint.utils.utils.createFoldersFor">createFoldersFor</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#scprint.utils.utils.fileToList">fileToList</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#scprint.utils.utils.find_required_colums">find_required_colums</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#scprint.utils.utils.getBiomartTable">getBiomartTable</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#scprint.utils.utils.get_free_gpu">get_free_gpu</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#scprint.utils.utils.get_git_commit">get_git_commit</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#scprint.utils.utils.histogram">histogram</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#scprint.utils.utils.inf_loop">inf_loop</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#scprint.utils.utils.isnotebook">isnotebook</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#scprint.utils.utils.listToFile">listToFile</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#scprint.utils.utils.load_genes">load_genes</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#scprint.utils.utils.map_raw_id_to_vocab_id">map_raw_id_to_vocab_id</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#scprint.utils.utils.onto_to_name">onto_to_name</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#scprint.utils.utils.pd_load_cached">pd_load_cached</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#scprint.utils.utils.prepare_device">prepare_device</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#scprint.utils.utils.run_command">run_command</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#scprint.utils.utils.set_seed">set_seed</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#scprint.utils.get_seq">get_seq</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#scprint.utils.get_seq.load_fasta_species">load_fasta_species</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#scprint.utils.get_seq.seq">seq</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#scprint.utils.get_seq.subset_fasta">subset_fasta</a>
    </li>
        </ul>
    </li>
    </ul>
                  </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="..">scprint</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href=".." class="icon icon-home" aria-label="Docs"></a></li>
          <li class="breadcrumb-item">documentation</li>
      <li class="breadcrumb-item active">utils</li>
    <li class="wy-breadcrumbs-aside">
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <h1 id="documentation-for-the-utils-modules">Documentation for the <code>utils</code> modules</h1>


<div class="doc doc-object doc-module">



<h2 id="scprint.utils.sinkhorn" class="doc doc-heading">
            <code>scprint.utils.sinkhorn</code>


</h2>

    <div class="doc doc-contents first">



  <div class="doc doc-children">








<div class="doc doc-object doc-class">



<h3 id="scprint.utils.sinkhorn.SinkhornDistance" class="doc doc-heading">
              <code class="highlight language-python"><span class="n">SinkhornDistance</span></code>

</h3>


    <div class="doc doc-contents ">
            <p class="doc doc-class-bases">
              Bases: <code><span title="torch.nn.Module">Module</span></code></p>


      <p>SinkhornDistance Initialize the SinkhornDistance class</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>eps</code></b>
                  (<code>float</code>, default:
                      <code>0.01</code>
)
              –
              <div class="doc-md-description">
                <p>Regularization parameter. Defaults to 1e-2.</p>
              </div>
            </li>
            <li>
              <b><code>max_iter</code></b>
                  (<code>int</code>, default:
                      <code>100</code>
)
              –
              <div class="doc-md-description">
                <p>Maximum number of Sinkhorn iterations. Defaults to 100.</p>
              </div>
            </li>
            <li>
              <b><code>reduction</code></b>
                  (<code>str</code>, default:
                      <code>&#39;none&#39;</code>
)
              –
              <div class="doc-md-description">
                <p>Specifies the reduction to apply to the output. Defaults to "none".</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
                  <details class="quote">
                    <summary>Source code in <code>scprint/utils/sinkhorn.py</code></summary>
                    <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">eps</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">1e-2</span><span class="p">,</span> <span class="n">max_iter</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">100</span><span class="p">,</span> <span class="n">reduction</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;none&quot;</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    SinkhornDistance Initialize the SinkhornDistance class</span>

<span class="sd">    Args:</span>
<span class="sd">        eps (float, optional): Regularization parameter. Defaults to 1e-2.</span>
<span class="sd">        max_iter (int, optional): Maximum number of Sinkhorn iterations. Defaults to 100.</span>
<span class="sd">        reduction (str, optional): Specifies the reduction to apply to the output. Defaults to &quot;none&quot;.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="nb">super</span><span class="p">(</span><span class="n">SinkhornDistance</span><span class="p">,</span> <span class="bp">self</span><span class="p">)</span><span class="o">.</span><span class="fm">__init__</span><span class="p">()</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">eps</span> <span class="o">=</span> <span class="n">eps</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">max_iter</span> <span class="o">=</span> <span class="n">max_iter</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">reduction</span> <span class="o">=</span> <span class="n">reduction</span>
</code></pre></div></td></tr></table></div>
                  </details>



  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h4 id="scprint.utils.sinkhorn.SinkhornDistance.M" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">M</span></code>

</h4>


    <div class="doc doc-contents ">

      <p>Modified cost for logarithmic updates</p>

            <details class="quote">
              <summary>Source code in <code>scprint/utils/sinkhorn.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">M</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">C</span><span class="p">,</span> <span class="n">u</span><span class="p">,</span> <span class="n">v</span><span class="p">):</span>
    <span class="s2">&quot;Modified cost for logarithmic updates&quot;</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;$M_{ij} = (-c_{ij} + u_i + v_j) / epsilon$&quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="p">(</span><span class="o">-</span><span class="n">C</span> <span class="o">+</span> <span class="n">u</span><span class="o">.</span><span class="n">unsqueeze</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">)</span> <span class="o">+</span> <span class="n">v</span><span class="o">.</span><span class="n">unsqueeze</span><span class="p">(</span><span class="mi">1</span><span class="p">))</span> <span class="o">/</span> <span class="bp">self</span><span class="o">.</span><span class="n">eps</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="scprint.utils.sinkhorn.SinkhornDistance.ave" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">ave</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-staticmethod"><code>staticmethod</code></small>
  </span>

</h4>


    <div class="doc doc-contents ">

      <p>Barycenter subroutine, used by kinetic acceleration through extrapolation.</p>

            <details class="quote">
              <summary>Source code in <code>scprint/utils/sinkhorn.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@staticmethod</span>
<span class="k">def</span> <span class="nf">ave</span><span class="p">(</span><span class="n">u</span><span class="p">,</span> <span class="n">u1</span><span class="p">,</span> <span class="n">tau</span><span class="p">):</span>
    <span class="s2">&quot;Barycenter subroutine, used by kinetic acceleration through extrapolation.&quot;</span>
    <span class="k">return</span> <span class="n">tau</span> <span class="o">*</span> <span class="n">u</span> <span class="o">+</span> <span class="p">(</span><span class="mi">1</span> <span class="o">-</span> <span class="n">tau</span><span class="p">)</span> <span class="o">*</span> <span class="n">u1</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="scprint.utils.sinkhorn.SinkhornDistance.forward" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">forward</span></code>

</h4>


    <div class="doc doc-contents ">

      <p>forward Compute the Sinkhorn distance between two measures with cost matrix c</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>c</code></b>
                  (<code><span title="torch.Tensor">Tensor</span></code>)
              –
              <div class="doc-md-description">
                <p>The cost matrix between the two measures.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              –
              <div class="doc-md-description">
                <p>torch.Tensor: The computed Sinkhorn distance.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
            <details class="quote">
              <summary>Source code in <code>scprint/utils/sinkhorn.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span>
<span class="normal">86</span>
<span class="normal">87</span>
<span class="normal">88</span>
<span class="normal">89</span>
<span class="normal">90</span>
<span class="normal">91</span>
<span class="normal">92</span>
<span class="normal">93</span>
<span class="normal">94</span>
<span class="normal">95</span>
<span class="normal">96</span>
<span class="normal">97</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">forward</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">c</span><span class="p">:</span> <span class="n">torch</span><span class="o">.</span><span class="n">Tensor</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    forward Compute the Sinkhorn distance between two measures with cost matrix c</span>

<span class="sd">    Args:</span>
<span class="sd">        c (torch.Tensor): The cost matrix between the two measures.</span>

<span class="sd">    Returns:</span>
<span class="sd">        torch.Tensor: The computed Sinkhorn distance.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">C</span> <span class="o">=</span> <span class="o">-</span><span class="n">c</span>
    <span class="n">x_points</span> <span class="o">=</span> <span class="n">C</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="o">-</span><span class="mi">2</span><span class="p">]</span>
    <span class="n">batch_size</span> <span class="o">=</span> <span class="n">C</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

    <span class="c1"># both marginals are fixed with equal weights</span>
    <span class="n">mu</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">torch</span><span class="o">.</span><span class="n">empty</span><span class="p">(</span>
            <span class="n">batch_size</span><span class="p">,</span>
            <span class="n">x_points</span><span class="p">,</span>
            <span class="n">dtype</span><span class="o">=</span><span class="n">C</span><span class="o">.</span><span class="n">dtype</span><span class="p">,</span>
            <span class="n">requires_grad</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
            <span class="n">device</span><span class="o">=</span><span class="n">C</span><span class="o">.</span><span class="n">device</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">fill_</span><span class="p">(</span><span class="mf">1.0</span> <span class="o">/</span> <span class="n">x_points</span><span class="p">)</span>
        <span class="o">.</span><span class="n">squeeze</span><span class="p">()</span>
    <span class="p">)</span>
    <span class="n">nu</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">torch</span><span class="o">.</span><span class="n">empty</span><span class="p">(</span>
            <span class="n">batch_size</span><span class="p">,</span>
            <span class="n">x_points</span><span class="p">,</span>
            <span class="n">dtype</span><span class="o">=</span><span class="n">C</span><span class="o">.</span><span class="n">dtype</span><span class="p">,</span>
            <span class="n">requires_grad</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
            <span class="n">device</span><span class="o">=</span><span class="n">C</span><span class="o">.</span><span class="n">device</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">fill_</span><span class="p">(</span><span class="mf">1.0</span> <span class="o">/</span> <span class="n">x_points</span><span class="p">)</span>
        <span class="o">.</span><span class="n">squeeze</span><span class="p">()</span>
    <span class="p">)</span>
    <span class="n">u</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">zeros_like</span><span class="p">(</span><span class="n">mu</span><span class="p">)</span>
    <span class="n">v</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">zeros_like</span><span class="p">(</span><span class="n">nu</span><span class="p">)</span>

    <span class="c1"># Stopping criterion</span>
    <span class="n">thresh</span> <span class="o">=</span> <span class="mf">1e-12</span>

    <span class="c1"># Sinkhorn iterations</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">max_iter</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">i</span> <span class="o">%</span> <span class="mi">2</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">u1</span> <span class="o">=</span> <span class="n">u</span>  <span class="c1"># useful to check the update</span>
            <span class="n">u</span> <span class="o">=</span> <span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">eps</span>
                <span class="o">*</span> <span class="p">(</span><span class="n">torch</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">mu</span><span class="p">)</span> <span class="o">-</span> <span class="n">torch</span><span class="o">.</span><span class="n">logsumexp</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">M</span><span class="p">(</span><span class="n">C</span><span class="p">,</span> <span class="n">u</span><span class="p">,</span> <span class="n">v</span><span class="p">),</span> <span class="n">dim</span><span class="o">=-</span><span class="mi">1</span><span class="p">))</span>
                <span class="o">+</span> <span class="n">u</span>
            <span class="p">)</span>
            <span class="n">err</span> <span class="o">=</span> <span class="p">(</span><span class="n">u</span> <span class="o">-</span> <span class="n">u1</span><span class="p">)</span><span class="o">.</span><span class="n">abs</span><span class="p">()</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">v</span> <span class="o">=</span> <span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">eps</span>
                <span class="o">*</span> <span class="p">(</span>
                    <span class="n">torch</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">nu</span><span class="p">)</span>
                    <span class="o">-</span> <span class="n">torch</span><span class="o">.</span><span class="n">logsumexp</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">M</span><span class="p">(</span><span class="n">C</span><span class="p">,</span> <span class="n">u</span><span class="p">,</span> <span class="n">v</span><span class="p">)</span><span class="o">.</span><span class="n">transpose</span><span class="p">(</span><span class="o">-</span><span class="mi">2</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">),</span> <span class="n">dim</span><span class="o">=-</span><span class="mi">1</span><span class="p">)</span>
                <span class="p">)</span>
                <span class="o">+</span> <span class="n">v</span>
            <span class="p">)</span>
            <span class="n">v</span> <span class="o">=</span> <span class="n">v</span><span class="o">.</span><span class="n">detach</span><span class="p">()</span><span class="o">.</span><span class="n">requires_grad_</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
            <span class="n">v</span><span class="p">[</span><span class="n">v</span> <span class="o">&gt;</span> <span class="mi">9</span> <span class="o">*</span> <span class="mf">1e8</span><span class="p">]</span> <span class="o">=</span> <span class="mf">0.0</span>
            <span class="n">v</span> <span class="o">=</span> <span class="n">v</span><span class="o">.</span><span class="n">detach</span><span class="p">()</span><span class="o">.</span><span class="n">requires_grad_</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">err</span><span class="o">.</span><span class="n">item</span><span class="p">()</span> <span class="o">&lt;</span> <span class="n">thresh</span><span class="p">:</span>
            <span class="k">break</span>

    <span class="n">U</span><span class="p">,</span> <span class="n">V</span> <span class="o">=</span> <span class="n">u</span><span class="p">,</span> <span class="n">v</span>
    <span class="c1"># Transport plan pi = diag(a)*K*diag(b)</span>
    <span class="n">pi</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">exp</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">M</span><span class="p">(</span><span class="n">C</span><span class="p">,</span> <span class="n">U</span><span class="p">,</span> <span class="n">V</span><span class="p">))</span>

    <span class="c1"># Sinkhorn distance</span>

    <span class="k">return</span> <span class="n">pi</span><span class="p">,</span> <span class="n">C</span><span class="p">,</span> <span class="n">U</span><span class="p">,</span> <span class="n">V</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>




  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="scprint.utils.utils" class="doc doc-heading">
            <code>scprint.utils.utils</code>


</h2>

    <div class="doc doc-contents first">



  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="scprint.utils.utils.category_str2int" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">category_str2int</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>category_str2int converts a list of category strings to a list of category integers.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>category_strs</code></b>
                  (<code><span title="typing.List">List</span>[str]</code>)
              –
              <div class="doc-md-description">
                <p>A list of category strings to be converted.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
                  <code><span title="typing.List">List</span>[int]</code>
              –
              <div class="doc-md-description">
                <p>List[int]: A list of integers corresponding to the input category strings.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
            <details class="quote">
              <summary>Source code in <code>scprint/utils/utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">category_str2int</span><span class="p">(</span><span class="n">category_strs</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">int</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    category_str2int converts a list of category strings to a list of category integers.</span>

<span class="sd">    Args:</span>
<span class="sd">        category_strs (List[str]): A list of category strings to be converted.</span>

<span class="sd">    Returns:</span>
<span class="sd">        List[int]: A list of integers corresponding to the input category strings.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">set_category_strs</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">category_strs</span><span class="p">)</span>
    <span class="n">name2id</span> <span class="o">=</span> <span class="p">{</span><span class="n">name</span><span class="p">:</span> <span class="n">i</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">name</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">set_category_strs</span><span class="p">)}</span>
    <span class="k">return</span> <span class="p">[</span><span class="n">name2id</span><span class="p">[</span><span class="n">name</span><span class="p">]</span> <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="n">category_strs</span><span class="p">]</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="scprint.utils.utils.createFoldersFor" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">createFoldersFor</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>will recursively create folders if needed until having all the folders required to save the file in this filepath</p>

            <details class="quote">
              <summary>Source code in <code>scprint/utils/utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">createFoldersFor</span><span class="p">(</span><span class="n">filepath</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    will recursively create folders if needed until having all the folders required to save the file in this filepath</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">prevval</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
    <span class="k">for</span> <span class="n">val</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">expanduser</span><span class="p">(</span><span class="n">filepath</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;/&quot;</span><span class="p">)[:</span><span class="o">-</span><span class="mi">1</span><span class="p">]:</span>
        <span class="n">prevval</span> <span class="o">+=</span> <span class="n">val</span> <span class="o">+</span> <span class="s2">&quot;/&quot;</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">prevval</span><span class="p">):</span>
            <span class="n">os</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">prevval</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="scprint.utils.utils.fileToList" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">fileToList</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>loads an input file with a\n b\n.. into a list [a,b,..]</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>input_str</code></b>
                  (<code>str</code>)
              –
              <div class="doc-md-description">
                <p>The input string to be completed.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>str</code></b>(                  <code>list</code>
)              –
              <div class="doc-md-description">
                <p>The completed string with 'complete' appended.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
            <details class="quote">
              <summary>Source code in <code>scprint/utils/utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">fileToList</span><span class="p">(</span><span class="n">filename</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">strconv</span><span class="p">:</span> <span class="nb">callable</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    loads an input file with a\\n b\\n.. into a list [a,b,..]</span>

<span class="sd">    Args:</span>
<span class="sd">        input_str (str): The input string to be completed.</span>

<span class="sd">    Returns:</span>
<span class="sd">        str: The completed string with &#39;complete&#39; appended.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="k">return</span> <span class="p">[</span><span class="n">strconv</span><span class="p">(</span><span class="n">val</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span> <span class="k">for</span> <span class="n">val</span> <span class="ow">in</span> <span class="n">f</span><span class="o">.</span><span class="n">readlines</span><span class="p">()]</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="scprint.utils.utils.find_required_colums" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">find_required_colums</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Find the required columns in AnnData, including celltype column, str_celltype
column, the gene name column, and the experimental batch key.</p>
<p>This function asks the user to input the required column names if the first
time loading the data. The names are saved in the config file and will be
automatically loaded next time.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>adata</code></b>
                  (<code><span title="anndata.AnnData">AnnData</span></code>)
              –
              <div class="doc-md-description">
                <p>The AnnData object.</p>
              </div>
            </li>
            <li>
              <b><code>id</code></b>
                  (<code>str</code>)
              –
              <div class="doc-md-description">
                <p>The id of the AnnData object, will be used as the file name for
saving the config file.</p>
              </div>
            </li>
            <li>
              <b><code>configs_dir</code></b>
                  (<code><span title="typing.Union">Union</span>[str, <span title="pathlib.Path">Path</span>]</code>)
              –
              <div class="doc-md-description">
                <p>The directory of saved config files.</p>
              </div>
            </li>
            <li>
              <b><code>update</code></b>
                  (<code>bool</code>, default:
                      <code>False</code>
)
              –
              <div class="doc-md-description">
                <p>Whether to update the config file.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
                  <code><span title="typing.List">List</span>[<span title="typing.Optional">Optional</span>[str]]</code>
              –
              <div class="doc-md-description">
                <p>List[Optional[str]]: The required columns, including celltype_col, str_celltype_col,
gene_col, and batch_col.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
            <details class="quote">
              <summary>Source code in <code>scprint/utils/utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">397</span>
<span class="normal">398</span>
<span class="normal">399</span>
<span class="normal">400</span>
<span class="normal">401</span>
<span class="normal">402</span>
<span class="normal">403</span>
<span class="normal">404</span>
<span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span>
<span class="normal">408</span>
<span class="normal">409</span>
<span class="normal">410</span>
<span class="normal">411</span>
<span class="normal">412</span>
<span class="normal">413</span>
<span class="normal">414</span>
<span class="normal">415</span>
<span class="normal">416</span>
<span class="normal">417</span>
<span class="normal">418</span>
<span class="normal">419</span>
<span class="normal">420</span>
<span class="normal">421</span>
<span class="normal">422</span>
<span class="normal">423</span>
<span class="normal">424</span>
<span class="normal">425</span>
<span class="normal">426</span>
<span class="normal">427</span>
<span class="normal">428</span>
<span class="normal">429</span>
<span class="normal">430</span>
<span class="normal">431</span>
<span class="normal">432</span>
<span class="normal">433</span>
<span class="normal">434</span>
<span class="normal">435</span>
<span class="normal">436</span>
<span class="normal">437</span>
<span class="normal">438</span>
<span class="normal">439</span>
<span class="normal">440</span>
<span class="normal">441</span>
<span class="normal">442</span>
<span class="normal">443</span>
<span class="normal">444</span>
<span class="normal">445</span>
<span class="normal">446</span>
<span class="normal">447</span>
<span class="normal">448</span>
<span class="normal">449</span>
<span class="normal">450</span>
<span class="normal">451</span>
<span class="normal">452</span>
<span class="normal">453</span>
<span class="normal">454</span>
<span class="normal">455</span>
<span class="normal">456</span>
<span class="normal">457</span>
<span class="normal">458</span>
<span class="normal">459</span>
<span class="normal">460</span>
<span class="normal">461</span>
<span class="normal">462</span>
<span class="normal">463</span>
<span class="normal">464</span>
<span class="normal">465</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">find_required_colums</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">:</span> <span class="n">AnnData</span><span class="p">,</span>
    <span class="nb">id</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">configs_dir</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Path</span><span class="p">],</span>
    <span class="n">update</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Find the required columns in AnnData, including celltype column, str_celltype</span>
<span class="sd">    column, the gene name column, and the experimental batch key.</span>

<span class="sd">    This function asks the user to input the required column names if the first</span>
<span class="sd">    time loading the data. The names are saved in the config file and will be</span>
<span class="sd">    automatically loaded next time.</span>

<span class="sd">    Args:</span>
<span class="sd">        adata (AnnData): The AnnData object.</span>
<span class="sd">        id (str): The id of the AnnData object, will be used as the file name for</span>
<span class="sd">            saving the config file.</span>
<span class="sd">        configs_dir (Union[str, Path]): The directory of saved config files.</span>
<span class="sd">        update (bool): Whether to update the config file.</span>

<span class="sd">    Returns:</span>
<span class="sd">        List[Optional[str]]: The required columns, including celltype_col, str_celltype_col,</span>
<span class="sd">            gene_col, and batch_col.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">configs_dir</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
        <span class="n">configs_dir</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">configs_dir</span><span class="p">)</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">configs_dir</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
        <span class="n">configs_dir</span><span class="o">.</span><span class="n">mkdir</span><span class="p">()</span>

    <span class="n">config_file</span> <span class="o">=</span> <span class="n">configs_dir</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">id</span><span class="si">}</span><span class="s2">.json&quot;</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">config_file</span><span class="o">.</span><span class="n">exists</span><span class="p">()</span> <span class="ow">or</span> <span class="n">update</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span>
            <span class="s2">&quot;The config file does not exist, this may be the first time &quot;</span>
            <span class="s2">&quot;loading the data. </span><span class="se">\n</span><span class="s2">Please input the required column names.&quot;</span>
        <span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
        <span class="n">celltype_col</span> <span class="o">=</span> <span class="n">_indicate_col_name</span><span class="p">(</span>
            <span class="n">adata</span><span class="p">,</span>
            <span class="s2">&quot;Please input the celltype column name (skip if not applicable): &quot;</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">str_celltype_col</span> <span class="o">=</span> <span class="n">_indicate_col_name</span><span class="p">(</span>
            <span class="n">adata</span><span class="p">,</span> <span class="s2">&quot;Please input the str_celltype column name: &quot;</span>
        <span class="p">)</span>
        <span class="n">gene_col</span> <span class="o">=</span> <span class="n">_indicate_col_name</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="s2">&quot;Please input the gene column name: &quot;</span><span class="p">)</span>
        <span class="n">batch_col</span> <span class="o">=</span> <span class="n">_indicate_col_name</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="s2">&quot;Please input the batch column name: &quot;</span><span class="p">)</span>

        <span class="n">config</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;celltype_col&quot;</span><span class="p">:</span> <span class="n">celltype_col</span><span class="p">,</span>
            <span class="s2">&quot;str_celltype_col&quot;</span><span class="p">:</span> <span class="n">str_celltype_col</span><span class="p">,</span>
            <span class="s2">&quot;gene_col&quot;</span><span class="p">:</span> <span class="n">gene_col</span><span class="p">,</span>
            <span class="s2">&quot;batch_col&quot;</span><span class="p">:</span> <span class="n">batch_col</span><span class="p">,</span>
        <span class="p">}</span>

        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">config_file</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
            <span class="n">json</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="n">config</span><span class="p">,</span> <span class="n">f</span><span class="p">)</span>

    <span class="k">else</span><span class="p">:</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">config_file</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
            <span class="n">config</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>

    <span class="k">return</span> <span class="p">[</span>
        <span class="n">config</span><span class="p">[</span><span class="s2">&quot;celltype_col&quot;</span><span class="p">],</span>
        <span class="n">config</span><span class="p">[</span><span class="s2">&quot;str_celltype_col&quot;</span><span class="p">],</span>
        <span class="n">config</span><span class="p">[</span><span class="s2">&quot;gene_col&quot;</span><span class="p">],</span>
        <span class="n">config</span><span class="p">[</span><span class="s2">&quot;batch_col&quot;</span><span class="p">],</span>
    <span class="p">]</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="scprint.utils.utils.getBiomartTable" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">getBiomartTable</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>generate a genelist dataframe from ensembl's biomart</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>ensemble_server</code></b>
                  (<code>str</code>, default:
                      <code>&#39;http://jul2023.archive.ensembl.org/biomart&#39;</code>
)
              –
              <div class="doc-md-description">
                <p>The URL of the Ensembl Biomart server. Defaults to "http://jul2023.archive.ensembl.org/biomart".</p>
              </div>
            </li>
            <li>
              <b><code>useCache</code></b>
                  (<code>bool</code>, default:
                      <code>False</code>
)
              –
              <div class="doc-md-description">
                <p>Whether to use cached data if available. Defaults to False.</p>
              </div>
            </li>
            <li>
              <b><code>cache_folder</code></b>
                  (<code>str</code>, default:
                      <code>&#39;/tmp/biomart/&#39;</code>
)
              –
              <div class="doc-md-description">
                <p>The directory where cached data will be stored. Defaults to "/tmp/biomart/".</p>
              </div>
            </li>
            <li>
              <b><code>attributes</code></b>
                  (<code>list</code>, default:
                      <code>[]</code>
)
              –
              <div class="doc-md-description">
                <p>Additional attributes to fetch from the server. Defaults to an empty list.</p>
              </div>
            </li>
            <li>
              <b><code>bypass_attributes</code></b>
                  (<code>bool</code>, default:
                      <code>False</code>
)
              –
              <div class="doc-md-description">
                <p>Whether to bypass the default attributes. Defaults to False.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Raises:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
                  <code>ValueError</code>
              –
              <div class="doc-md-description">
                <p>If the result is not a pandas DataFrame.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
              –
              <div class="doc-md-description">
                <p>pd.DataFrame: A DataFrame containing the gene list from Ensembl's Biomart.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
            <details class="quote">
              <summary>Source code in <code>scprint/utils/utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">getBiomartTable</span><span class="p">(</span>
    <span class="n">ensemble_server</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;http://jul2023.archive.ensembl.org/biomart&quot;</span><span class="p">,</span>
    <span class="n">useCache</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">cache_folder</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;/tmp/biomart/&quot;</span><span class="p">,</span>
    <span class="n">attributes</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[],</span>
    <span class="n">bypass_attributes</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;generate a genelist dataframe from ensembl&#39;s biomart</span>

<span class="sd">    Args:</span>
<span class="sd">        ensemble_server (str, optional): The URL of the Ensembl Biomart server. Defaults to &quot;http://jul2023.archive.ensembl.org/biomart&quot;.</span>
<span class="sd">        useCache (bool, optional): Whether to use cached data if available. Defaults to False.</span>
<span class="sd">        cache_folder (str, optional): The directory where cached data will be stored. Defaults to &quot;/tmp/biomart/&quot;.</span>
<span class="sd">        attributes (list, optional): Additional attributes to fetch from the server. Defaults to an empty list.</span>
<span class="sd">        bypass_attributes (bool, optional): Whether to bypass the default attributes. Defaults to False.</span>

<span class="sd">    Raises:</span>
<span class="sd">        ValueError: If the result is not a pandas DataFrame.</span>

<span class="sd">    Returns:</span>
<span class="sd">        pd.DataFrame: A DataFrame containing the gene list from Ensembl&#39;s Biomart.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">attr</span> <span class="o">=</span> <span class="p">(</span>
        <span class="p">[</span>
            <span class="s2">&quot;ensembl_gene_id&quot;</span><span class="p">,</span>
            <span class="s2">&quot;hgnc_symbol&quot;</span><span class="p">,</span>
            <span class="s2">&quot;gene_biotype&quot;</span><span class="p">,</span>
            <span class="s2">&quot;entrezgene_id&quot;</span><span class="p">,</span>
        <span class="p">]</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">bypass_attributes</span>
        <span class="k">else</span> <span class="p">[]</span>
    <span class="p">)</span>
    <span class="k">assert</span> <span class="n">cache_folder</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;/&quot;</span>

    <span class="n">cache_folder</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">expanduser</span><span class="p">(</span><span class="n">cache_folder</span><span class="p">)</span>
    <span class="n">createFoldersFor</span><span class="p">(</span><span class="n">cache_folder</span><span class="p">)</span>
    <span class="n">cachefile</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">cache_folder</span><span class="p">,</span> <span class="s2">&quot;.biomart.csv&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">useCache</span> <span class="o">&amp;</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">cachefile</span><span class="p">):</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;fetching gene names from biomart cache&quot;</span><span class="p">)</span>
        <span class="n">res</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">cachefile</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;downloading gene names from biomart&quot;</span><span class="p">)</span>

        <span class="n">res</span> <span class="o">=</span> <span class="n">_fetchFromServer</span><span class="p">(</span><span class="n">ensemble_server</span><span class="p">,</span> <span class="n">attr</span> <span class="o">+</span> <span class="n">attributes</span><span class="p">)</span>
        <span class="n">res</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">cachefile</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

    <span class="n">res</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="n">attr</span> <span class="o">+</span> <span class="n">attributes</span>
    <span class="k">if</span> <span class="nb">type</span><span class="p">(</span><span class="n">res</span><span class="p">)</span> <span class="ow">is</span> <span class="ow">not</span> <span class="nb">type</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()):</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;should be a dataframe&quot;</span><span class="p">)</span>
    <span class="n">res</span> <span class="o">=</span> <span class="n">res</span><span class="p">[</span><span class="o">~</span><span class="p">(</span><span class="n">res</span><span class="p">[</span><span class="s2">&quot;ensembl_gene_id&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isna</span><span class="p">()</span> <span class="o">&amp;</span> <span class="n">res</span><span class="p">[</span><span class="s2">&quot;hgnc_symbol&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isna</span><span class="p">())]</span>
    <span class="n">res</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">res</span><span class="p">[</span><span class="n">res</span><span class="o">.</span><span class="n">hgnc_symbol</span><span class="o">.</span><span class="n">isna</span><span class="p">()]</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="s2">&quot;hgnc_symbol&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">res</span><span class="p">[</span>
        <span class="n">res</span><span class="o">.</span><span class="n">hgnc_symbol</span><span class="o">.</span><span class="n">isna</span><span class="p">()</span>
    <span class="p">][</span><span class="s2">&quot;ensembl_gene_id&quot;</span><span class="p">]</span>

    <span class="k">return</span> <span class="n">res</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="scprint.utils.utils.get_free_gpu" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">get_free_gpu</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>get_free_gpu finds the GPU with the most free memory using nvidia-smi.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>int</code></b>              –
              <div class="doc-md-description">
                <p>The index of the GPU with the most free memory.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
            <details class="quote">
              <summary>Source code in <code>scprint/utils/utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_free_gpu</span><span class="p">():</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    get_free_gpu finds the GPU with the most free memory using nvidia-smi.</span>

<span class="sd">    Returns:</span>
<span class="sd">        int: The index of the GPU with the most free memory.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="kn">import</span> <span class="nn">subprocess</span>
    <span class="kn">import</span> <span class="nn">sys</span>
    <span class="kn">from</span> <span class="nn">io</span> <span class="kn">import</span> <span class="n">StringIO</span>
    <span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

    <span class="n">gpu_stats</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">check_output</span><span class="p">(</span>
        <span class="p">[</span>
            <span class="s2">&quot;nvidia-smi&quot;</span><span class="p">,</span>
            <span class="s2">&quot;--format=csv&quot;</span><span class="p">,</span>
            <span class="s2">&quot;--query-gpu=memory.used,memory.free&quot;</span><span class="p">,</span>
        <span class="p">]</span>
    <span class="p">)</span><span class="o">.</span><span class="n">decode</span><span class="p">(</span><span class="s2">&quot;utf-8&quot;</span><span class="p">)</span>
    <span class="n">gpu_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
        <span class="n">StringIO</span><span class="p">(</span><span class="n">gpu_stats</span><span class="p">),</span> <span class="n">names</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;memory.used&quot;</span><span class="p">,</span> <span class="s2">&quot;memory.free&quot;</span><span class="p">],</span> <span class="n">skiprows</span><span class="o">=</span><span class="mi">1</span>
    <span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;GPU usage:</span><span class="se">\n</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">gpu_df</span><span class="p">))</span>
    <span class="n">gpu_df</span><span class="p">[</span><span class="s2">&quot;memory.free&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">gpu_df</span><span class="p">[</span><span class="s2">&quot;memory.free&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">rstrip</span><span class="p">(</span><span class="s2">&quot; [MiB]&quot;</span><span class="p">)))</span>
    <span class="n">idx</span> <span class="o">=</span> <span class="n">gpu_df</span><span class="p">[</span><span class="s2">&quot;memory.free&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">idxmax</span><span class="p">()</span>
    <span class="nb">print</span><span class="p">(</span>
        <span class="s2">&quot;Find free GPU</span><span class="si">{}</span><span class="s2"> with </span><span class="si">{}</span><span class="s2"> free MiB&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">idx</span><span class="p">,</span> <span class="n">gpu_df</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">idx</span><span class="p">][</span><span class="s2">&quot;memory.free&quot;</span><span class="p">])</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="n">idx</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="scprint.utils.utils.get_git_commit" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">get_git_commit</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>get_git_commit gets the current git commit hash.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>str</code></b>              –
              <div class="doc-md-description">
                <p>The current git commit</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
            <details class="quote">
              <summary>Source code in <code>scprint/utils/utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_git_commit</span><span class="p">():</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    get_git_commit gets the current git commit hash.</span>

<span class="sd">    Returns:</span>
<span class="sd">        str: The current git commit</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">check_output</span><span class="p">([</span><span class="s2">&quot;git&quot;</span><span class="p">,</span> <span class="s2">&quot;rev-parse&quot;</span><span class="p">,</span> <span class="s2">&quot;HEAD&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">decode</span><span class="p">(</span><span class="s2">&quot;utf-8&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="scprint.utils.utils.histogram" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">histogram</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Plot histogram of the data.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>data</code></b>
                  (<code><span title="typing.List">List</span>[<span title="numpy.ndarray">ndarray</span>]</code>, default:
                      <code>()</code>
)
              –
              <div class="doc-md-description">
                <p>The data to plot.</p>
              </div>
            </li>
            <li>
              <b><code>label</code></b>
                  (<code><span title="typing.List">List</span>[str]</code>, default:
                      <code>[&#39;train&#39;, &#39;valid&#39;]</code>
)
              –
              <div class="doc-md-description">
                <p>The label of the data.</p>
              </div>
            </li>
            <li>
              <b><code>color</code></b>
                  (<code><span title="typing.List">List</span>[str]</code>, default:
                      <code>[&#39;blue&#39;, &#39;red&#39;]</code>
)
              –
              <div class="doc-md-description">
                <p>The color of the data.</p>
              </div>
            </li>
            <li>
              <b><code>figsize</code></b>
                  (<code><span title="typing.Tuple">Tuple</span>[int, int]</code>, default:
                      <code>(9, 4)</code>
)
              –
              <div class="doc-md-description">
                <p>The size of the figure.</p>
              </div>
            </li>
            <li>
              <b><code>title</code></b>
                  (<code><span title="typing.Optional">Optional</span>[str]</code>, default:
                      <code>None</code>
)
              –
              <div class="doc-md-description">
                <p>The title of the figure.</p>
              </div>
            </li>
            <li>
              <b><code>show</code></b>
                  (<code>bool</code>, default:
                      <code>False</code>
)
              –
              <div class="doc-md-description">
                <p>Whether to show the figure.</p>
              </div>
            </li>
            <li>
              <b><code>save</code></b>
                  (<code><span title="typing.Optional">Optional</span>[str]</code>, default:
                      <code>None</code>
)
              –
              <div class="doc-md-description">
                <p>The path to save the figure.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
                  <code><span title="matplotlib.axes.Axes">Axes</span></code>
              –
              <div class="doc-md-description">
                <p>axes.Axes: The axes of the figure.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
            <details class="quote">
              <summary>Source code in <code>scprint/utils/utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">histogram</span><span class="p">(</span>
    <span class="o">*</span><span class="n">data</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">],</span>
    <span class="n">label</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;train&quot;</span><span class="p">,</span> <span class="s2">&quot;valid&quot;</span><span class="p">],</span>
    <span class="n">color</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;blue&quot;</span><span class="p">,</span> <span class="s2">&quot;red&quot;</span><span class="p">],</span>
    <span class="n">figsize</span><span class="p">:</span> <span class="n">Tuple</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="mi">9</span><span class="p">,</span> <span class="mi">4</span><span class="p">),</span>
    <span class="n">title</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">show</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">save</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">axes</span><span class="o">.</span><span class="n">Axes</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Plot histogram of the data.</span>

<span class="sd">    Args:</span>
<span class="sd">        data (List[np.ndarray]): The data to plot.</span>
<span class="sd">        label (List[str]): The label of the data.</span>
<span class="sd">        color (List[str]): The color of the data.</span>
<span class="sd">        figsize (Tuple[int, int]): The size of the figure.</span>
<span class="sd">        title (Optional[str]): The title of the figure.</span>
<span class="sd">        show (bool): Whether to show the figure.</span>
<span class="sd">        save (Optional[str]): The path to save the figure.</span>

<span class="sd">    Returns:</span>
<span class="sd">        axes.Axes: The axes of the figure.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># show histogram of the clipped values</span>
    <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">data</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">label</span><span class="p">),</span> <span class="s2">&quot;The number of data and labels must be equal.&quot;</span>

    <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="n">figsize</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">150</span><span class="p">)</span>
    <span class="n">max_value</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">data</span><span class="p">)</span> <span class="k">for</span> <span class="n">data</span> <span class="ow">in</span> <span class="n">data</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span>
        <span class="p">[</span><span class="n">d</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span> <span class="k">for</span> <span class="n">d</span> <span class="ow">in</span> <span class="n">data</span><span class="p">],</span>
        <span class="n">bins</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">max_value</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span> <span class="o">+</span> <span class="mf">0.5</span> <span class="k">if</span> <span class="n">max_value</span> <span class="o">&lt;</span> <span class="mi">60</span> <span class="k">else</span> <span class="mi">60</span><span class="p">,</span>
        <span class="n">label</span><span class="o">=</span><span class="n">label</span><span class="p">,</span>
        <span class="n">density</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="n">histtype</span><span class="o">=</span><span class="s2">&quot;bar&quot;</span><span class="p">,</span>
        <span class="n">linewidth</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
        <span class="n">rwidth</span><span class="o">=</span><span class="mf">0.85</span><span class="p">,</span>
        <span class="n">color</span><span class="o">=</span><span class="n">color</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;counts&quot;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;density&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">title</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="n">title</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">show</span><span class="p">:</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>

    <span class="k">if</span> <span class="n">save</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">save</span><span class="p">,</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s2">&quot;tight&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">ax</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="scprint.utils.utils.inf_loop" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">inf_loop</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>wrapper function for endless data loader.</p>

            <details class="quote">
              <summary>Source code in <code>scprint/utils/utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">615</span>
<span class="normal">616</span>
<span class="normal">617</span>
<span class="normal">618</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">inf_loop</span><span class="p">(</span><span class="n">data_loader</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;wrapper function for endless data loader.&quot;&quot;&quot;</span>
    <span class="k">for</span> <span class="n">loader</span> <span class="ow">in</span> <span class="n">repeat</span><span class="p">(</span><span class="n">data_loader</span><span class="p">):</span>
        <span class="k">yield from</span> <span class="n">loader</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="scprint.utils.utils.isnotebook" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">isnotebook</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>check whether excuting in jupyter notebook.</p>

            <details class="quote">
              <summary>Source code in <code>scprint/utils/utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">isnotebook</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;check whether excuting in jupyter notebook.&quot;&quot;&quot;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">shell</span> <span class="o">=</span> <span class="n">get_ipython</span><span class="p">()</span><span class="o">.</span><span class="vm">__class__</span><span class="o">.</span><span class="vm">__name__</span>
        <span class="k">if</span> <span class="n">shell</span> <span class="o">==</span> <span class="s2">&quot;ZMQInteractiveShell&quot;</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">True</span>  <span class="c1"># Jupyter notebook or qtconsole</span>
        <span class="k">elif</span> <span class="n">shell</span> <span class="o">==</span> <span class="s2">&quot;TerminalInteractiveShell&quot;</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">True</span>  <span class="c1"># Terminal running IPython</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">False</span>  <span class="c1"># Other type (?)</span>
    <span class="k">except</span> <span class="ne">NameError</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">False</span>  <span class="c1"># Probably standard Python interpreter</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="scprint.utils.utils.listToFile" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">listToFile</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>listToFile loads a list with [a,b,..] into an input file a\n b\n..</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>l</code></b>
                  (<code>list</code>)
              –
              <div class="doc-md-description">
                <p>The list of elements to be written to the file.</p>
              </div>
            </li>
            <li>
              <b><code>filename</code></b>
                  (<code>str</code>)
              –
              <div class="doc-md-description">
                <p>The name of the file where the list will be written.</p>
              </div>
            </li>
            <li>
              <b><code>strconv</code></b>
                  (<code>callable</code>, default:
                      <code>lambda x: str(x)</code>
)
              –
              <div class="doc-md-description">
                <p>A function to convert each element of the list to a string. Defaults to str.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
                  <code>None</code>
              –
              <div class="doc-md-description">
                <p>None</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
            <details class="quote">
              <summary>Source code in <code>scprint/utils/utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">listToFile</span><span class="p">(</span><span class="n">l</span><span class="p">:</span> <span class="nb">list</span><span class="p">,</span> <span class="n">filename</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">strconv</span><span class="p">:</span> <span class="nb">callable</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">))</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    listToFile loads a list with [a,b,..] into an input file a\\n b\\n..</span>

<span class="sd">    Args:</span>
<span class="sd">        l (list): The list of elements to be written to the file.</span>
<span class="sd">        filename (str): The name of the file where the list will be written.</span>
<span class="sd">        strconv (callable, optional): A function to convert each element of the list to a string. Defaults to str.</span>

<span class="sd">    Returns:</span>
<span class="sd">        None</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">l</span><span class="p">:</span>
            <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">strconv</span><span class="p">(</span><span class="n">item</span><span class="p">))</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="scprint.utils.utils.load_genes" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">load_genes</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>load_genes loads the genes for a given organism.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>organisms</code></b>
                  (<code><span title="typing.Union">Union</span>[str, list]</code>, default:
                      <code>&#39;NCBITaxon:9606&#39;</code>
)
              –
              <div class="doc-md-description">
                <p>A string or list of strings representing the organism(s) to load genes for. Defaults to "NCBITaxon:9606".</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              –
              <div class="doc-md-description">
                <p>pd.DataFrame: A DataFrame containing gene information for the specified organism(s).</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
            <details class="quote">
              <summary>Source code in <code>scprint/utils/utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">load_genes</span><span class="p">(</span><span class="n">organisms</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">list</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;NCBITaxon:9606&quot;</span><span class="p">):</span>  <span class="c1"># &quot;NCBITaxon:10090&quot;,</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    load_genes loads the genes for a given organism.</span>

<span class="sd">    Args:</span>
<span class="sd">        organisms (Union[str, list], optional): A string or list of strings representing the organism(s) to load genes for. Defaults to &quot;NCBITaxon:9606&quot;.</span>

<span class="sd">    Returns:</span>
<span class="sd">        pd.DataFrame: A DataFrame containing gene information for the specified organism(s).</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">organismdf</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">if</span> <span class="nb">type</span><span class="p">(</span><span class="n">organisms</span><span class="p">)</span> <span class="o">==</span> <span class="nb">str</span><span class="p">:</span>
        <span class="n">organisms</span> <span class="o">=</span> <span class="p">[</span><span class="n">organisms</span><span class="p">]</span>
    <span class="k">for</span> <span class="n">organism</span> <span class="ow">in</span> <span class="n">organisms</span><span class="p">:</span>
        <span class="n">genesdf</span> <span class="o">=</span> <span class="n">bt</span><span class="o">.</span><span class="n">Gene</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
            <span class="n">organism_id</span><span class="o">=</span><span class="n">bt</span><span class="o">.</span><span class="n">Organism</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">ontology_id</span><span class="o">=</span><span class="n">organism</span><span class="p">)</span><span class="o">.</span><span class="n">first</span><span class="p">()</span><span class="o">.</span><span class="n">id</span>
        <span class="p">)</span><span class="o">.</span><span class="n">df</span><span class="p">()</span>
        <span class="n">genesdf</span> <span class="o">=</span> <span class="n">genesdf</span><span class="p">[</span><span class="o">~</span><span class="n">genesdf</span><span class="p">[</span><span class="s2">&quot;public_source_id&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isna</span><span class="p">()]</span>
        <span class="n">genesdf</span> <span class="o">=</span> <span class="n">genesdf</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">(</span><span class="n">subset</span><span class="o">=</span><span class="s2">&quot;ensembl_gene_id&quot;</span><span class="p">)</span>
        <span class="n">genesdf</span> <span class="o">=</span> <span class="n">genesdf</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s2">&quot;ensembl_gene_id&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">sort_index</span><span class="p">()</span>
        <span class="c1"># mitochondrial genes</span>
        <span class="n">genesdf</span><span class="p">[</span><span class="s2">&quot;mt&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">genesdf</span><span class="o">.</span><span class="n">symbol</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;MT-&quot;</span><span class="p">)</span>
        <span class="c1"># ribosomal genes</span>
        <span class="n">genesdf</span><span class="p">[</span><span class="s2">&quot;ribo&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">genesdf</span><span class="o">.</span><span class="n">symbol</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">startswith</span><span class="p">((</span><span class="s2">&quot;RPS&quot;</span><span class="p">,</span> <span class="s2">&quot;RPL&quot;</span><span class="p">))</span>
        <span class="c1"># hemoglobin genes.</span>
        <span class="n">genesdf</span><span class="p">[</span><span class="s2">&quot;hb&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">genesdf</span><span class="o">.</span><span class="n">symbol</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">((</span><span class="s2">&quot;^HB[^(P)]&quot;</span><span class="p">))</span>
        <span class="n">genesdf</span><span class="p">[</span><span class="s2">&quot;organism&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">organism</span>
        <span class="n">organismdf</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">genesdf</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">organismdf</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="scprint.utils.utils.map_raw_id_to_vocab_id" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">map_raw_id_to_vocab_id</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Map some raw ids which are indices of the raw gene names to the indices of the</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>raw_ids</code></b>
                  (<code><span title="typing.Union">Union</span>[<span title="numpy.ndarray">ndarray</span>, <span title="torch.Tensor">Tensor</span>]</code>)
              –
              <div class="doc-md-description">
                <p>the raw ids to map</p>
              </div>
            </li>
            <li>
              <b><code>gene_ids</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>the gene ids to map to</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
            <details class="quote">
              <summary>Source code in <code>scprint/utils/utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">479</span>
<span class="normal">480</span>
<span class="normal">481</span>
<span class="normal">482</span>
<span class="normal">483</span>
<span class="normal">484</span>
<span class="normal">485</span>
<span class="normal">486</span>
<span class="normal">487</span>
<span class="normal">488</span>
<span class="normal">489</span>
<span class="normal">490</span>
<span class="normal">491</span>
<span class="normal">492</span>
<span class="normal">493</span>
<span class="normal">494</span>
<span class="normal">495</span>
<span class="normal">496</span>
<span class="normal">497</span>
<span class="normal">498</span>
<span class="normal">499</span>
<span class="normal">500</span>
<span class="normal">501</span>
<span class="normal">502</span>
<span class="normal">503</span>
<span class="normal">504</span>
<span class="normal">505</span>
<span class="normal">506</span>
<span class="normal">507</span>
<span class="normal">508</span>
<span class="normal">509</span>
<span class="normal">510</span>
<span class="normal">511</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">map_raw_id_to_vocab_id</span><span class="p">(</span>
    <span class="n">raw_ids</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">torch</span><span class="o">.</span><span class="n">Tensor</span><span class="p">],</span>
    <span class="n">gene_ids</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Union</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">torch</span><span class="o">.</span><span class="n">Tensor</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Map some raw ids which are indices of the raw gene names to the indices of the</span>

<span class="sd">    Args:</span>
<span class="sd">        raw_ids: the raw ids to map</span>
<span class="sd">        gene_ids: the gene ids to map to</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">raw_ids</span><span class="p">,</span> <span class="n">torch</span><span class="o">.</span><span class="n">Tensor</span><span class="p">):</span>
        <span class="n">device</span> <span class="o">=</span> <span class="n">raw_ids</span><span class="o">.</span><span class="n">device</span>
        <span class="n">dtype</span> <span class="o">=</span> <span class="n">raw_ids</span><span class="o">.</span><span class="n">dtype</span>
        <span class="n">return_pt</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="n">raw_ids</span> <span class="o">=</span> <span class="n">raw_ids</span><span class="o">.</span><span class="n">cpu</span><span class="p">()</span><span class="o">.</span><span class="n">numpy</span><span class="p">()</span>
    <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">raw_ids</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">):</span>
        <span class="n">return_pt</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="n">dtype</span> <span class="o">=</span> <span class="n">raw_ids</span><span class="o">.</span><span class="n">dtype</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;raw_ids must be either torch.Tensor or np.ndarray.&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">raw_ids</span><span class="o">.</span><span class="n">ndim</span> <span class="o">!=</span> <span class="mi">1</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;raw_ids must be 1d, got </span><span class="si">{</span><span class="n">raw_ids</span><span class="o">.</span><span class="n">ndim</span><span class="si">}</span><span class="s2">d.&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">gene_ids</span><span class="o">.</span><span class="n">ndim</span> <span class="o">!=</span> <span class="mi">1</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;gene_ids must be 1d, got </span><span class="si">{</span><span class="n">gene_ids</span><span class="o">.</span><span class="n">ndim</span><span class="si">}</span><span class="s2">d.&quot;</span><span class="p">)</span>

    <span class="n">mapped_ids</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span> <span class="o">=</span> <span class="n">gene_ids</span><span class="p">[</span><span class="n">raw_ids</span><span class="p">]</span>
    <span class="k">assert</span> <span class="n">mapped_ids</span><span class="o">.</span><span class="n">shape</span> <span class="o">==</span> <span class="n">raw_ids</span><span class="o">.</span><span class="n">shape</span>
    <span class="k">if</span> <span class="n">return_pt</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">torch</span><span class="o">.</span><span class="n">from_numpy</span><span class="p">(</span><span class="n">mapped_ids</span><span class="p">)</span><span class="o">.</span><span class="n">type</span><span class="p">(</span><span class="n">dtype</span><span class="p">)</span><span class="o">.</span><span class="n">to</span><span class="p">(</span><span class="n">device</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">mapped_ids</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="n">dtype</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="scprint.utils.utils.onto_to_name" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">onto_to_name</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Convert ontology IDs to names using a given ontology and schema.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>ids</code></b>
                  (<code>list</code>)
              –
              <div class="doc-md-description">
                <p>A list of ontology IDs to be converted.</p>
              </div>
            </li>
            <li>
              <b><code>onto</code></b>
              –
              <div class="doc-md-description">
                <p>The ontology object used for searching.</p>
              </div>
            </li>
            <li>
              <b><code>schema</code></b>
                  (<code>str</code>, default:
                      <code>&#39;http://www.ebi.ac.uk/efo/&#39;</code>
)
              –
              <div class="doc-md-description">
                <p>The schema URL to be used for constructing the search IRI. Defaults to "http://www.ebi.ac.uk/efo/".</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>list</code></b>(                  <code>list</code>
)              –
              <div class="doc-md-description">
                <p>A list of names corresponding to the given ontology IDs.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
            <details class="quote">
              <summary>Source code in <code>scprint/utils/utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">onto_to_name</span><span class="p">(</span><span class="n">ids</span><span class="p">:</span> <span class="nb">list</span><span class="p">,</span> <span class="n">onto</span><span class="p">,</span> <span class="n">schema</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;http://www.ebi.ac.uk/efo/&quot;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Convert ontology IDs to names using a given ontology and schema.</span>

<span class="sd">    Args:</span>
<span class="sd">        ids (list): A list of ontology IDs to be converted.</span>
<span class="sd">        onto: The ontology object used for searching.</span>
<span class="sd">        schema (str, optional): The schema URL to be used for constructing the search IRI. Defaults to &quot;http://www.ebi.ac.uk/efo/&quot;.</span>

<span class="sd">    Returns:</span>
<span class="sd">        list: A list of names corresponding to the given ontology IDs.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">names</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">val</span> <span class="ow">in</span> <span class="n">ids</span><span class="p">:</span>
        <span class="n">res</span> <span class="o">=</span> <span class="n">onto</span><span class="o">.</span><span class="n">search_one</span><span class="p">(</span><span class="n">iri</span><span class="o">=</span><span class="n">schema</span> <span class="o">+</span> <span class="n">val</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;:&quot;</span><span class="p">,</span> <span class="s2">&quot;_&quot;</span><span class="p">))</span>
        <span class="k">if</span> <span class="n">res</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="n">val</span><span class="p">,</span> <span class="s2">&quot;was not found&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">names</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">res</span><span class="o">.</span><span class="n">label</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">names</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="scprint.utils.utils.pd_load_cached" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">pd_load_cached</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>pd_load_cached loads a csv file from a url, and caches it locally</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>url</code></b>
                  (<code>str</code>)
              –
              <div class="doc-md-description">
                <p>The URL of the CSV file to be loaded.</p>
              </div>
            </li>
            <li>
              <b><code>loc</code></b>
                  (<code>str</code>, default:
                      <code>&#39;/tmp/&#39;</code>
)
              –
              <div class="doc-md-description">
                <p>The local directory where the file will be cached. Defaults to "/tmp/".</p>
              </div>
            </li>
            <li>
              <b><code>cache</code></b>
                  (<code>bool</code>, default:
                      <code>True</code>
)
              –
              <div class="doc-md-description">
                <p>Whether to use the cached file if it exists. Defaults to True.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
              –
              <div class="doc-md-description">
                <p>pd.DataFrame: The loaded data as a pandas DataFrame.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
            <details class="quote">
              <summary>Source code in <code>scprint/utils/utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">pd_load_cached</span><span class="p">(</span><span class="n">url</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">loc</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;/tmp/&quot;</span><span class="p">,</span> <span class="n">cache</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    pd_load_cached loads a csv file from a url, and caches it locally</span>

<span class="sd">    Args:</span>
<span class="sd">        url (str): The URL of the CSV file to be loaded.</span>
<span class="sd">        loc (str, optional): The local directory where the file will be cached. Defaults to &quot;/tmp/&quot;.</span>
<span class="sd">        cache (bool, optional): Whether to use the cached file if it exists. Defaults to True.</span>

<span class="sd">    Returns:</span>
<span class="sd">        pd.DataFrame: The loaded data as a pandas DataFrame.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Check if the file exists, if not, download it</span>
    <span class="n">loc</span> <span class="o">+=</span> <span class="n">url</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;/&quot;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">loc</span><span class="p">)</span> <span class="ow">or</span> <span class="ow">not</span> <span class="n">cache</span><span class="p">:</span>
        <span class="n">urllib</span><span class="o">.</span><span class="n">request</span><span class="o">.</span><span class="n">urlretrieve</span><span class="p">(</span><span class="n">url</span><span class="p">,</span> <span class="n">loc</span><span class="p">)</span>
    <span class="c1"># Load the data from the file</span>
    <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">loc</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="scprint.utils.utils.prepare_device" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">prepare_device</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>setup GPU device if available. get gpu device indices which are used for DataParallel</p>

            <details class="quote">
              <summary>Source code in <code>scprint/utils/utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">621</span>
<span class="normal">622</span>
<span class="normal">623</span>
<span class="normal">624</span>
<span class="normal">625</span>
<span class="normal">626</span>
<span class="normal">627</span>
<span class="normal">628</span>
<span class="normal">629</span>
<span class="normal">630</span>
<span class="normal">631</span>
<span class="normal">632</span>
<span class="normal">633</span>
<span class="normal">634</span>
<span class="normal">635</span>
<span class="normal">636</span>
<span class="normal">637</span>
<span class="normal">638</span>
<span class="normal">639</span>
<span class="normal">640</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">prepare_device</span><span class="p">(</span><span class="n">n_gpu_use</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    setup GPU device if available. get gpu device indices which are used for DataParallel</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">n_gpu</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">cuda</span><span class="o">.</span><span class="n">device_count</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">n_gpu_use</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="ow">and</span> <span class="n">n_gpu</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span>
            <span class="s2">&quot;Warning: There&#39;s no GPU available on this machine,&quot;</span>
            <span class="s2">&quot;training will be performed on CPU.&quot;</span>
        <span class="p">)</span>
        <span class="n">n_gpu_use</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">if</span> <span class="n">n_gpu_use</span> <span class="o">&gt;</span> <span class="n">n_gpu</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Warning: The number of GPU&#39;s configured to use is </span><span class="si">{</span><span class="n">n_gpu_use</span><span class="si">}</span><span class="s2">, but only </span><span class="si">{</span><span class="n">n_gpu</span><span class="si">}</span><span class="s2"> are &quot;</span>
            <span class="s2">&quot;available on this machine.&quot;</span>
        <span class="p">)</span>
        <span class="n">n_gpu_use</span> <span class="o">=</span> <span class="n">n_gpu</span>
    <span class="n">device</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">device</span><span class="p">(</span><span class="s2">&quot;cuda:0&quot;</span> <span class="k">if</span> <span class="n">n_gpu_use</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="s2">&quot;cpu&quot;</span><span class="p">)</span>
    <span class="n">list_ids</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="n">n_gpu_use</span><span class="p">))</span>
    <span class="k">return</span> <span class="n">device</span><span class="p">,</span> <span class="n">list_ids</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="scprint.utils.utils.run_command" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">run_command</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>run_command runs a command in the shell and prints the output.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>command</code></b>
                  (<code>str</code>)
              –
              <div class="doc-md-description">
                <p>The command to be executed in the shell.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>int</code></b>              –
              <div class="doc-md-description">
                <p>The return code of the command executed.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
            <details class="quote">
              <summary>Source code in <code>scprint/utils/utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">run_command</span><span class="p">(</span><span class="n">command</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    run_command runs a command in the shell and prints the output.</span>

<span class="sd">    Args:</span>
<span class="sd">        command (str): The command to be executed in the shell.</span>

<span class="sd">    Returns:</span>
<span class="sd">        int: The return code of the command executed.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">process</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">Popen</span><span class="p">(</span><span class="n">command</span><span class="p">,</span> <span class="n">stdout</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
    <span class="k">while</span> <span class="kc">True</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">process</span><span class="o">.</span><span class="n">poll</span><span class="p">()</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">break</span>
        <span class="n">output</span> <span class="o">=</span> <span class="n">process</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">readline</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">output</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="n">output</span><span class="o">.</span><span class="n">strip</span><span class="p">())</span>
    <span class="n">rc</span> <span class="o">=</span> <span class="n">process</span><span class="o">.</span><span class="n">poll</span><span class="p">()</span>
    <span class="k">return</span> <span class="n">rc</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="scprint.utils.utils.set_seed" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">set_seed</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>set random seed.</p>

            <details class="quote">
              <summary>Source code in <code>scprint/utils/utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">set_seed</span><span class="p">(</span><span class="n">seed</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">42</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;set random seed.&quot;&quot;&quot;</span>
    <span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="n">seed</span><span class="p">)</span>
    <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="n">seed</span><span class="p">)</span>
    <span class="n">torch</span><span class="o">.</span><span class="n">manual_seed</span><span class="p">(</span><span class="n">seed</span><span class="p">)</span>
    <span class="n">torch</span><span class="o">.</span><span class="n">backends</span><span class="o">.</span><span class="n">cudnn</span><span class="o">.</span><span class="n">deterministic</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="n">torch</span><span class="o">.</span><span class="n">backends</span><span class="o">.</span><span class="n">cudnn</span><span class="o">.</span><span class="n">benchmark</span> <span class="o">=</span> <span class="kc">False</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-module">



<h2 id="scprint.utils.get_seq" class="doc doc-heading">
            <code>scprint.utils.get_seq</code>


</h2>

    <div class="doc doc-contents first">



  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="scprint.utils.get_seq.load_fasta_species" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">load_fasta_species</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Downloads and caches FASTA files for a given species from the Ensembl FTP server.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>species</code></b>
                  (<code>str</code>, default:
                      <code>&#39;homo_sapiens&#39;</code>
)
              –
              <div class="doc-md-description">
                <p>The species name for which to download FASTA files. Defaults to "homo_sapiens".</p>
              </div>
            </li>
            <li>
              <b><code>output_path</code></b>
                  (<code>str</code>, default:
                      <code>&#39;/tmp/data/fasta/&#39;</code>
)
              –
              <div class="doc-md-description">
                <p>The local directory path where the FASTA files will be saved. Defaults to "/tmp/data/fasta/".</p>
              </div>
            </li>
            <li>
              <b><code>cache</code></b>
                  (<code>bool</code>, default:
                      <code>True</code>
)
              –
              <div class="doc-md-description">
                <p>If True, use cached files if they exist. If False, re-download the files. Defaults to True.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
            <details class="quote">
              <summary>Source code in <code>scprint/utils/get_seq.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">load_fasta_species</span><span class="p">(</span>
    <span class="n">species</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;homo_sapiens&quot;</span><span class="p">,</span>
    <span class="n">output_path</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;/tmp/data/fasta/&quot;</span><span class="p">,</span>
    <span class="n">cache</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Downloads and caches FASTA files for a given species from the Ensembl FTP server.</span>

<span class="sd">    Args:</span>
<span class="sd">        species (str, optional): The species name for which to download FASTA files. Defaults to &quot;homo_sapiens&quot;.</span>
<span class="sd">        output_path (str, optional): The local directory path where the FASTA files will be saved. Defaults to &quot;/tmp/data/fasta/&quot;.</span>
<span class="sd">        cache (bool, optional): If True, use cached files if they exist. If False, re-download the files. Defaults to True.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">ftp</span> <span class="o">=</span> <span class="n">ftplib</span><span class="o">.</span><span class="n">FTP</span><span class="p">(</span><span class="s2">&quot;ftp.ensembl.org&quot;</span><span class="p">)</span>
    <span class="n">ftp</span><span class="o">.</span><span class="n">login</span><span class="p">()</span>
    <span class="n">ftp</span><span class="o">.</span><span class="n">cwd</span><span class="p">(</span><span class="s2">&quot;/pub/release-110/fasta/&quot;</span> <span class="o">+</span> <span class="n">species</span> <span class="o">+</span> <span class="s2">&quot;/pep/&quot;</span><span class="p">)</span>
    <span class="n">file</span> <span class="o">=</span> <span class="n">list_files</span><span class="p">(</span><span class="n">ftp</span><span class="p">,</span> <span class="s2">&quot;.all.fa.gz&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">local_file_path</span> <span class="o">=</span> <span class="n">output_path</span> <span class="o">+</span> <span class="n">file</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">local_file_path</span><span class="p">)</span> <span class="ow">or</span> <span class="ow">not</span> <span class="n">cache</span><span class="p">:</span>
        <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">local_file_path</span><span class="p">),</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">local_file_path</span><span class="p">,</span> <span class="s2">&quot;wb&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">local_file</span><span class="p">:</span>
            <span class="n">ftp</span><span class="o">.</span><span class="n">retrbinary</span><span class="p">(</span><span class="s2">&quot;RETR &quot;</span> <span class="o">+</span> <span class="n">file</span><span class="p">,</span> <span class="n">local_file</span><span class="o">.</span><span class="n">write</span><span class="p">)</span>
    <span class="n">ftp</span><span class="o">.</span><span class="n">cwd</span><span class="p">(</span><span class="s2">&quot;/pub/release-110/fasta/&quot;</span> <span class="o">+</span> <span class="n">species</span> <span class="o">+</span> <span class="s2">&quot;/ncrna/&quot;</span><span class="p">)</span>
    <span class="n">file</span> <span class="o">=</span> <span class="n">list_files</span><span class="p">(</span><span class="n">ftp</span><span class="p">,</span> <span class="s2">&quot;.ncrna.fa.gz&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">local_file_path</span> <span class="o">=</span> <span class="n">output_path</span> <span class="o">+</span> <span class="n">file</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">local_file_path</span><span class="p">)</span> <span class="ow">or</span> <span class="ow">not</span> <span class="n">cache</span><span class="p">:</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">local_file_path</span><span class="p">,</span> <span class="s2">&quot;wb&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">local_file</span><span class="p">:</span>
            <span class="n">ftp</span><span class="o">.</span><span class="n">retrbinary</span><span class="p">(</span><span class="s2">&quot;RETR &quot;</span> <span class="o">+</span> <span class="n">file</span><span class="p">,</span> <span class="n">local_file</span><span class="o">.</span><span class="n">write</span><span class="p">)</span>
    <span class="n">ftp</span><span class="o">.</span><span class="n">quit</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="scprint.utils.get_seq.seq" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">seq</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Fetch nucleotide or amino acid sequence (FASTA) of a gene (and all its isoforms) or transcript by Ensembl, WormBase, or FlyBase ID.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>ens_ids</code></b>
                  (<code><span title="typing.Union">Union</span>[str, <span title="typing.List">List</span>[str]]</code>)
              –
              <div class="doc-md-description">
                <p>One or more Ensembl IDs (passed as string or list of strings).
                             Also supports WormBase and FlyBase IDs.</p>
              </div>
            </li>
            <li>
              <b><code>translate</code></b>
                  (<code>bool</code>, default:
                      <code>False</code>
)
              –
              <div class="doc-md-description">
                <p>Defines whether nucleotide or amino acid sequences are returned.
                        Defaults to False (returns nucleotide sequences).
                        Nucleotide sequences are fetched from the Ensembl REST API server.
                        Amino acid sequences are fetched from the UniProt REST API server.</p>
              </div>
            </li>
            <li>
              <b><code>isoforms</code></b>
                  (<code>bool</code>, default:
                      <code>False</code>
)
              –
              <div class="doc-md-description">
                <p>If True, returns the sequences of all known transcripts. Defaults to False.
                       (Only for gene IDs.)</p>
              </div>
            </li>
            <li>
              <b><code>parallel</code></b>
                  (<code>bool</code>, default:
                      <code>True</code>
)
              –
              <div class="doc-md-description">
                <p>If True, fetches sequences in parallel. Defaults to True.</p>
              </div>
            </li>
            <li>
              <b><code>save</code></b>
                  (<code>bool</code>, default:
                      <code>False</code>
)
              –
              <div class="doc-md-description">
                <p>If True, saves output FASTA to current directory. Defaults to False.</p>
              </div>
            </li>
            <li>
              <b><code>transcribe</code></b>
                  (<code>bool</code>, default:
                      <code>None</code>
)
              –
              <div class="doc-md-description">
                <p>Deprecated. Use 'translate' instead.</p>
              </div>
            </li>
            <li>
              <b><code>seqtype</code></b>
                  (<code>str</code>, default:
                      <code>None</code>
)
              –
              <div class="doc-md-description">
                <p>Deprecated. Use 'translate' instead.</p>
              </div>
            </li>
            <li>
              <b><code>verbose</code></b>
                  (<code>bool</code>, default:
                      <code>True</code>
)
              –
              <div class="doc-md-description">
                <p>If True, prints progress information. Defaults to True.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
                  <code><span title="typing.List">List</span>[str]</code>
              –
              <div class="doc-md-description">
                <p>List[str]: A list containing the requested sequences, or a FASTA file if 'save' is True.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Raises:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
                  <code>ValueError</code>
              –
              <div class="doc-md-description">
                <p>If an invalid Ensembl ID is provided.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
            <details class="quote">
              <summary>Source code in <code>scprint/utils/get_seq.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span>
<span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span>
<span class="normal">387</span>
<span class="normal">388</span>
<span class="normal">389</span>
<span class="normal">390</span>
<span class="normal">391</span>
<span class="normal">392</span>
<span class="normal">393</span>
<span class="normal">394</span>
<span class="normal">395</span>
<span class="normal">396</span>
<span class="normal">397</span>
<span class="normal">398</span>
<span class="normal">399</span>
<span class="normal">400</span>
<span class="normal">401</span>
<span class="normal">402</span>
<span class="normal">403</span>
<span class="normal">404</span>
<span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span>
<span class="normal">408</span>
<span class="normal">409</span>
<span class="normal">410</span>
<span class="normal">411</span>
<span class="normal">412</span>
<span class="normal">413</span>
<span class="normal">414</span>
<span class="normal">415</span>
<span class="normal">416</span>
<span class="normal">417</span>
<span class="normal">418</span>
<span class="normal">419</span>
<span class="normal">420</span>
<span class="normal">421</span>
<span class="normal">422</span>
<span class="normal">423</span>
<span class="normal">424</span>
<span class="normal">425</span>
<span class="normal">426</span>
<span class="normal">427</span>
<span class="normal">428</span>
<span class="normal">429</span>
<span class="normal">430</span>
<span class="normal">431</span>
<span class="normal">432</span>
<span class="normal">433</span>
<span class="normal">434</span>
<span class="normal">435</span>
<span class="normal">436</span>
<span class="normal">437</span>
<span class="normal">438</span>
<span class="normal">439</span>
<span class="normal">440</span>
<span class="normal">441</span>
<span class="normal">442</span>
<span class="normal">443</span>
<span class="normal">444</span>
<span class="normal">445</span>
<span class="normal">446</span>
<span class="normal">447</span>
<span class="normal">448</span>
<span class="normal">449</span>
<span class="normal">450</span>
<span class="normal">451</span>
<span class="normal">452</span>
<span class="normal">453</span>
<span class="normal">454</span>
<span class="normal">455</span>
<span class="normal">456</span>
<span class="normal">457</span>
<span class="normal">458</span>
<span class="normal">459</span>
<span class="normal">460</span>
<span class="normal">461</span>
<span class="normal">462</span>
<span class="normal">463</span>
<span class="normal">464</span>
<span class="normal">465</span>
<span class="normal">466</span>
<span class="normal">467</span>
<span class="normal">468</span>
<span class="normal">469</span>
<span class="normal">470</span>
<span class="normal">471</span>
<span class="normal">472</span>
<span class="normal">473</span>
<span class="normal">474</span>
<span class="normal">475</span>
<span class="normal">476</span>
<span class="normal">477</span>
<span class="normal">478</span>
<span class="normal">479</span>
<span class="normal">480</span>
<span class="normal">481</span>
<span class="normal">482</span>
<span class="normal">483</span>
<span class="normal">484</span>
<span class="normal">485</span>
<span class="normal">486</span>
<span class="normal">487</span>
<span class="normal">488</span>
<span class="normal">489</span>
<span class="normal">490</span>
<span class="normal">491</span>
<span class="normal">492</span>
<span class="normal">493</span>
<span class="normal">494</span>
<span class="normal">495</span>
<span class="normal">496</span>
<span class="normal">497</span>
<span class="normal">498</span>
<span class="normal">499</span>
<span class="normal">500</span>
<span class="normal">501</span>
<span class="normal">502</span>
<span class="normal">503</span>
<span class="normal">504</span>
<span class="normal">505</span>
<span class="normal">506</span>
<span class="normal">507</span>
<span class="normal">508</span>
<span class="normal">509</span>
<span class="normal">510</span>
<span class="normal">511</span>
<span class="normal">512</span>
<span class="normal">513</span>
<span class="normal">514</span>
<span class="normal">515</span>
<span class="normal">516</span>
<span class="normal">517</span>
<span class="normal">518</span>
<span class="normal">519</span>
<span class="normal">520</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">seq</span><span class="p">(</span>
    <span class="n">ens_ids</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]],</span>
    <span class="n">translate</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">isoforms</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">parallel</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
    <span class="n">save</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">transcribe</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">bool</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">seqtype</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">verbose</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Fetch nucleotide or amino acid sequence (FASTA) of a gene (and all its isoforms) or transcript by Ensembl, WormBase, or FlyBase ID.</span>

<span class="sd">    Args:</span>
<span class="sd">        ens_ids (Union[str, List[str]]): One or more Ensembl IDs (passed as string or list of strings).</span>
<span class="sd">                                         Also supports WormBase and FlyBase IDs.</span>
<span class="sd">        translate (bool, optional): Defines whether nucleotide or amino acid sequences are returned.</span>
<span class="sd">                                    Defaults to False (returns nucleotide sequences).</span>
<span class="sd">                                    Nucleotide sequences are fetched from the Ensembl REST API server.</span>
<span class="sd">                                    Amino acid sequences are fetched from the UniProt REST API server.</span>
<span class="sd">        isoforms (bool, optional): If True, returns the sequences of all known transcripts. Defaults to False.</span>
<span class="sd">                                   (Only for gene IDs.)</span>
<span class="sd">        parallel (bool, optional): If True, fetches sequences in parallel. Defaults to True.</span>
<span class="sd">        save (bool, optional): If True, saves output FASTA to current directory. Defaults to False.</span>
<span class="sd">        transcribe (bool, optional): Deprecated. Use &#39;translate&#39; instead.</span>
<span class="sd">        seqtype (str, optional): Deprecated. Use &#39;translate&#39; instead.</span>
<span class="sd">        verbose (bool, optional): If True, prints progress information. Defaults to True.</span>

<span class="sd">    Returns:</span>
<span class="sd">        List[str]: A list containing the requested sequences, or a FASTA file if &#39;save&#39; is True.</span>

<span class="sd">    Raises:</span>
<span class="sd">        ValueError: If an invalid Ensembl ID is provided.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Handle deprecated arguments</span>
    <span class="k">if</span> <span class="n">seqtype</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span>
            <span class="s2">&quot;&#39;seqtype&#39; argument deprecated! Please use True/False argument &#39;translate&#39; instead.&quot;</span>
        <span class="p">)</span>
        <span class="k">return</span>
    <span class="k">if</span> <span class="n">transcribe</span><span class="p">:</span>
        <span class="n">translate</span> <span class="o">=</span> <span class="n">transcribe</span>

    <span class="c1">## Clean up arguments</span>
    <span class="c1"># Clean up Ensembl IDs</span>
    <span class="c1"># If single Ensembl ID passed as string, convert to list</span>
    <span class="k">if</span> <span class="nb">type</span><span class="p">(</span><span class="n">ens_ids</span><span class="p">)</span> <span class="o">==</span> <span class="nb">str</span><span class="p">:</span>
        <span class="n">ens_ids</span> <span class="o">=</span> <span class="p">[</span><span class="n">ens_ids</span><span class="p">]</span>
    <span class="c1"># Remove Ensembl ID version if passed</span>
    <span class="n">ens_ids_clean</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">temp</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">for</span> <span class="n">ensembl_ID</span> <span class="ow">in</span> <span class="n">ens_ids</span><span class="p">:</span>
        <span class="c1"># But only for Ensembl ID (and not for flybase/wormbase IDs)</span>
        <span class="k">if</span> <span class="n">ensembl_ID</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;ENS&quot;</span><span class="p">):</span>
            <span class="n">ens_ids_clean</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ensembl_ID</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">])</span>

            <span class="k">if</span> <span class="s2">&quot;.&quot;</span> <span class="ow">in</span> <span class="n">ensembl_ID</span> <span class="ow">and</span> <span class="n">temp</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
                    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                        <span class="s2">&quot;We noticed that you may have passed a version number with your Ensembl ID.</span><span class="se">\n</span><span class="s2">&quot;</span>
                        <span class="s2">&quot;Please note that gget seq will return information linked to the latest Ensembl ID version.&quot;</span>
                    <span class="p">)</span>
                <span class="n">temp</span> <span class="o">=</span> <span class="o">+</span><span class="mi">1</span>

        <span class="k">else</span><span class="p">:</span>
            <span class="n">ens_ids_clean</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ensembl_ID</span><span class="p">)</span>

    <span class="c1"># Initiate empty &#39;fasta&#39;</span>
    <span class="n">fasta</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="c1">## Fetch nucleotide sequece</span>
    <span class="k">if</span> <span class="n">translate</span> <span class="ow">is</span> <span class="kc">False</span><span class="p">:</span>
        <span class="c1"># Define Ensembl REST API server</span>
        <span class="n">server</span> <span class="o">=</span> <span class="n">ENSEMBL_REST_API</span>
        <span class="c1"># Define type of returned content from REST</span>
        <span class="n">content_type</span> <span class="o">=</span> <span class="s2">&quot;application/json&quot;</span>

        <span class="c1"># Initiate dictionary to save results for all IDs in</span>
        <span class="n">master_dict</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="c1"># Query REST APIs from https://rest.ensembl.org/</span>
        <span class="k">for</span> <span class="n">ensembl_ID</span> <span class="ow">in</span> <span class="n">ens_ids_clean</span><span class="p">:</span>
            <span class="c1"># Create dict to save query results</span>
            <span class="n">results_dict</span> <span class="o">=</span> <span class="p">{</span><span class="n">ensembl_ID</span><span class="p">:</span> <span class="p">{}}</span>

            <span class="c1"># If isoforms False, just fetch sequences of passed Ensembl ID</span>
            <span class="k">if</span> <span class="n">isoforms</span> <span class="o">==</span> <span class="kc">False</span><span class="p">:</span>
                <span class="c1"># sequence/id/ query: Request sequence by stable identifier</span>
                <span class="n">query</span> <span class="o">=</span> <span class="s2">&quot;sequence/id/&quot;</span> <span class="o">+</span> <span class="n">ensembl_ID</span> <span class="o">+</span> <span class="s2">&quot;?&quot;</span>

                <span class="c1"># Try if query valid</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="c1"># Submit query; this will throw RuntimeError if ID not found</span>
                    <span class="n">df_temp</span> <span class="o">=</span> <span class="n">rest_query</span><span class="p">(</span><span class="n">server</span><span class="p">,</span> <span class="n">query</span><span class="p">,</span> <span class="n">content_type</span><span class="p">)</span>

                    <span class="c1"># Delete superfluous entries</span>
                    <span class="n">keys_to_delete</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;query&quot;</span><span class="p">,</span> <span class="s2">&quot;id&quot;</span><span class="p">,</span> <span class="s2">&quot;version&quot;</span><span class="p">,</span> <span class="s2">&quot;molecule&quot;</span><span class="p">]</span>
                    <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">keys_to_delete</span><span class="p">:</span>
                        <span class="c1"># Pop keys, None -&gt; do not raise an error if key to delete not found</span>
                        <span class="n">df_temp</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="n">key</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>

                    <span class="c1"># Add results to main dict</span>
                    <span class="n">results_dict</span><span class="p">[</span><span class="n">ensembl_ID</span><span class="p">]</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="s2">&quot;seq&quot;</span><span class="p">:</span> <span class="n">df_temp</span><span class="p">})</span>

                    <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
                        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                            <span class="sa">f</span><span class="s2">&quot;Requesting nucleotide sequence of </span><span class="si">{</span><span class="n">ensembl_ID</span><span class="si">}</span><span class="s2"> from Ensembl.&quot;</span>
                        <span class="p">)</span>

                <span class="k">except</span> <span class="ne">RuntimeError</span><span class="p">:</span>
                    <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span>
                        <span class="sa">f</span><span class="s2">&quot;ID </span><span class="si">{</span><span class="n">ensembl_ID</span><span class="si">}</span><span class="s2"> not found. Please double-check spelling/arguments and try again.&quot;</span>
                    <span class="p">)</span>

            <span class="c1"># If isoforms true, fetch sequences of isoforms instead</span>
            <span class="k">if</span> <span class="n">isoforms</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span>
                <span class="c1"># Get ID type (gene, transcript, ...) using gget info</span>
                <span class="n">info_df</span> <span class="o">=</span> <span class="n">info</span><span class="p">(</span>
                    <span class="n">ensembl_ID</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">pdb</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">ncbi</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">uniprot</span><span class="o">=</span><span class="kc">False</span>
                <span class="p">)</span>

                <span class="c1"># Check if Ensembl ID was found</span>
                <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">info_df</span><span class="p">,</span> <span class="nb">type</span><span class="p">(</span><span class="kc">None</span><span class="p">)):</span>
                    <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                        <span class="sa">f</span><span class="s2">&quot;ID &#39;</span><span class="si">{</span><span class="n">ensembl_ID</span><span class="si">}</span><span class="s2">&#39; not found. Please double-check spelling/arguments and try again.&quot;</span>
                    <span class="p">)</span>
                    <span class="k">continue</span>

                <span class="n">ens_ID_type</span> <span class="o">=</span> <span class="n">info_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ensembl_ID</span><span class="p">][</span><span class="s2">&quot;object_type&quot;</span><span class="p">]</span>

                <span class="c1"># If the ID is a gene, get the IDs of all its transcripts</span>
                <span class="k">if</span> <span class="n">ens_ID_type</span> <span class="o">==</span> <span class="s2">&quot;Gene&quot;</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
                        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                            <span class="sa">f</span><span class="s2">&quot;Requesting nucleotide sequences of all transcripts of </span><span class="si">{</span><span class="n">ensembl_ID</span><span class="si">}</span><span class="s2"> from Ensembl.&quot;</span>
                        <span class="p">)</span>

                    <span class="k">for</span> <span class="n">transcipt_id</span> <span class="ow">in</span> <span class="n">info_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ensembl_ID</span><span class="p">][</span><span class="s2">&quot;all_transcripts&quot;</span><span class="p">]:</span>
                        <span class="c1"># Remove version number for Ensembl IDs (not for flybase/wormbase IDs)</span>
                        <span class="k">if</span> <span class="n">transcipt_id</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;ENS&quot;</span><span class="p">):</span>
                            <span class="n">transcipt_id</span> <span class="o">=</span> <span class="n">transcipt_id</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>

                        <span class="c1"># Try if query is valid</span>
                        <span class="k">try</span><span class="p">:</span>
                            <span class="c1"># Define the REST query</span>
                            <span class="n">query</span> <span class="o">=</span> <span class="s2">&quot;sequence/id/&quot;</span> <span class="o">+</span> <span class="n">transcipt_id</span> <span class="o">+</span> <span class="s2">&quot;?&quot;</span>
                            <span class="c1"># Submit query</span>
                            <span class="n">df_temp</span> <span class="o">=</span> <span class="n">rest_query</span><span class="p">(</span><span class="n">server</span><span class="p">,</span> <span class="n">query</span><span class="p">,</span> <span class="n">content_type</span><span class="p">)</span>

                            <span class="c1"># Delete superfluous entries</span>
                            <span class="n">keys_to_delete</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;query&quot;</span><span class="p">,</span> <span class="s2">&quot;version&quot;</span><span class="p">,</span> <span class="s2">&quot;molecule&quot;</span><span class="p">]</span>
                            <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">keys_to_delete</span><span class="p">:</span>
                                <span class="c1"># Pop keys, None -&gt; do not raise an error if key to delete not found</span>
                                <span class="n">df_temp</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="n">key</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>

                            <span class="c1"># Add results to main dict</span>
                            <span class="n">results_dict</span><span class="p">[</span><span class="n">ensembl_ID</span><span class="p">]</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
                                <span class="p">{</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">transcipt_id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">:</span> <span class="n">df_temp</span><span class="p">}</span>
                            <span class="p">)</span>

                        <span class="k">except</span> <span class="ne">RuntimeError</span><span class="p">:</span>
                            <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span>
                                <span class="sa">f</span><span class="s2">&quot;ID </span><span class="si">{</span><span class="n">transcipt_id</span><span class="si">}</span><span class="s2"> not found. &quot;</span>
                                <span class="s2">&quot;Please double-check spelling/arguments and try again.&quot;</span>
                            <span class="p">)</span>

                <span class="c1"># If isoform true, but ID is not a gene; ignore the isoform parameter</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="c1"># Try if query is valid</span>
                    <span class="k">try</span><span class="p">:</span>
                        <span class="c1"># Define the REST query</span>
                        <span class="n">query</span> <span class="o">=</span> <span class="s2">&quot;sequence/id/&quot;</span> <span class="o">+</span> <span class="n">ensembl_ID</span> <span class="o">+</span> <span class="s2">&quot;?&quot;</span>

                        <span class="c1"># Submit query</span>
                        <span class="n">df_temp</span> <span class="o">=</span> <span class="n">rest_query</span><span class="p">(</span><span class="n">server</span><span class="p">,</span> <span class="n">query</span><span class="p">,</span> <span class="n">content_type</span><span class="p">)</span>

                        <span class="c1"># Delete superfluous entries</span>
                        <span class="n">keys_to_delete</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;query&quot;</span><span class="p">,</span> <span class="s2">&quot;id&quot;</span><span class="p">,</span> <span class="s2">&quot;version&quot;</span><span class="p">,</span> <span class="s2">&quot;molecule&quot;</span><span class="p">]</span>
                        <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">keys_to_delete</span><span class="p">:</span>
                            <span class="c1"># Pop keys, None -&gt; do not raise an error if key to delete not found</span>
                            <span class="n">df_temp</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="n">key</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>

                        <span class="c1"># Add results to main dict</span>
                        <span class="n">results_dict</span><span class="p">[</span><span class="n">ensembl_ID</span><span class="p">]</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="s2">&quot;seq&quot;</span><span class="p">:</span> <span class="n">df_temp</span><span class="p">})</span>

                        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                            <span class="sa">f</span><span class="s2">&quot;Requesting nucleotide sequence of </span><span class="si">{</span><span class="n">ensembl_ID</span><span class="si">}</span><span class="s2"> from Ensembl.&quot;</span>
                        <span class="p">)</span>
                        <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;The isoform option only applies to gene IDs.&quot;</span><span class="p">)</span>

                    <span class="k">except</span> <span class="ne">RuntimeError</span><span class="p">:</span>
                        <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span>
                            <span class="sa">f</span><span class="s2">&quot;ID </span><span class="si">{</span><span class="n">ensembl_ID</span><span class="si">}</span><span class="s2"> not found. &quot;</span>
                            <span class="s2">&quot;Please double-check spelling/arguments and try again.&quot;</span>
                        <span class="p">)</span>

            <span class="c1"># Add results to master dict</span>
            <span class="n">master_dict</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">results_dict</span><span class="p">)</span>

        <span class="c1"># Build FASTA file</span>
        <span class="k">for</span> <span class="n">ens_ID</span> <span class="ow">in</span> <span class="n">master_dict</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">master_dict</span><span class="p">[</span><span class="n">ens_ID</span><span class="p">]</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                <span class="k">if</span> <span class="n">key</span> <span class="o">==</span> <span class="s2">&quot;seq&quot;</span><span class="p">:</span>
                    <span class="n">fasta</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;&gt;&quot;</span> <span class="o">+</span> <span class="n">ens_ID</span> <span class="o">+</span> <span class="s2">&quot; &quot;</span> <span class="o">+</span> <span class="n">master_dict</span><span class="p">[</span><span class="n">ens_ID</span><span class="p">][</span><span class="n">key</span><span class="p">][</span><span class="s2">&quot;desc&quot;</span><span class="p">])</span>
                    <span class="n">fasta</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">master_dict</span><span class="p">[</span><span class="n">ens_ID</span><span class="p">][</span><span class="n">key</span><span class="p">][</span><span class="s2">&quot;seq&quot;</span><span class="p">])</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">fasta</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                        <span class="s2">&quot;&gt;&quot;</span>
                        <span class="o">+</span> <span class="n">master_dict</span><span class="p">[</span><span class="n">ens_ID</span><span class="p">][</span><span class="n">key</span><span class="p">][</span><span class="s2">&quot;id&quot;</span><span class="p">]</span>
                        <span class="o">+</span> <span class="s2">&quot; &quot;</span>
                        <span class="o">+</span> <span class="n">master_dict</span><span class="p">[</span><span class="n">ens_ID</span><span class="p">][</span><span class="n">key</span><span class="p">][</span><span class="s2">&quot;desc&quot;</span><span class="p">]</span>
                    <span class="p">)</span>
                    <span class="n">fasta</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">master_dict</span><span class="p">[</span><span class="n">ens_ID</span><span class="p">][</span><span class="n">key</span><span class="p">][</span><span class="s2">&quot;seq&quot;</span><span class="p">])</span>

    <span class="c1">## Fetch amino acid sequences from UniProt</span>
    <span class="k">if</span> <span class="n">translate</span> <span class="ow">is</span> <span class="kc">True</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">isoforms</span> <span class="ow">is</span> <span class="kc">False</span><span class="p">:</span>
            <span class="c1"># List to collect transcript IDs</span>
            <span class="n">trans_ids</span> <span class="o">=</span> <span class="p">[]</span>

            <span class="c1"># Get ID type (gene, transcript, ...) using gget info</span>
            <span class="n">info_df</span> <span class="o">=</span> <span class="n">info</span><span class="p">(</span>
                <span class="n">ens_ids_clean</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">pdb</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">ncbi</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">uniprot</span><span class="o">=</span><span class="kc">False</span>
            <span class="p">)</span>

            <span class="c1"># Check that Ensembl ID was found</span>
            <span class="n">missing</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">ens_ids_clean</span><span class="p">)</span> <span class="o">-</span> <span class="nb">set</span><span class="p">(</span><span class="n">info_df</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">missing</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="n">missing</span><span class="p">)</span><span class="si">}</span><span class="s2"> IDs not found. Please double-check spelling/arguments.&quot;</span>
                <span class="p">)</span>

            <span class="n">ens_ID_type</span> <span class="o">=</span> <span class="n">info_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ens_ids_clean</span><span class="p">[</span><span class="mi">0</span><span class="p">]][</span><span class="s2">&quot;object_type&quot;</span><span class="p">]</span>

            <span class="c1"># If the ID is a gene, use the ID of its canonical transcript</span>
            <span class="k">if</span> <span class="n">ens_ID_type</span> <span class="o">==</span> <span class="s2">&quot;Gene&quot;</span><span class="p">:</span>
                <span class="c1"># Get ID of canonical transcript</span>
                <span class="k">for</span> <span class="n">ensembl_ID</span> <span class="ow">in</span> <span class="n">info_df</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">values</span><span class="p">:</span>
                    <span class="n">can_trans</span> <span class="o">=</span> <span class="n">info_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ensembl_ID</span><span class="p">][</span><span class="s2">&quot;canonical_transcript&quot;</span><span class="p">]</span>

                    <span class="k">if</span> <span class="n">ensembl_ID</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;ENS&quot;</span><span class="p">):</span>
                        <span class="c1"># Remove Ensembl ID version from transcript IDs and append to transcript IDs list</span>
                        <span class="n">temp_trans_id</span> <span class="o">=</span> <span class="n">can_trans</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
                        <span class="n">trans_ids</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">temp_trans_id</span><span class="p">)</span>

                    <span class="k">elif</span> <span class="n">ensembl_ID</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;WB&quot;</span><span class="p">):</span>
                        <span class="c1"># Remove added &quot;.&quot; at the end of transcript IDs</span>
                        <span class="n">temp_trans_id</span> <span class="o">=</span> <span class="s2">&quot;.&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">can_trans</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">)[:</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                        <span class="c1"># # For WormBase transcript IDs, also remove the version number for submission to UniProt API</span>
                        <span class="c1"># temp_trans_id = &quot;.&quot;.join(temp_trans_id1.split(&quot;.&quot;)[:-1])</span>
                        <span class="n">trans_ids</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">temp_trans_id</span><span class="p">)</span>

                    <span class="k">else</span><span class="p">:</span>
                        <span class="c1"># Remove added &quot;.&quot; at the end of other transcript IDs</span>
                        <span class="n">temp_trans_id</span> <span class="o">=</span> <span class="s2">&quot;.&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">can_trans</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">)[:</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
                        <span class="n">trans_ids</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">temp_trans_id</span><span class="p">)</span>

                    <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
                        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                            <span class="sa">f</span><span class="s2">&quot;Requesting amino acid sequence of the canonical transcript </span><span class="si">{</span><span class="n">temp_trans_id</span><span class="si">}</span><span class="s2"> of gene </span><span class="si">{</span><span class="n">ensembl_ID</span><span class="si">}</span><span class="s2"> from UniProt.&quot;</span>
                        <span class="p">)</span>

            <span class="c1"># If the ID is a transcript, append the ID directly</span>
            <span class="k">elif</span> <span class="n">ens_ID_type</span> <span class="o">==</span> <span class="s2">&quot;Transcript&quot;</span><span class="p">:</span>
                <span class="c1"># # For WormBase transcript IDs, remove the version number for submission to UniProt API</span>
                <span class="c1"># if ensembl_ID.startswith(&quot;T&quot;):</span>
                <span class="c1">#     trans_ids.append(&quot;.&quot;.join(ensembl_ID.split(&quot;.&quot;)[:-1]))</span>
                <span class="c1"># else:</span>
                <span class="n">trans_ids</span> <span class="o">=</span> <span class="n">ens_ids_clean</span>

                <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
                    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                        <span class="sa">f</span><span class="s2">&quot;Requesting amino acid sequence of </span><span class="si">{</span><span class="n">trans_ids</span><span class="si">}</span><span class="s2"> from UniProt.&quot;</span>
                    <span class="p">)</span>

            <span class="k">else</span><span class="p">:</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;ensembl_IDs not recognized as either a gene or transcript ID. It will not be included in the UniProt query.&quot;</span>
                <span class="p">)</span>

            <span class="c1"># Fetch the amino acid sequences of the transcript Ensembl IDs</span>
            <span class="n">df_uniprot</span> <span class="o">=</span> <span class="n">get_uniprot_seqs</span><span class="p">(</span><span class="n">UNIPROT_REST_API</span><span class="p">,</span> <span class="n">trans_ids</span><span class="p">)</span>
            <span class="c1"># Add info_df.loc[ensembl_ID] to df_uniprot by joining on &quot;canonical_transcript&quot; / &quot;gene_name&quot; respectively</span>
            <span class="kn">import</span> <span class="nn">pdb</span>

            <span class="n">pdb</span><span class="o">.</span><span class="n">set_trace</span><span class="p">()</span>
            <span class="n">info_df</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s2">&quot;canonical_transcript&quot;</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

            <span class="n">df_uniprot</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="s2">&quot;gene_id&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">info_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span>
                <span class="n">df_uniprot</span><span class="p">[</span><span class="s2">&quot;query&quot;</span><span class="p">],</span> <span class="s2">&quot;gene_name&quot;</span>
            <span class="p">]</span><span class="o">.</span><span class="n">values</span>

        <span class="k">if</span> <span class="n">isoforms</span> <span class="ow">is</span> <span class="kc">True</span><span class="p">:</span>
            <span class="c1"># List to collect transcript IDs</span>
            <span class="n">trans_ids</span> <span class="o">=</span> <span class="p">[]</span>

            <span class="k">for</span> <span class="n">ensembl_ID</span> <span class="ow">in</span> <span class="n">ens_ids_clean</span><span class="p">:</span>
                <span class="c1"># Get ID type (gene, transcript, ...) using gget info</span>
                <span class="n">info_df</span> <span class="o">=</span> <span class="n">info</span><span class="p">(</span>
                    <span class="n">ensembl_ID</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">pdb</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">ncbi</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">uniprot</span><span class="o">=</span><span class="kc">False</span>
                <span class="p">)</span>

                <span class="c1"># Check that Ensembl ID was found</span>
                <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">info_df</span><span class="p">,</span> <span class="nb">type</span><span class="p">(</span><span class="kc">None</span><span class="p">)):</span>
                    <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                        <span class="sa">f</span><span class="s2">&quot;ID &#39;</span><span class="si">{</span><span class="n">ensembl_ID</span><span class="si">}</span><span class="s2">&#39; not found. Please double-check spelling/arguments.&quot;</span>
                    <span class="p">)</span>
                    <span class="k">continue</span>

                <span class="n">ens_ID_type</span> <span class="o">=</span> <span class="n">info_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ensembl_ID</span><span class="p">][</span><span class="s2">&quot;object_type&quot;</span><span class="p">]</span>

                <span class="c1"># If the ID is a gene, get the IDs of all isoforms</span>
                <span class="k">if</span> <span class="n">ens_ID_type</span> <span class="o">==</span> <span class="s2">&quot;Gene&quot;</span><span class="p">:</span>
                    <span class="c1"># Get the IDs of all transcripts from the gget info results</span>
                    <span class="k">for</span> <span class="n">transcipt_id</span> <span class="ow">in</span> <span class="n">info_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ensembl_ID</span><span class="p">][</span><span class="s2">&quot;all_transcripts&quot;</span><span class="p">]:</span>
                        <span class="k">if</span> <span class="n">ensembl_ID</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;ENS&quot;</span><span class="p">):</span>
                            <span class="c1"># Append transcript ID (without Ensembl version number) to list of transcripts to fetch</span>
                            <span class="n">trans_ids</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">transcipt_id</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">])</span>

                        <span class="c1"># elif ensembl_ID.startswith(&quot;WB&quot;):</span>
                        <span class="c1">#     # For WormBase transcript IDs, remove the version number for submission to UniProt API</span>
                        <span class="c1">#     temp_trans_id = &quot;.&quot;.join(transcipt_id.split(&quot;.&quot;)[:-1])</span>
                        <span class="c1">#     trans_ids.append(temp_trans_id)</span>

                        <span class="k">else</span><span class="p">:</span>
                            <span class="c1"># Note: No need to remove the added &quot;.&quot; at the end of unversioned transcripts here, because &quot;all_transcripts&quot; are returned without it</span>
                            <span class="n">trans_ids</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">transcipt_id</span><span class="p">)</span>

                    <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
                        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                            <span class="sa">f</span><span class="s2">&quot;Requesting amino acid sequences of all transcripts of gene </span><span class="si">{</span><span class="n">ensembl_ID</span><span class="si">}</span><span class="s2"> from UniProt.&quot;</span>
                        <span class="p">)</span>

                <span class="k">elif</span> <span class="n">ens_ID_type</span> <span class="o">==</span> <span class="s2">&quot;Transcript&quot;</span><span class="p">:</span>
                    <span class="c1"># # For WormBase transcript IDs, remove the version number for submission to UniProt API</span>
                    <span class="c1"># if ensembl_ID.startswith(&quot;T&quot;):</span>
                    <span class="c1">#     trans_ids.append(&quot;.&quot;.join(ensembl_ID.split(&quot;.&quot;)[:-1]))</span>

                    <span class="c1"># else:</span>
                    <span class="n">trans_ids</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ensembl_ID</span><span class="p">)</span>

                    <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
                        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                            <span class="sa">f</span><span class="s2">&quot;Requesting amino acid sequence of </span><span class="si">{</span><span class="n">ensembl_ID</span><span class="si">}</span><span class="s2"> from UniProt.&quot;</span>
                        <span class="p">)</span>
                    <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;The isoform option only applies to gene IDs.&quot;</span><span class="p">)</span>

                <span class="k">else</span><span class="p">:</span>
                    <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ensembl_ID</span><span class="si">}</span><span class="s2"> not recognized as either a gene or transcript ID. It will not be included in the UniProt query.&quot;</span>
                    <span class="p">)</span>

            <span class="c1"># Fetch amino acid sequences of all isoforms from the UniProt REST API</span>
            <span class="n">df_uniprot</span> <span class="o">=</span> <span class="n">get_uniprot_seqs</span><span class="p">(</span><span class="n">UNIPROT_REST_API</span><span class="p">,</span> <span class="n">trans_ids</span><span class="p">)</span>

        <span class="c1"># Check if any results were found</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">df_uniprot</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">1</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s2">&quot;No UniProt amino acid sequences were found for these ID(s).&quot;</span><span class="p">)</span>

        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># Build FASTA file from UniProt results</span>
            <span class="k">for</span> <span class="p">(</span>
                <span class="n">uniprot_id</span><span class="p">,</span>
                <span class="n">query_ensembl_id</span><span class="p">,</span>
                <span class="n">gene_name</span><span class="p">,</span>
                <span class="n">organism</span><span class="p">,</span>
                <span class="n">sequence_length</span><span class="p">,</span>
                <span class="n">uniprot_seq</span><span class="p">,</span>
            <span class="p">)</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span>
                <span class="n">df_uniprot</span><span class="p">[</span><span class="s2">&quot;uniprot_id&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">,</span>
                <span class="n">df_uniprot</span><span class="p">[</span><span class="s2">&quot;query&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">,</span>
                <span class="n">df_uniprot</span><span class="p">[</span><span class="s2">&quot;gene_name&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">,</span>
                <span class="n">df_uniprot</span><span class="p">[</span><span class="s2">&quot;gene_id&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">,</span>
                <span class="n">df_uniprot</span><span class="p">[</span><span class="s2">&quot;organism&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">,</span>
                <span class="n">df_uniprot</span><span class="p">[</span><span class="s2">&quot;sequence_length&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">,</span>
                <span class="n">df_uniprot</span><span class="p">[</span><span class="s2">&quot;sequence&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">,</span>
            <span class="p">):</span>
                <span class="n">fasta</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                    <span class="s2">&quot;&gt;&quot;</span>
                    <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">query_ensembl_id</span><span class="p">)</span>
                    <span class="o">+</span> <span class="s2">&quot; uniprot_id: &quot;</span>
                    <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">uniprot_id</span><span class="p">)</span>
                    <span class="o">+</span> <span class="s2">&quot; ensembl_id: &quot;</span>
                    <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">query_ensembl_id</span><span class="p">)</span>
                    <span class="o">+</span> <span class="s2">&quot; gene_name: &quot;</span>
                    <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">gene_name</span><span class="p">)</span>
                    <span class="o">+</span> <span class="s2">&quot; gene_id: &quot;</span>
                    <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">gene_id</span><span class="p">)</span>
                    <span class="o">+</span> <span class="s2">&quot; organism: &quot;</span>
                    <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">organism</span><span class="p">)</span>
                    <span class="o">+</span> <span class="s2">&quot; sequence_length: &quot;</span>
                    <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">sequence_length</span><span class="p">)</span>
                <span class="p">)</span>
                <span class="n">fasta</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">uniprot_seq</span><span class="p">))</span>

    <span class="c1"># Save</span>
    <span class="k">if</span> <span class="n">save</span><span class="p">:</span>
        <span class="n">file</span><span class="p">:</span> <span class="n">TextIOWrapper</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;gget_seq_results.fa&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">element</span> <span class="ow">in</span> <span class="n">fasta</span><span class="p">:</span>
            <span class="n">file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">element</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">file</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
        <span class="c1"># missed samples</span>
        <span class="k">return</span> <span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">trans_ids</span><span class="p">)</span> <span class="o">-</span> <span class="nb">set</span><span class="p">(</span><span class="n">df_uniprot</span><span class="p">[</span><span class="s2">&quot;query&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">))</span> <span class="o">|</span> <span class="nb">set</span><span class="p">(</span><span class="n">missing</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">fasta</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="scprint.utils.get_seq.subset_fasta" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">subset_fasta</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>subset_fasta: creates a new fasta file with only the sequence which names contain one of gene_names</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>gene_tosubset</code></b>
                  (<code>set</code>)
              –
              <div class="doc-md-description">
                <p>A set of gene names to subset from the original FASTA file.</p>
              </div>
            </li>
            <li>
              <b><code>fasta_path</code></b>
                  (<code>str</code>)
              –
              <div class="doc-md-description">
                <p>The path to the original FASTA file.</p>
              </div>
            </li>
            <li>
              <b><code>subfasta_path</code></b>
                  (<code>str</code>, default:
                      <code>&#39;./data/fasta/subset.fa&#39;</code>
)
              –
              <div class="doc-md-description">
                <p>The path to save the subsetted FASTA file. Defaults to "./data/fasta/subset.fa".</p>
              </div>
            </li>
            <li>
              <b><code>drop_unknown_seq</code></b>
                  (<code>bool</code>, default:
                      <code>True</code>
)
              –
              <div class="doc-md-description">
                <p>If True, drop sequences containing unknown amino acids (denoted by '*'). Defaults to True.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>set</code></b>(                  <code>set</code>
)              –
              <div class="doc-md-description">
                <p>A set of gene names that were found and included in the subsetted FASTA file.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Raises:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
                  <code>ValueError</code>
              –
              <div class="doc-md-description">
                <p>If a gene name does not start with "ENS".</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
            <details class="quote">
              <summary>Source code in <code>scprint/utils/get_seq.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">subset_fasta</span><span class="p">(</span>
    <span class="n">gene_tosubset</span><span class="p">:</span> <span class="nb">set</span><span class="p">,</span>
    <span class="n">fasta_path</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">subfasta_path</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;./data/fasta/subset.fa&quot;</span><span class="p">,</span>
    <span class="n">drop_unknown_seq</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">set</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    subset_fasta: creates a new fasta file with only the sequence which names contain one of gene_names</span>

<span class="sd">    Args:</span>
<span class="sd">        gene_tosubset (set): A set of gene names to subset from the original FASTA file.</span>
<span class="sd">        fasta_path (str): The path to the original FASTA file.</span>
<span class="sd">        subfasta_path (str, optional): The path to save the subsetted FASTA file. Defaults to &quot;./data/fasta/subset.fa&quot;.</span>
<span class="sd">        drop_unknown_seq (bool, optional): If True, drop sequences containing unknown amino acids (denoted by &#39;*&#39;). Defaults to True.</span>

<span class="sd">    Returns:</span>
<span class="sd">        set: A set of gene names that were found and included in the subsetted FASTA file.</span>

<span class="sd">    Raises:</span>
<span class="sd">        ValueError: If a gene name does not start with &quot;ENS&quot;.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">dup</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
    <span class="n">weird</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">genes_found</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
    <span class="n">gene_tosubset</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">gene_tosubset</span><span class="p">)</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">fasta_path</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">original_fasta</span><span class="p">,</span> <span class="nb">open</span><span class="p">(</span>
        <span class="n">subfasta_path</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span>
    <span class="p">)</span> <span class="k">as</span> <span class="n">subset_fasta</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">original_fasta</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">):</span>
            <span class="n">gene_name</span> <span class="o">=</span> <span class="p">(</span>
                <span class="n">record</span><span class="o">.</span><span class="n">description</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot; gene:&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span>
                <span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot; transcript&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
                <span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
            <span class="p">)</span>
            <span class="k">if</span> <span class="n">gene_name</span> <span class="ow">in</span> <span class="n">gene_tosubset</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">drop_unknown_seq</span><span class="p">:</span>
                    <span class="k">if</span> <span class="s2">&quot;*&quot;</span> <span class="ow">in</span> <span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="p">:</span>
                        <span class="n">weird</span> <span class="o">+=</span> <span class="mi">1</span>

                        <span class="k">continue</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">gene_name</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;ENS&quot;</span><span class="p">):</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;issue&quot;</span><span class="p">,</span> <span class="n">gene_name</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">gene_name</span> <span class="ow">in</span> <span class="n">genes_found</span><span class="p">:</span>
                    <span class="n">dup</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">gene_name</span><span class="p">)</span>
                    <span class="k">continue</span>
                <span class="n">record</span><span class="o">.</span><span class="n">description</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
                <span class="n">record</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="n">gene_name</span>
                <span class="n">SeqIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">record</span><span class="p">,</span> <span class="n">subset_fasta</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>
                <span class="n">genes_found</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">gene_name</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">dup</span><span class="p">),</span> <span class="s2">&quot; genes had duplicates&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;dropped&quot;</span><span class="p">,</span> <span class="n">weird</span><span class="p">,</span> <span class="s2">&quot;weird sequences&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">genes_found</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../embedder/" class="btn btn-neutral float-left" title="embedders"><span class="icon icon-circle-arrow-left"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
    
      <span><a href="../embedder/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
  </span>
</div>
    <script src="../js/jquery-3.6.0.min.js"></script>
    <script>var base_url = "..";</script>
    <script src="../js/theme_extra.js"></script>
    <script src="../js/theme.js"></script>
      <script src="../search/main.js"></script>
    <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>

</body>
</html>
